{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant disease phenotypes\n",
    "\n",
    "This analysis aimed to answer the following research question: **How many of the variants identified in African populations have known disease associations, and what are these associations?**\n",
    "\n",
    "To achieve this, the following steps were performed:\n",
    "\n",
    "1. Genetic Variant Phenotype Data Retrieval and Preparation: Known disease phenotypes for the variants were retrieved from [Favor v2.0](https://favor.genohub.org/). The retrieved data underwent processing and preparation following guidelines outlined in the [Notebooks\\Data_preparation\\6-Variant_phenotype_associations.ipynb](https://github.com/MeganHolborn/Genetic_data_analysis/blob/main/Notebooks/Data_preparation/6-Variant_phenotype_associations.ipynb) Jupyter notebook. The processed data can be found [here](https://github.com/MeganHolborn/Genetic_data_analysis/blob/main/Data/Processed/Variant_disease_phenotypes.csv).\n",
    "2. Analysis and Visualisation:\n",
    "    * To be completed..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import upsetplot\n",
    "from matplotlib import pyplot as plt\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant phenotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110149494_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149715</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>110149715_AAT_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110151168</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110151168_C_T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf         Rsid Clndn  CHROM        POS  REF ALT  \\\n",
       "0    13-110148917-C-G   rs59409892   NaN     13  110148917    C   G   \n",
       "1    13-110148891-C-G  rs552586867   NaN     13  110148891    C   G   \n",
       "2    13-110149494-C-T  rs552877576   NaN     13  110149494    C   T   \n",
       "3  13-110149715-AAT-A  rs886049952   NaN     13  110149715  AAT   A   \n",
       "4    13-110151168-C-T  rs557686466   NaN     13  110151168    C   T   \n",
       "\n",
       "                ID  \n",
       "0    110148917_C_G  \n",
       "1    110148891_C_G  \n",
       "2    110149494_C_T  \n",
       "3  110149715_AAT_A  \n",
       "4    110151168_C_T  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_disease_phenotypes.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "phenotype_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import genetic variant count data for African populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Bantu Kenya</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>EA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>0</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>San</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>SA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mende</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>166</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mbuti Pygmy</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID             VAR_NAME        POS REF ALT    GENE  \\\n",
       "0  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "1  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "2  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "3  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "4  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "\n",
       "       SUB_POP  IH_ALT_CTS  IH_TOTAL_CTS  IH_REF_CTS REG  IH_AF VARIANT_TYPE  \n",
       "0  Bantu Kenya           0            20          20  EA    0.0        INDEL  \n",
       "1       Yoruba           0           276         276  WA    0.0        INDEL  \n",
       "2          San           0            12          12  SA    0.0        INDEL  \n",
       "3        Mende           0           166         166  WA    0.0        INDEL  \n",
       "4  Mbuti Pygmy           0            24          24  CA    0.0        INDEL  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_afr = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "ih_afr.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant effect data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POLYPHEN_PRED</th>\n",
       "      <th>POLYPHEN_SCORE</th>\n",
       "      <th>SIFT_PRED</th>\n",
       "      <th>SIFT_SCORE</th>\n",
       "      <th>CADD_RAW_SCORE</th>\n",
       "      <th>CADD_PHRED_SCORE</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.437825</td>\n",
       "      <td>0.160</td>\n",
       "      <td>110148882_C_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.227221</td>\n",
       "      <td>0.446</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.269936</td>\n",
       "      <td>3.938</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.530972</td>\n",
       "      <td>6.825</td>\n",
       "      <td>110148920_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.380228</td>\n",
       "      <td>14.950</td>\n",
       "      <td>110148959_A_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT    GENE POLYPHEN_PRED  POLYPHEN_SCORE SIFT_PRED  \\\n",
       "0     13  110148882   C  CT  COL4A1           NaN             NaN       NaN   \n",
       "1     13  110148891   C   G  COL4A1           NaN             NaN       NaN   \n",
       "2     13  110148917   C   G  COL4A1           NaN             NaN       NaN   \n",
       "3     13  110148920   G   C  COL4A1           NaN             NaN       NaN   \n",
       "4     13  110148959   A   G  COL4A1           NaN             NaN       NaN   \n",
       "\n",
       "   SIFT_SCORE  CADD_RAW_SCORE  CADD_PHRED_SCORE              ID  \n",
       "0         NaN       -0.437825             0.160  110148882_C_CT  \n",
       "1         NaN       -0.227221             0.446   110148891_C_G  \n",
       "2         NaN        0.269936             3.938   110148917_C_G  \n",
       "3         NaN        0.530972             6.825   110148920_G_C  \n",
       "4         NaN        1.380228            14.950   110148959_A_G  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_effects.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant consequence data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>CONSEQUENCE_CLASSIFICATION</th>\n",
       "      <th>CONSEQUENCE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>INS</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148882_C_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148920_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148959_A_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT TYPE CONSEQUENCE_CLASSIFICATION CONSEQUENCE  \\\n",
       "0     13  110148882   C  CT  INS                 downstream  downstream   \n",
       "1     13  110148891   C   G  SNV                 downstream  downstream   \n",
       "2     13  110148917   C   G  SNV                 downstream  downstream   \n",
       "3     13  110148920   G   C  SNV                 downstream  downstream   \n",
       "4     13  110148959   A   G  SNV                 downstream  downstream   \n",
       "\n",
       "     GENE              ID  \n",
       "0  COL4A1  110148882_C_CT  \n",
       "1  COL4A1   110148891_C_G  \n",
       "2  COL4A1   110148917_C_G  \n",
       "3  COL4A1   110148920_G_C  \n",
       "4  COL4A1   110148959_A_G  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consequence_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_consequences.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "consequence_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis and Visualisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select effect data on rare variants within African subpopulation (ethnolinguistic) groups for analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148891_C_G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148917_C_G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>119</td>\n",
       "      <td>1220</td>\n",
       "      <td>1101</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.097541</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110149176_T_A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>110149176</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149176-T-A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110149349_G_A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1220</td>\n",
       "      <td>1215</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149349-G-A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110149494_C_T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID     VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  110148891_C_G  rs552586867  110148891   C   G  COL4A1     NaN           1   \n",
       "1  110148917_C_G   rs59409892  110148917   C   G  COL4A1     NaN         119   \n",
       "2  110149176_T_A  rs546124548  110149176   T   A  COL4A1     NaN           1   \n",
       "3  110149349_G_A  rs139916479  110149349   G   A  COL4A1     NaN           5   \n",
       "4  110149494_C_T  rs552877576  110149494   C   T  COL4A1     NaN           1   \n",
       "\n",
       "   IH_TOTAL_CTS  IH_REF_CTS             REG     IH_AF VARIANT_TYPE  \\\n",
       "0          1220        1219  Recent African  0.000820          SNP   \n",
       "1          1220        1101  Recent African  0.097541          SNP   \n",
       "2          1220        1219  Recent African  0.000820          SNP   \n",
       "3          1220        1215  Recent African  0.004098          SNP   \n",
       "4          1220        1219  Recent African  0.000820          SNP   \n",
       "\n",
       "         VariantVcf         Rsid  \\\n",
       "0  13-110148891-C-G  rs552586867   \n",
       "1  13-110148917-C-G   rs59409892   \n",
       "2  13-110149176-T-A  rs546124548   \n",
       "3  13-110149349-G-A  rs139916479   \n",
       "4  13-110149494-C-T  rs552877576   \n",
       "\n",
       "                                               Clndn  CHROM  \n",
       "0                                                NaN   13.0  \n",
       "1                                                NaN   13.0  \n",
       "2                                                NaN   13.0  \n",
       "3  Brain_small_vessel_disease_1_with_or_without_o...   13.0  \n",
       "4                                                NaN   13.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select aggregated variant count and frequency data for Recent Africans. Remove variants with an alternate allele count of 0. These variants are not present in Recent Africans.\n",
    "\n",
    "ih_afr_subpops = ih_afr[(ih_afr[\"REG\"] == \"Recent African\") & (ih_afr[\"IH_ALT_CTS\"] > 0)]\n",
    "\n",
    "# Add in effect data for rare variants that are in the Recent African populations\n",
    "ih_afr_subpops_phenotype_data = (\n",
    "    ih_afr_subpops.merge(\n",
    "        phenotype_data,\n",
    "        how=\"left\",\n",
    "        left_on=[\"REF\", \"ALT\", \"POS\"],\n",
    "        right_on=[\"REF\", \"ALT\", \"POS\"],\n",
    "    )\n",
    "    .drop(columns=\"ID_y\")\n",
    "    .rename(columns={\"ID_x\": \"ID\"})\n",
    ")\n",
    "\n",
    "ih_afr_subpops_phenotype_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many variants have known disease phenotypes?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What disease phenotypes are present?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do the variants with known disease phenotypes have predicted deleteriousness (CADD>=20)?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
