{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant disease phenotypes\n",
    "\n",
    "This analysis aimed to answer the following research question: **How many of the variants identified in African populations have known disease associations, and what are these associations?**\n",
    "\n",
    "To achieve this, the following steps were performed:\n",
    "\n",
    "1. Genetic Variant Phenotype Data Retrieval and Preparation: Known disease phenotypes for the variants were retrieved from [Favor v2.0](https://favor.genohub.org/). The retrieved data underwent processing and preparation following guidelines outlined in the [Notebooks\\Data_preparation\\6-Variant_phenotype_associations.ipynb](https://github.com/MeganHolborn/Genetic_data_analysis/blob/main/Notebooks/Data_preparation/6-Variant_phenotype_associations.ipynb) Jupyter notebook. The processed data can be found [here](https://github.com/MeganHolborn/Genetic_data_analysis/blob/main/Data/Processed/Variant_disease_phenotypes.csv).\n",
    "2. Analysis and Visualisation:\n",
    "    * To be completed..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import upsetplot\n",
    "from matplotlib import pyplot as plt\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant phenotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110149494_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149715</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>110149715_AAT_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110151168</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110151168_C_T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf         Rsid Clndn  CHROM        POS  REF ALT  \\\n",
       "0    13-110148917-C-G   rs59409892   NaN     13  110148917    C   G   \n",
       "1    13-110148891-C-G  rs552586867   NaN     13  110148891    C   G   \n",
       "2    13-110149494-C-T  rs552877576   NaN     13  110149494    C   T   \n",
       "3  13-110149715-AAT-A  rs886049952   NaN     13  110149715  AAT   A   \n",
       "4    13-110151168-C-T  rs557686466   NaN     13  110151168    C   T   \n",
       "\n",
       "                ID  \n",
       "0    110148917_C_G  \n",
       "1    110148891_C_G  \n",
       "2    110149494_C_T  \n",
       "3  110149715_AAT_A  \n",
       "4    110151168_C_T  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_disease_phenotypes.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "phenotype_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import genetic variant count data for African populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Bantu Kenya</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>EA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>0</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>San</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>SA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mende</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>166</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mbuti Pygmy</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID             VAR_NAME        POS REF ALT    GENE  \\\n",
       "0  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "1  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "2  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "3  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "4  110148882_C_CT  chr13:110148882C-CT  110148882   C  CT  COL4A1   \n",
       "\n",
       "       SUB_POP  IH_ALT_CTS  IH_TOTAL_CTS  IH_REF_CTS REG  IH_AF VARIANT_TYPE  \n",
       "0  Bantu Kenya           0            20          20  EA    0.0        INDEL  \n",
       "1       Yoruba           0           276         276  WA    0.0        INDEL  \n",
       "2          San           0            12          12  SA    0.0        INDEL  \n",
       "3        Mende           0           166         166  WA    0.0        INDEL  \n",
       "4  Mbuti Pygmy           0            24          24  CA    0.0        INDEL  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_afr = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "ih_afr.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import genetic variant count data for global populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>ALT_CT_ALFA_East Asian</th>\n",
       "      <th>ALT_CT_ALFA_European</th>\n",
       "      <th>ALT_CT_ALFA_Latin American 1</th>\n",
       "      <th>ALT_CT_ALFA_Latin American 2</th>\n",
       "      <th>ALT_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_Latin American 1</th>\n",
       "      <th>REF_CT_ALFA_Latin American 2</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000343</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>109377.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>2200.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000989</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>55.0</td>\n",
       "      <td>21489.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>1330.0</td>\n",
       "      <td>1685.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>37269.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>2052.0</td>\n",
       "      <td>3283.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000990</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>32.0</td>\n",
       "      <td>5355.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>261.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>8931.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>62.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1005573</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>35.0</td>\n",
       "      <td>10693.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>4955.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>956.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1007311</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9154.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>11242.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>456.0</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  variant_id reference_allele alternate_allele  ALT_CT_ALFA_East Asian  \\\n",
       "0  rs1000343                C                T                     0.0   \n",
       "1  rs1000989                T                C                    55.0   \n",
       "2  rs1000990                T                C                    32.0   \n",
       "3  rs1005573                C                T                    35.0   \n",
       "4  rs1007311                A                G                    56.0   \n",
       "\n",
       "   ALT_CT_ALFA_European  ALT_CT_ALFA_Latin American 1  \\\n",
       "0                  49.0                           5.0   \n",
       "1               21489.0                         123.0   \n",
       "2                5355.0                          40.0   \n",
       "3               10693.0                         209.0   \n",
       "4                9154.0                          61.0   \n",
       "\n",
       "   ALT_CT_ALFA_Latin American 2  ALT_CT_ALFA_South Asian  \\\n",
       "0                          10.0                      0.0   \n",
       "1                        1330.0                   1685.0   \n",
       "2                         261.0                     36.0   \n",
       "3                        1810.0                     79.0   \n",
       "4                         154.0                     43.0   \n",
       "\n",
       "   REF_CT_ALFA_East Asian  REF_CT_ALFA_European  REF_CT_ALFA_Latin American 1  \\\n",
       "0                   490.0              109377.0                         673.0   \n",
       "1                   109.0               37269.0                         273.0   \n",
       "2                    54.0                8931.0                         106.0   \n",
       "3                    69.0                4955.0                          87.0   \n",
       "4                    56.0               11242.0                          85.0   \n",
       "\n",
       "   REF_CT_ALFA_Latin American 2  REF_CT_ALFA_South Asian  \n",
       "0                        2200.0                    184.0  \n",
       "1                        2052.0                   3283.0  \n",
       "2                         349.0                     62.0  \n",
       "3                         956.0                     31.0  \n",
       "4                         456.0                     55.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alfa_global = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"ALFA_allele_counts_b.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "alfa_global.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant effect data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POLYPHEN_PRED</th>\n",
       "      <th>POLYPHEN_SCORE</th>\n",
       "      <th>SIFT_PRED</th>\n",
       "      <th>SIFT_SCORE</th>\n",
       "      <th>CADD_RAW_SCORE</th>\n",
       "      <th>CADD_PHRED_SCORE</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.437825</td>\n",
       "      <td>0.160</td>\n",
       "      <td>110148882_C_CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.227221</td>\n",
       "      <td>0.446</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.269936</td>\n",
       "      <td>3.938</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.530972</td>\n",
       "      <td>6.825</td>\n",
       "      <td>110148920_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.380228</td>\n",
       "      <td>14.950</td>\n",
       "      <td>110148959_A_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT    GENE POLYPHEN_PRED  POLYPHEN_SCORE SIFT_PRED  \\\n",
       "0     13  110148882   C  CT  COL4A1           NaN             NaN       NaN   \n",
       "1     13  110148891   C   G  COL4A1           NaN             NaN       NaN   \n",
       "2     13  110148917   C   G  COL4A1           NaN             NaN       NaN   \n",
       "3     13  110148920   G   C  COL4A1           NaN             NaN       NaN   \n",
       "4     13  110148959   A   G  COL4A1           NaN             NaN       NaN   \n",
       "\n",
       "   SIFT_SCORE  CADD_RAW_SCORE  CADD_PHRED_SCORE              ID  \n",
       "0         NaN       -0.437825             0.160  110148882_C_CT  \n",
       "1         NaN       -0.227221             0.446   110148891_C_G  \n",
       "2         NaN        0.269936             3.938   110148917_C_G  \n",
       "3         NaN        0.530972             6.825   110148920_G_C  \n",
       "4         NaN        1.380228            14.950   110148959_A_G  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_effects.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis and Visualisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select effect data on rare variants within African subpopulation (ethnolinguistic) groups for analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148891_C_G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148917_C_G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>119</td>\n",
       "      <td>1220</td>\n",
       "      <td>1101</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.097541</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110149176_T_A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>110149176</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149176-T-A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110149349_G_A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1220</td>\n",
       "      <td>1215</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149349-G-A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110149494_C_T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>1219</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>SNP</td>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID     VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  110148891_C_G  rs552586867  110148891   C   G  COL4A1     NaN           1   \n",
       "1  110148917_C_G   rs59409892  110148917   C   G  COL4A1     NaN         119   \n",
       "2  110149176_T_A  rs546124548  110149176   T   A  COL4A1     NaN           1   \n",
       "3  110149349_G_A  rs139916479  110149349   G   A  COL4A1     NaN           5   \n",
       "4  110149494_C_T  rs552877576  110149494   C   T  COL4A1     NaN           1   \n",
       "\n",
       "   IH_TOTAL_CTS  IH_REF_CTS             REG     IH_AF VARIANT_TYPE  \\\n",
       "0          1220        1219  Recent African  0.000820          SNP   \n",
       "1          1220        1101  Recent African  0.097541          SNP   \n",
       "2          1220        1219  Recent African  0.000820          SNP   \n",
       "3          1220        1215  Recent African  0.004098          SNP   \n",
       "4          1220        1219  Recent African  0.000820          SNP   \n",
       "\n",
       "         VariantVcf         Rsid  \\\n",
       "0  13-110148891-C-G  rs552586867   \n",
       "1  13-110148917-C-G   rs59409892   \n",
       "2  13-110149176-T-A  rs546124548   \n",
       "3  13-110149349-G-A  rs139916479   \n",
       "4  13-110149494-C-T  rs552877576   \n",
       "\n",
       "                                               Clndn  CHROM  \n",
       "0                                                NaN   13.0  \n",
       "1                                                NaN   13.0  \n",
       "2                                                NaN   13.0  \n",
       "3  Brain_small_vessel_disease_1_with_or_without_o...   13.0  \n",
       "4                                                NaN   13.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select aggregated variant count and frequency data for Recent Africans. Remove variants with an alternate allele count of 0. These variants are not present in Recent Africans.\n",
    "\n",
    "ih_afr_subpops = ih_afr[(ih_afr[\"REG\"] == \"Recent African\") & (ih_afr[\"IH_ALT_CTS\"] > 0)]\n",
    "\n",
    "# Add in effect data for rare variants that are in the Recent African populations\n",
    "ih_afr_subpops_phenotype_data = (\n",
    "    ih_afr_subpops.merge(\n",
    "        phenotype_data,\n",
    "        how=\"left\",\n",
    "        left_on=[\"REF\", \"ALT\", \"POS\"],\n",
    "        right_on=[\"REF\", \"ALT\", \"POS\"],\n",
    "    )\n",
    "    .drop(columns=\"ID_y\")\n",
    "    .rename(columns={\"ID_x\": \"ID\"})\n",
    ")\n",
    "\n",
    "ih_afr_subpops_phenotype_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many variants have previously been associated with HIE and are present in Africans?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a list of variants (with rsids) associated with HIE in the genes of interest. This information was retrieved from: https://doi.org/10.1016/j.ygeno.2022.110508."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "hie_variant_rsids = [\n",
    "    \"rs2067853\",\n",
    "    \"rs1217401\",\n",
    "    \"rs2043211\",\n",
    "    \"rs1001179\",\n",
    "    \"rs1800896\",\n",
    "    \"rs1071676\",\n",
    "    \"rs1143623\",\n",
    "    \"rs16944\",\n",
    "    \"rs1800795\",\n",
    "    \"rs1801133\",\n",
    "    \"rs1808593\",\n",
    "    \"rs2070744\",\n",
    "    \"rs6517135\",\n",
    "    \"rs1799964\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many HIE-associated variants with rsids are there?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(hie_variant_rsids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which of these variants are present in Africans?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rs1071676', 'rs1217401', 'rs1801133', 'rs2070744', 'rs1808593',\n",
       "       'rs1800896', 'rs1800795', 'rs2067853', 'rs2043211'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_with_hie_assoc = ih_afr_subpops_phenotype_data[ih_afr_subpops_phenotype_data.VAR_NAME.isin(hie_variant_rsids)]\n",
    "variants_with_hie_assoc.VAR_NAME.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many of the variants are present in Africans?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_with_hie_assoc.VAR_NAME.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many variants have a known ClinVar disease phenotype?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split the phenotypes listed in each row of the Clndn column into a list\n",
    "\n",
    "ih_afr_subpops_phenotype_data['Clndn'] = ih_afr_subpops_phenotype_data['Clndn'].str.split('|')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the phenotypes list into multiple rows\n",
    "\n",
    "ih_afr_subpops_phenotype_explode = ih_afr_subpops_phenotype_data.explode('Clndn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows with missing data for the Clndn column\n",
    "\n",
    "variants_with_ClinVar_phenotypes = ih_afr_subpops_phenotype_explode[~((ih_afr_subpops_phenotype_explode['Clndn'].isna()) | (ih_afr_subpops_phenotype_explode['Clndn'] == 'not_provided') | (ih_afr_subpops_phenotype_explode['Clndn'] == 'not_specified') | (ih_afr_subpops_phenotype_explode['Clndn'] == 'not_specified|not_provided') | (ih_afr_subpops_phenotype_explode['Clndn'] == 'none_provided'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of variants with disease phenotype descriptions\n",
    "\n",
    "variants_with_ClinVar_phenotypes['ID'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_with_ClinVar_phenotypes[variants_with_ClinVar_phenotypes.VAR_NAME.isin(hie_variant_rsids)].ID.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Hypertension,_essential,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Preeclampsia,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Susceptibility_to_progression_to_renal_failure...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_with_hemorrhage</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Graft-versus-host_disease,_resistance_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Human_immunodeficiency_virus_type_1,_susceptib...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>Antisynthetase_syndrome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>Gastric_cancer_susceptibility_after_h._pylori_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Crohn_disease-associated_growth_failure,_susce...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus,_type_1,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus_type_2,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IL6</td>\n",
       "      <td>INTERLEUKIN_6_POLYMORPHISM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Intracranial_hemorrhage_in_brain_cerebrovascul...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Kaposi_sarcoma</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Rheumatoid_arthritis,_systemic_juvenile,_susce...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Intellectual_disability</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neoplasm_of_stomach</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>carboplatin_response_-_Efficacy</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>cyclophosphamide_response_-_Toxicity/ADR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>methotrexate_response_-_Dosage,_Efficacy,_Toxi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Gastrointestinal_stromal_tumor</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>MTHFR_deficiency,_thermolabile_type</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Thrombophilia_due_to_thrombin_defect</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neural_tube_defects,_folate-sensitive</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_MTHFR_deficiency</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Alzheimer_disease,_late-onset,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Coronary_artery_spasm_1,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension,_pregnancy-induced,_susceptibilit...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension_resistant_to_conventional_therapy</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_heart_disease,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_stroke,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Metabolic_syndrome,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE                                              Clndn  Count\n",
       "0      AGT         Hypertension,_essential,_susceptibility_to      1\n",
       "1      AGT                    Preeclampsia,_susceptibility_to      1\n",
       "3      AGT  Susceptibility_to_progression_to_renal_failure...      1\n",
       "2      AGT                                    Renal_dysplasia     22\n",
       "4    AP4B1             History_of_neurodevelopmental_disorder      7\n",
       "5    AP4B1         Spastic_paraplegia_47,_autosomal_recessive     10\n",
       "9   COL4A1                                 Porencephalic_cyst     17\n",
       "8   COL4A1         Brain_small_vessel_disease_with_hemorrhage     26\n",
       "7   COL4A1  Brain_small_vessel_disease_1_with_or_without_o...     38\n",
       "6   COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...     40\n",
       "10    IL10           Graft-versus-host_disease,_resistance_to      1\n",
       "11    IL10  Human_immunodeficiency_virus_type_1,_susceptib...      1\n",
       "12    IL10                         Inflammatory_bowel_disease     10\n",
       "13    IL1B                            Antisynthetase_syndrome      1\n",
       "14    IL1B  Gastric_cancer_susceptibility_after_h._pylori_...      1\n",
       "15     IL6  Crohn_disease-associated_growth_failure,_susce...      1\n",
       "16     IL6       Diabetes_mellitus,_type_1,_susceptibility_to      1\n",
       "17     IL6        Diabetes_mellitus_type_2,_susceptibility_to      1\n",
       "18     IL6                         INTERLEUKIN_6_POLYMORPHISM      1\n",
       "19     IL6  Intracranial_hemorrhage_in_brain_cerebrovascul...      1\n",
       "20     IL6                                     Kaposi_sarcoma      1\n",
       "21     IL6  Rheumatoid_arthritis,_systemic_juvenile,_susce...      1\n",
       "25   MTHFR                            Intellectual_disability      1\n",
       "27   MTHFR                                Neoplasm_of_stomach      1\n",
       "29   MTHFR                                      Schizophrenia      1\n",
       "30   MTHFR                   Schizophrenia,_susceptibility_to      1\n",
       "32   MTHFR                    carboplatin_response_-_Efficacy      1\n",
       "33   MTHFR           cyclophosphamide_response_-_Toxicity/ADR      1\n",
       "34   MTHFR  methotrexate_response_-_Dosage,_Efficacy,_Toxi...      1\n",
       "22   MTHFR                     Gastrointestinal_stromal_tumor      2\n",
       "26   MTHFR                MTHFR_deficiency,_thermolabile_type      2\n",
       "31   MTHFR               Thrombophilia_due_to_thrombin_defect      2\n",
       "28   MTHFR              Neural_tube_defects,_folate-sensitive      7\n",
       "23   MTHFR             Homocystinuria_due_to_MTHFR_deficiency     13\n",
       "24   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...     39\n",
       "35    NOS3   Alzheimer_disease,_late-onset,_susceptibility_to      1\n",
       "36    NOS3         Coronary_artery_spasm_1,_susceptibility_to      1\n",
       "37    NOS3  Hypertension,_pregnancy-induced,_susceptibilit...      1\n",
       "38    NOS3     Hypertension_resistant_to_conventional_therapy      1\n",
       "39    NOS3          Ischemic_heart_disease,_susceptibility_to      1\n",
       "40    NOS3                 Ischemic_stroke,_susceptibility_to      1\n",
       "41    NOS3              Metabolic_syndrome,_susceptibility_to      1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_with_known_phenotypes_grouped_by_gene = variants_with_ClinVar_phenotypes[['GENE','Clndn','ID']].groupby(['GENE','Clndn']).count().reset_index().sort_values(by=['GENE','ID'], ascending=True)\n",
    "variants_with_known_phenotypes_grouped_by_gene.rename(columns={'ID':'Count'}, inplace=True)\n",
    "\n",
    "variants_with_known_phenotypes_grouped_by_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine information on variants associated with HIE and variants associated with diseases in ClinVar databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "137"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate information on variants with known ClinVar phenotypes and HIE associations. Drop duplicate information.\n",
    "\n",
    "variants_with_known_phenotypes = pd.concat([variants_with_ClinVar_phenotypes, variants_with_hie_assoc]).drop_duplicates()\n",
    "\n",
    "variants_with_known_phenotypes.VAR_NAME.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many of the variants with known disease phenotypes (ClinVar and HIE studies) have predicted deleteriousness (CADD>=10)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join the phenotype and cadd data\n",
    "\n",
    "variants_phenotypes_and_cadd_phred = pd.merge(variants_with_known_phenotypes, vep_data, how='left', on=['ID','CHROM','POS','REF','ALT','GENE'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for variants with cadd >=10\n",
    "\n",
    "deleterious_variants_with_known_phenotype = variants_phenotypes_and_cadd_phred[variants_phenotypes_and_cadd_phred.CADD_PHRED_SCORE >= 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of variants with known disease phenotypes and cadd >= 10\n",
    "\n",
    "deleterious_variants_with_known_phenotype.ID.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29.927007299270077"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Percentage of variants with known disease phenotypes and cadd >= 10 of the variants with known disease phenotypes\n",
    "\n",
    "(deleterious_variants_with_known_phenotype.ID.nunique()/variants_with_known_phenotypes['ID'].nunique())*100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What phenotypes are represented in the variants with cadd scores >= 10?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive|His...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_with_hemorrhage</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>Gastric_cancer_susceptibility_after_h._pylori_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Crohn_disease-associated_growth_failure,_susce...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus,_type_1,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus_type_2,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus_type_2,_susceptibility_to|Di...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Intracranial_hemorrhage_in_brain_cerebrovascul...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Kaposi_sarcoma</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Rheumatoid_arthritis,_systemic_juvenile,_susce...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Intellectual_disability</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neoplasm_of_stomach</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neoplasm_of_stomach|Gastrointestinal_stromal_t...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>carboplatin_response_-_Efficacy</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>cyclophosphamide_response_-_Toxicity/ADR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>methotrexate_response_-_Dosage,_Efficacy,_Toxi...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Gastrointestinal_stromal_tumor</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>MTHFR_deficiency,_thermolabile_type</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Thrombophilia_due_to_thrombin_defect</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neural_tube_defects,_folate-sensitive</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_MTHFR_deficiency</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Alzheimer_disease,_late-onset,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Coronary_artery_spasm_1,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension,_pregnancy-induced,_susceptibilit...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension_resistant_to_conventional_therapy</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_heart_disease,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_stroke,_susceptibility_to</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE                                              Clndn  Count\n",
       "0      AGT                                    Renal_dysplasia      5\n",
       "3    AP4B1  Spastic_paraplegia_47,_autosomal_recessive|His...      1\n",
       "1    AP4B1             History_of_neurodevelopmental_disorder      5\n",
       "2    AP4B1         Spastic_paraplegia_47,_autosomal_recessive      7\n",
       "6   COL4A1         Brain_small_vessel_disease_with_hemorrhage      3\n",
       "5   COL4A1  Brain_small_vessel_disease_1_with_or_without_o...      5\n",
       "7   COL4A1                                 Porencephalic_cyst      5\n",
       "4   COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...      6\n",
       "8     IL10                         Inflammatory_bowel_disease      5\n",
       "9     IL1B  Gastric_cancer_susceptibility_after_h._pylori_...      1\n",
       "10     IL6  Crohn_disease-associated_growth_failure,_susce...      1\n",
       "11     IL6       Diabetes_mellitus,_type_1,_susceptibility_to      1\n",
       "12     IL6        Diabetes_mellitus_type_2,_susceptibility_to      1\n",
       "13     IL6  Diabetes_mellitus_type_2,_susceptibility_to|Di...      1\n",
       "14     IL6  Intracranial_hemorrhage_in_brain_cerebrovascul...      1\n",
       "15     IL6                                     Kaposi_sarcoma      1\n",
       "16     IL6  Rheumatoid_arthritis,_systemic_juvenile,_susce...      1\n",
       "20   MTHFR                            Intellectual_disability      1\n",
       "22   MTHFR                                Neoplasm_of_stomach      1\n",
       "23   MTHFR  Neoplasm_of_stomach|Gastrointestinal_stromal_t...      1\n",
       "25   MTHFR                                      Schizophrenia      1\n",
       "26   MTHFR                   Schizophrenia,_susceptibility_to      1\n",
       "28   MTHFR                    carboplatin_response_-_Efficacy      1\n",
       "29   MTHFR           cyclophosphamide_response_-_Toxicity/ADR      1\n",
       "30   MTHFR  methotrexate_response_-_Dosage,_Efficacy,_Toxi...      1\n",
       "17   MTHFR                     Gastrointestinal_stromal_tumor      2\n",
       "21   MTHFR                MTHFR_deficiency,_thermolabile_type      2\n",
       "27   MTHFR               Thrombophilia_due_to_thrombin_defect      2\n",
       "24   MTHFR              Neural_tube_defects,_folate-sensitive      3\n",
       "18   MTHFR             Homocystinuria_due_to_MTHFR_deficiency      5\n",
       "19   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...     13\n",
       "31    NOS3   Alzheimer_disease,_late-onset,_susceptibility_to      1\n",
       "32    NOS3         Coronary_artery_spasm_1,_susceptibility_to      1\n",
       "33    NOS3  Hypertension,_pregnancy-induced,_susceptibilit...      1\n",
       "34    NOS3     Hypertension_resistant_to_conventional_therapy      1\n",
       "35    NOS3          Ischemic_heart_disease,_susceptibility_to      1\n",
       "36    NOS3                 Ischemic_stroke,_susceptibility_to      1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deleterious_variants_with_known_phenotype_grouped_by_gene = deleterious_variants_with_known_phenotype[['GENE','Clndn','ID']].groupby(['GENE','Clndn']).count().reset_index().sort_values(by=['GENE','ID'], ascending=True)\n",
    "deleterious_variants_with_known_phenotype_grouped_by_gene.rename(columns={'ID':'Count'}, inplace=True)\n",
    "\n",
    "deleterious_variants_with_known_phenotype_grouped_by_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deleterious_variants_with_known_phenotype[deleterious_variants_with_known_phenotype.VAR_NAME.isin(hie_variant_rsids)].ID.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rs1217401', 'rs1801133', 'rs1800795'], dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deleterious_variants_with_known_phenotype[deleterious_variants_with_known_phenotype.VAR_NAME.isin(hie_variant_rsids)].VAR_NAME.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the known phenotypes of the variants with cadd phred scores >= 10 and common frequencies (>0.1)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>CADD_PHRED_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>113896329_A_G</td>\n",
       "      <td>rs1217401</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.797541</td>\n",
       "      <td>21.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>113896329_A_G</td>\n",
       "      <td>rs1217401</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.797541</td>\n",
       "      <td>21.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive|His...</td>\n",
       "      <td>113896329_A_G</td>\n",
       "      <td>rs1217401</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.797541</td>\n",
       "      <td>21.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110149776_G_T</td>\n",
       "      <td>rs13260</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.266393</td>\n",
       "      <td>11.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110205548_A_G</td>\n",
       "      <td>rs677877</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>10.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110307009_C_G</td>\n",
       "      <td>rs9515185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>13.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110307117_C_A</td>\n",
       "      <td>rs113651836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.107084</td>\n",
       "      <td>14.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>110149776_G_T</td>\n",
       "      <td>rs13260</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.266393</td>\n",
       "      <td>11.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>110205548_A_G</td>\n",
       "      <td>rs677877</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>10.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>110307009_C_G</td>\n",
       "      <td>rs9515185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>13.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>110307117_C_A</td>\n",
       "      <td>rs113651836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.107084</td>\n",
       "      <td>14.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_with_hemorrhage</td>\n",
       "      <td>110307009_C_G</td>\n",
       "      <td>rs9515185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>13.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>110149776_G_T</td>\n",
       "      <td>rs13260</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.266393</td>\n",
       "      <td>11.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>110205548_A_G</td>\n",
       "      <td>rs677877</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>10.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>110307009_C_G</td>\n",
       "      <td>rs9515185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>13.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>110307117_C_A</td>\n",
       "      <td>rs113651836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.107084</td>\n",
       "      <td>14.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>206771300_T_C</td>\n",
       "      <td>rs1518111</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.555738</td>\n",
       "      <td>10.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>Gastric_cancer_susceptibility_after_h._pylori_...</td>\n",
       "      <td>112836810_G_A</td>\n",
       "      <td>rs1143627</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.351639</td>\n",
       "      <td>10.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Crohn_disease-associated_growth_failure,_susce...</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus,_type_1,_susceptibility_to</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus_type_2,_susceptibility_to</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Diabetes_mellitus_type_2,_susceptibility_to|Di...</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Intracranial_hemorrhage_in_brain_cerebrovascul...</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Kaposi_sarcoma</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>IL6</td>\n",
       "      <td>Rheumatoid_arthritis,_systemic_juvenile,_susce...</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>18.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Gastrointestinal_stromal_tumor</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_MTHFR_deficiency</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>MTHFR_deficiency,_thermolabile_type</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neural_tube_defects,_folate-sensitive</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia,_susceptibility_to</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>20.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Alzheimer_disease,_late-onset,_susceptibility_to</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Coronary_artery_spasm_1,_susceptibility_to</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension,_pregnancy-induced,_susceptibilit...</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Hypertension_resistant_to_conventional_therapy</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_heart_disease,_susceptibility_to</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>Ischemic_stroke,_susceptibility_to</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>10.07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GENE                                              Clndn             ID  \\\n",
       "125   AP4B1             History_of_neurodevelopmental_disorder  113896329_A_G   \n",
       "124   AP4B1         Spastic_paraplegia_47,_autosomal_recessive  113896329_A_G   \n",
       "264   AP4B1  Spastic_paraplegia_47,_autosomal_recessive|His...  113896329_A_G   \n",
       "8    COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...  110149776_G_T   \n",
       "83   COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...  110205548_A_G   \n",
       "116  COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...  110307009_C_G   \n",
       "120  COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...  110307117_C_A   \n",
       "7    COL4A1  Brain_small_vessel_disease_1_with_or_without_o...  110149776_G_T   \n",
       "82   COL4A1  Brain_small_vessel_disease_1_with_or_without_o...  110205548_A_G   \n",
       "115  COL4A1  Brain_small_vessel_disease_1_with_or_without_o...  110307009_C_G   \n",
       "119  COL4A1  Brain_small_vessel_disease_1_with_or_without_o...  110307117_C_A   \n",
       "117  COL4A1         Brain_small_vessel_disease_with_hemorrhage  110307009_C_G   \n",
       "6    COL4A1                                 Porencephalic_cyst  110149776_G_T   \n",
       "81   COL4A1                                 Porencephalic_cyst  110205548_A_G   \n",
       "114  COL4A1                                 Porencephalic_cyst  110307009_C_G   \n",
       "118  COL4A1                                 Porencephalic_cyst  110307117_C_A   \n",
       "224    IL10                         Inflammatory_bowel_disease  206771300_T_C   \n",
       "122    IL1B  Gastric_cancer_susceptibility_after_h._pylori_...  112836810_G_A   \n",
       "237     IL6  Crohn_disease-associated_growth_failure,_susce...   22727026_C_G   \n",
       "233     IL6       Diabetes_mellitus,_type_1,_susceptibility_to   22727026_C_G   \n",
       "232     IL6        Diabetes_mellitus_type_2,_susceptibility_to   22727026_C_G   \n",
       "268     IL6  Diabetes_mellitus_type_2,_susceptibility_to|Di...   22727026_C_G   \n",
       "234     IL6  Intracranial_hemorrhage_in_brain_cerebrovascul...   22727026_C_G   \n",
       "236     IL6                                     Kaposi_sarcoma   22727026_C_G   \n",
       "235     IL6  Rheumatoid_arthritis,_systemic_juvenile,_susce...   22727026_C_G   \n",
       "169   MTHFR                     Gastrointestinal_stromal_tumor   11794419_T_G   \n",
       "173   MTHFR             Homocystinuria_due_to_MTHFR_deficiency   11794419_T_G   \n",
       "170   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   11794419_T_G   \n",
       "172   MTHFR                MTHFR_deficiency,_thermolabile_type   11794419_T_G   \n",
       "171   MTHFR              Neural_tube_defects,_folate-sensitive   11794419_T_G   \n",
       "168   MTHFR                   Schizophrenia,_susceptibility_to   11794419_T_G   \n",
       "215    NOS3   Alzheimer_disease,_late-onset,_susceptibility_to  150999023_T_G   \n",
       "218    NOS3         Coronary_artery_spasm_1,_susceptibility_to  150999023_T_G   \n",
       "216    NOS3  Hypertension,_pregnancy-induced,_susceptibilit...  150999023_T_G   \n",
       "217    NOS3     Hypertension_resistant_to_conventional_therapy  150999023_T_G   \n",
       "213    NOS3          Ischemic_heart_disease,_susceptibility_to  150999023_T_G   \n",
       "214    NOS3                 Ischemic_stroke,_susceptibility_to  150999023_T_G   \n",
       "\n",
       "        VAR_NAME REF ALT     IH_AF  CADD_PHRED_SCORE  \n",
       "125    rs1217401   A   G  0.797541             21.90  \n",
       "124    rs1217401   A   G  0.797541             21.90  \n",
       "264    rs1217401   A   G  0.797541             21.90  \n",
       "8        rs13260   G   T  0.266393             11.67  \n",
       "83      rs677877   A   G  0.388060             10.31  \n",
       "116    rs9515185   C   G  0.216172             13.89  \n",
       "120  rs113651836   C   A  0.107084             14.63  \n",
       "7        rs13260   G   T  0.266393             11.67  \n",
       "82      rs677877   A   G  0.388060             10.31  \n",
       "115    rs9515185   C   G  0.216172             13.89  \n",
       "119  rs113651836   C   A  0.107084             14.63  \n",
       "117    rs9515185   C   G  0.216172             13.89  \n",
       "6        rs13260   G   T  0.266393             11.67  \n",
       "81      rs677877   A   G  0.388060             10.31  \n",
       "114    rs9515185   C   G  0.216172             13.89  \n",
       "118  rs113651836   C   A  0.107084             14.63  \n",
       "224    rs1518111   T   C  0.555738             10.64  \n",
       "122    rs1143627   G   A  0.351639             10.26  \n",
       "237    rs1800795   C   G  0.999180             18.77  \n",
       "233    rs1800795   C   G  0.999180             18.77  \n",
       "232    rs1800795   C   G  0.999180             18.77  \n",
       "268    rs1800795   C   G  0.999180             18.77  \n",
       "234    rs1800795   C   G  0.999180             18.77  \n",
       "236    rs1800795   C   G  0.999180             18.77  \n",
       "235    rs1800795   C   G  0.999180             18.77  \n",
       "169    rs1801131   T   G  0.145902             20.90  \n",
       "173    rs1801131   T   G  0.145902             20.90  \n",
       "170    rs1801131   T   G  0.145902             20.90  \n",
       "172    rs1801131   T   G  0.145902             20.90  \n",
       "171    rs1801131   T   G  0.145902             20.90  \n",
       "168    rs1801131   T   G  0.145902             20.90  \n",
       "215    rs1799983   T   G  0.935246             10.07  \n",
       "218    rs1799983   T   G  0.935246             10.07  \n",
       "216    rs1799983   T   G  0.935246             10.07  \n",
       "217    rs1799983   T   G  0.935246             10.07  \n",
       "213    rs1799983   T   G  0.935246             10.07  \n",
       "214    rs1799983   T   G  0.935246             10.07  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display common variants, and their associated clinical disease phenotype\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype = deleterious_variants_with_known_phenotype[deleterious_variants_with_known_phenotype.IH_AF>=0.1][['GENE','Clndn','ID','VAR_NAME','REF','ALT','IH_AF','CADD_PHRED_SCORE']].sort_values(by=['GENE','Clndn','ID'], ascending=True)\n",
    "common_deleterious_variants_with_known_phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of variants with known phenotypes, CADD >=10 and that are common\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype.ID.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the frequencies of the common variants with known phenotypes and CADD Phred scores in Africans >=10 compare to that of other global populations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ALFA global allele count data to African data in common_deleterious_variants_with_known_phenotype dataframe\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa = pd.merge(common_deleterious_variants_with_known_phenotype.drop_duplicates(subset=['VAR_NAME','REF','ALT']).drop(columns=['Clndn','ID','CADD_PHRED_SCORE']), alfa_global, how='left', left_on=['VAR_NAME','REF','ALT'], right_on=['variant_id','reference_allele','alternate_allele'])\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa = common_deleterious_variants_with_known_phenotype_incl_alfa.drop(columns=['variant_id','reference_allele','alternate_allele'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>ALFA_AF_East Asian</th>\n",
       "      <th>ALFA_AF_South Asian</th>\n",
       "      <th>ALFA_AF_European</th>\n",
       "      <th>ALFA_AF_Latin American 1</th>\n",
       "      <th>ALFA_AF_Latin American 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs1217401</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.797541</td>\n",
       "      <td>0.057899</td>\n",
       "      <td>0.327415</td>\n",
       "      <td>0.321924</td>\n",
       "      <td>0.419217</td>\n",
       "      <td>0.192529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>rs13260</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.266393</td>\n",
       "      <td>0.004016</td>\n",
       "      <td>0.039773</td>\n",
       "      <td>0.091209</td>\n",
       "      <td>0.130312</td>\n",
       "      <td>0.226508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>rs677877</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.758065</td>\n",
       "      <td>0.660000</td>\n",
       "      <td>0.641808</td>\n",
       "      <td>0.613095</td>\n",
       "      <td>0.550746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>rs9515185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>0.488372</td>\n",
       "      <td>0.489796</td>\n",
       "      <td>0.406679</td>\n",
       "      <td>0.493151</td>\n",
       "      <td>0.372131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>rs113651836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.107084</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197838</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IL10</td>\n",
       "      <td>rs1518111</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.555738</td>\n",
       "      <td>0.295049</td>\n",
       "      <td>0.603571</td>\n",
       "      <td>0.784843</td>\n",
       "      <td>0.734914</td>\n",
       "      <td>0.661386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>rs1143627</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.351639</td>\n",
       "      <td>0.505700</td>\n",
       "      <td>0.415344</td>\n",
       "      <td>0.664999</td>\n",
       "      <td>0.585017</td>\n",
       "      <td>0.445177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IL6</td>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.559989</td>\n",
       "      <td>0.828767</td>\n",
       "      <td>0.819672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>0.214497</td>\n",
       "      <td>0.410439</td>\n",
       "      <td>0.314106</td>\n",
       "      <td>0.219092</td>\n",
       "      <td>0.192032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>rs1799983</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.679167</td>\n",
       "      <td>0.763889</td>\n",
       "      <td>0.240000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE     VAR_NAME REF ALT     IH_AF  ALFA_AF_East Asian  \\\n",
       "0   AP4B1    rs1217401   A   G  0.797541            0.057899   \n",
       "1  COL4A1      rs13260   G   T  0.266393            0.004016   \n",
       "2  COL4A1     rs677877   A   G  0.388060            0.758065   \n",
       "3  COL4A1    rs9515185   C   G  0.216172            0.488372   \n",
       "4  COL4A1  rs113651836   C   A  0.107084            0.026316   \n",
       "5    IL10    rs1518111   T   C  0.555738            0.295049   \n",
       "6    IL1B    rs1143627   G   A  0.351639            0.505700   \n",
       "7     IL6    rs1800795   C   G  0.999180            1.000000   \n",
       "8   MTHFR    rs1801131   T   G  0.145902            0.214497   \n",
       "9    NOS3    rs1799983   T   G  0.935246            0.857143   \n",
       "\n",
       "   ALFA_AF_South Asian  ALFA_AF_European  ALFA_AF_Latin American 1  \\\n",
       "0             0.327415          0.321924                  0.419217   \n",
       "1             0.039773          0.091209                  0.130312   \n",
       "2             0.660000          0.641808                  0.613095   \n",
       "3             0.489796          0.406679                  0.493151   \n",
       "4             0.000000          0.197838                  0.000000   \n",
       "5             0.603571          0.784843                  0.734914   \n",
       "6             0.415344          0.664999                  0.585017   \n",
       "7             0.857143          0.559989                  0.828767   \n",
       "8             0.410439          0.314106                  0.219092   \n",
       "9             0.333333          0.679167                  0.763889   \n",
       "\n",
       "   ALFA_AF_Latin American 2  \n",
       "0                  0.192529  \n",
       "1                  0.226508  \n",
       "2                  0.550746  \n",
       "3                  0.372131  \n",
       "4                  0.000000  \n",
       "5                  0.661386  \n",
       "6                  0.445177  \n",
       "7                  0.819672  \n",
       "8                  0.192032  \n",
       "9                  0.240000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the alternate allele frequencies for the global populations\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_East Asian'] = common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_East Asian']/(common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_East Asian']+common_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_East Asian'])\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_South Asian'] = common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_South Asian']/(common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_South Asian']+common_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_South Asian'])\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_European'] = common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_European']/(common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_European']+common_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_European'])\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_Latin American 1'] = common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 1']/(common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 1']+common_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_Latin American 1'])\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_Latin American 2'] = common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 2']/(common_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 2']+common_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_Latin American 2'])\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa.drop(columns=['ALT_CT_ALFA_East Asian','ALT_CT_ALFA_South Asian','ALT_CT_ALFA_European','ALT_CT_ALFA_Latin American 1','ALT_CT_ALFA_Latin American 2','REF_CT_ALFA_East Asian','REF_CT_ALFA_South Asian','REF_CT_ALFA_European','REF_CT_ALFA_Latin American 1','REF_CT_ALFA_Latin American 2'], inplace=True)\n",
    "common_deleterious_variants_with_known_phenotype_incl_alfa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the frequencies of the common variants with known phenotypes and CADD Phred scores >=10 in each African ethnolinguistic population group?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of variants to get frequencies for.\n",
    "\n",
    "common_deleterious_variants_with_known_phenotype_list = list(common_deleterious_variants_with_known_phenotype.ID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract ethnolinguistic population frequencies for each of these variants. \n",
    "\n",
    "ih_afr_subpops_common_deleterious_variants = ih_afr[(ih_afr.ID.isin(common_deleterious_variants_with_known_phenotype_list)) & ~ (ih_afr.REG == 'Recent African')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out irrelevant data. Pivot data.\n",
    "\n",
    "ih_afr_subpops_common_deleterious_variants_filtered = ih_afr_subpops_common_deleterious_variants.drop(columns=['POS','REF','ALT','IH_ALT_CTS','IH_TOTAL_CTS','IH_REF_CTS','VARIANT_TYPE','REG'])\n",
    "ih_afr_subpops_common_deleterious_variants_filtered = ih_afr_subpops_common_deleterious_variants_filtered.pivot(index=['ID','VAR_NAME','GENE'], columns=['SUB_POP'], values=['IH_AF']).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_2872\\3809755307.py:3: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left, 2 on the right)\n",
      "  ih_afr_subpops_common_deleterious_variants_filtered = pd.merge(left=common_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_common_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n",
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_2872\\3809755307.py:3: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "  ih_afr_subpops_common_deleterious_variants_filtered = pd.merge(left=common_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_common_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>(ID, )</th>\n",
       "      <th>(GENE, )</th>\n",
       "      <th>(IH_AF, Bantu Kenya)</th>\n",
       "      <th>(IH_AF, Bantu South Africa)</th>\n",
       "      <th>(IH_AF, Biaka Pygmy)</th>\n",
       "      <th>(IH_AF, Esan)</th>\n",
       "      <th>(IH_AF, Luhya)</th>\n",
       "      <th>(IH_AF, Mandenka)</th>\n",
       "      <th>(IH_AF, Mandinka)</th>\n",
       "      <th>(IH_AF, Mbuti Pygmy)</th>\n",
       "      <th>(IH_AF, Mende)</th>\n",
       "      <th>(IH_AF, San)</th>\n",
       "      <th>(IH_AF, Yoruba)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1217401</td>\n",
       "      <td>0.797541</td>\n",
       "      <td>113896329_A_G</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.812500</td>\n",
       "      <td>0.568182</td>\n",
       "      <td>0.815534</td>\n",
       "      <td>0.788043</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.831897</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>0.777108</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.840580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs13260</td>\n",
       "      <td>0.266393</td>\n",
       "      <td>110149776_G_T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.312500</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.271845</td>\n",
       "      <td>0.320652</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.265060</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.271739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs677877</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>110205548_A_G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.388350</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>0.448276</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.353659</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.380597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs9515185</td>\n",
       "      <td>0.216172</td>\n",
       "      <td>110307009_C_G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.169903</td>\n",
       "      <td>0.157609</td>\n",
       "      <td>0.236842</td>\n",
       "      <td>0.327586</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>0.265060</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.154412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs113651836</td>\n",
       "      <td>0.107084</td>\n",
       "      <td>110307117_C_A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.169903</td>\n",
       "      <td>0.097826</td>\n",
       "      <td>0.131579</td>\n",
       "      <td>0.081897</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.066265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.131387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1518111</td>\n",
       "      <td>0.555738</td>\n",
       "      <td>206771300_T_C</td>\n",
       "      <td>IL10</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.681818</td>\n",
       "      <td>0.533981</td>\n",
       "      <td>0.608696</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>0.534483</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.512048</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.525362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1143627</td>\n",
       "      <td>0.351639</td>\n",
       "      <td>112836810_G_A</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>0.340909</td>\n",
       "      <td>0.359223</td>\n",
       "      <td>0.277174</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>0.426724</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.319277</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.347826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1800795</td>\n",
       "      <td>0.999180</td>\n",
       "      <td>22727026_C_G</td>\n",
       "      <td>IL6</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995690</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>0.145902</td>\n",
       "      <td>11794419_T_G</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.386364</td>\n",
       "      <td>0.135922</td>\n",
       "      <td>0.184783</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.120690</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.138554</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.119565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1799983</td>\n",
       "      <td>0.935246</td>\n",
       "      <td>150999023_T_G</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>0.90</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.931818</td>\n",
       "      <td>0.898058</td>\n",
       "      <td>0.961957</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.922414</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.951807</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.952899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      VAR_NAME     IH_AF         (ID, ) (GENE, )  (IH_AF, Bantu Kenya)  \\\n",
       "0    rs1217401  0.797541  113896329_A_G    AP4B1                  0.75   \n",
       "1      rs13260  0.266393  110149776_G_T   COL4A1                  0.40   \n",
       "2     rs677877  0.388060  110205548_A_G   COL4A1                  0.55   \n",
       "3    rs9515185  0.216172  110307009_C_G   COL4A1                  0.15   \n",
       "4  rs113651836  0.107084  110307117_C_A   COL4A1                  0.05   \n",
       "5    rs1518111  0.555738  206771300_T_C     IL10                  0.70   \n",
       "6    rs1143627  0.351639  112836810_G_A     IL1B                  0.30   \n",
       "7    rs1800795  0.999180   22727026_C_G      IL6                  1.00   \n",
       "8    rs1801131  0.145902   11794419_T_G    MTHFR                  0.10   \n",
       "9    rs1799983  0.935246  150999023_T_G     NOS3                  0.90   \n",
       "\n",
       "   (IH_AF, Bantu South Africa)  (IH_AF, Biaka Pygmy)  (IH_AF, Esan)  \\\n",
       "0                     0.812500              0.568182       0.815534   \n",
       "1                     0.312500              0.454545       0.271845   \n",
       "2                     0.285714              0.318182       0.388350   \n",
       "3                     0.125000              0.204545       0.169903   \n",
       "4                     0.000000              0.090909       0.169903   \n",
       "5                     0.687500              0.681818       0.533981   \n",
       "6                     0.187500              0.340909       0.359223   \n",
       "7                     1.000000              1.000000       1.000000   \n",
       "8                     0.250000              0.386364       0.135922   \n",
       "9                     1.000000              0.931818       0.898058   \n",
       "\n",
       "   (IH_AF, Luhya)  (IH_AF, Mandenka)  (IH_AF, Mandinka)  (IH_AF, Mbuti Pygmy)  \\\n",
       "0        0.788043           0.950000           0.831897              0.291667   \n",
       "1        0.320652           0.150000           0.206897              0.166667   \n",
       "2        0.384615           0.550000           0.448276              0.041667   \n",
       "3        0.157609           0.236842           0.327586              0.458333   \n",
       "4        0.097826           0.131579           0.081897              0.041667   \n",
       "5        0.608696           0.450000           0.534483              0.750000   \n",
       "6        0.277174           0.550000           0.426724              0.333333   \n",
       "7        1.000000           1.000000           0.995690              1.000000   \n",
       "8        0.184783           0.025000           0.120690              0.166667   \n",
       "9        0.961957           0.950000           0.922414              1.000000   \n",
       "\n",
       "   (IH_AF, Mende)  (IH_AF, San)  (IH_AF, Yoruba)  \n",
       "0        0.777108      0.666667         0.840580  \n",
       "1        0.265060      0.000000         0.271739  \n",
       "2        0.353659      0.166667         0.380597  \n",
       "3        0.265060      0.200000         0.154412  \n",
       "4        0.066265      0.000000         0.131387  \n",
       "5        0.512048      0.916667         0.525362  \n",
       "6        0.319277      0.166667         0.347826  \n",
       "7        1.000000      1.000000         1.000000  \n",
       "8        0.138554      0.333333         0.119565  \n",
       "9        0.951807      0.583333         0.952899  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add overall African frequencies\n",
    "\n",
    "ih_afr_subpops_common_deleterious_variants_filtered = pd.merge(left=common_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_common_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n",
    "ih_afr_subpops_common_deleterious_variants_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the known phenotypes of the variants with cadd phred scores >= 10 and rare frequencies (>=0.01)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>CADD_PHRED_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>230703157_G_A</td>\n",
       "      <td>rs143479528</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>24.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>230703274_G_A</td>\n",
       "      <td>rs61751077</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>23.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>230704308_A_G</td>\n",
       "      <td>rs61731499</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>11.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>AGT</td>\n",
       "      <td>Renal_dysplasia</td>\n",
       "      <td>230706171_C_T</td>\n",
       "      <td>rs139685563</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>22.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>113897925_C_T</td>\n",
       "      <td>rs145803736</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>25.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>113900051_A_T</td>\n",
       "      <td>rs149335605</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>113900120_C_A</td>\n",
       "      <td>rs111785152</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>20.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>113895442_A_C</td>\n",
       "      <td>rs148748734</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>12.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>113900051_A_T</td>\n",
       "      <td>rs149335605</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>113900120_C_A</td>\n",
       "      <td>rs111785152</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>20.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>Spastic_paraplegia_47,_autosomal_recessive</td>\n",
       "      <td>113901851_T_C</td>\n",
       "      <td>rs143769705</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>14.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110149715_A_AAT</td>\n",
       "      <td>chr13:110149715A-AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>13.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Angiopathy,_hereditary,_with_nephropathy,_aneu...</td>\n",
       "      <td>110206692_T_C</td>\n",
       "      <td>rs140978802</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>10.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_1_with_or_without_o...</td>\n",
       "      <td>110206692_T_C</td>\n",
       "      <td>rs140978802</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>10.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Brain_small_vessel_disease_with_hemorrhage</td>\n",
       "      <td>110149715_A_AAT</td>\n",
       "      <td>chr13:110149715A-AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>13.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Porencephalic_cyst</td>\n",
       "      <td>110149715_A_AAT</td>\n",
       "      <td>chr13:110149715A-AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>13.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>206768666_T_G</td>\n",
       "      <td>rs568879359</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>24.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>206771057_A_G</td>\n",
       "      <td>rs139352858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>17.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>IL10</td>\n",
       "      <td>Inflammatory_bowel_disease</td>\n",
       "      <td>206772380_C_T</td>\n",
       "      <td>rs139073251</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>14.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_MTHFR_deficiency</td>\n",
       "      <td>11796278_G_A</td>\n",
       "      <td>rs34279942</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>10.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11792285_T_G</td>\n",
       "      <td>rs145641996</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>23.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11792354_C_A</td>\n",
       "      <td>rs45449298</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11794442_C_G</td>\n",
       "      <td>rs200688214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>27.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11795125_C_T</td>\n",
       "      <td>rs543016186</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>24.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11795341_T_G</td>\n",
       "      <td>rs142612062</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>22.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11796278_G_A</td>\n",
       "      <td>rs34279942</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>10.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11796319_C_T</td>\n",
       "      <td>rs150847674</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>22.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11802878_T_C</td>\n",
       "      <td>rs1413355</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.007377</td>\n",
       "      <td>19.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Homocystinuria_due_to_methylene_tetrahydrofola...</td>\n",
       "      <td>11806126_C_T</td>\n",
       "      <td>rs13306560</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>20.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Neural_tube_defects,_folate-sensitive</td>\n",
       "      <td>11792354_C_A</td>\n",
       "      <td>rs45449298</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>11792354_C_A</td>\n",
       "      <td>rs45449298</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>Thrombophilia_due_to_thrombin_defect</td>\n",
       "      <td>11792354_C_A</td>\n",
       "      <td>rs45449298</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>22.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GENE                                              Clndn  \\\n",
       "245     AGT                                    Renal_dysplasia   \n",
       "246     AGT                                    Renal_dysplasia   \n",
       "247     AGT                                    Renal_dysplasia   \n",
       "251     AGT                                    Renal_dysplasia   \n",
       "126   AP4B1             History_of_neurodevelopmental_disorder   \n",
       "129   AP4B1             History_of_neurodevelopmental_disorder   \n",
       "131   AP4B1             History_of_neurodevelopmental_disorder   \n",
       "123   AP4B1         Spastic_paraplegia_47,_autosomal_recessive   \n",
       "128   AP4B1         Spastic_paraplegia_47,_autosomal_recessive   \n",
       "130   AP4B1         Spastic_paraplegia_47,_autosomal_recessive   \n",
       "137   AP4B1         Spastic_paraplegia_47,_autosomal_recessive   \n",
       "4    COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...   \n",
       "85   COL4A1  Angiopathy,_hereditary,_with_nephropathy,_aneu...   \n",
       "84   COL4A1  Brain_small_vessel_disease_1_with_or_without_o...   \n",
       "5    COL4A1         Brain_small_vessel_disease_with_hemorrhage   \n",
       "3    COL4A1                                 Porencephalic_cyst   \n",
       "219    IL10                         Inflammatory_bowel_disease   \n",
       "223    IL10                         Inflammatory_bowel_disease   \n",
       "226    IL10                         Inflammatory_bowel_disease   \n",
       "192   MTHFR             Homocystinuria_due_to_MTHFR_deficiency   \n",
       "152   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "156   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "178   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "185   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "187   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "191   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "193   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "208   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "211   MTHFR  Homocystinuria_due_to_methylene_tetrahydrofola...   \n",
       "157   MTHFR              Neural_tube_defects,_folate-sensitive   \n",
       "154   MTHFR                                      Schizophrenia   \n",
       "155   MTHFR               Thrombophilia_due_to_thrombin_defect   \n",
       "\n",
       "                  ID              VAR_NAME REF  ALT     IH_AF  \\\n",
       "245    230703157_G_A           rs143479528   G    A  0.004098   \n",
       "246    230703274_G_A            rs61751077   G    A  0.005738   \n",
       "247    230704308_A_G            rs61731499   A    G  0.006557   \n",
       "251    230706171_C_T           rs139685563   C    T  0.001639   \n",
       "126    113897925_C_T           rs145803736   C    T  0.000820   \n",
       "129    113900051_A_T           rs149335605   A    T  0.000820   \n",
       "131    113900120_C_A           rs111785152   C    A  0.008197   \n",
       "123    113895442_A_C           rs148748734   A    C  0.001639   \n",
       "128    113900051_A_T           rs149335605   A    T  0.000820   \n",
       "130    113900120_C_A           rs111785152   C    A  0.008197   \n",
       "137    113901851_T_C           rs143769705   T    C  0.000820   \n",
       "4    110149715_A_AAT  chr13:110149715A-AAT   A  AAT  0.000825   \n",
       "85     110206692_T_C           rs140978802   T    C  0.001639   \n",
       "84     110206692_T_C           rs140978802   T    C  0.001639   \n",
       "5    110149715_A_AAT  chr13:110149715A-AAT   A  AAT  0.000825   \n",
       "3    110149715_A_AAT  chr13:110149715A-AAT   A  AAT  0.000825   \n",
       "219    206768666_T_G           rs568879359   T    G  0.001642   \n",
       "223    206771057_A_G           rs139352858   A    G  0.001639   \n",
       "226    206772380_C_T           rs139073251   C    T  0.000820   \n",
       "192     11796278_G_A            rs34279942   G    A  0.005738   \n",
       "152     11792285_T_G           rs145641996   T    G  0.001639   \n",
       "156     11792354_C_A            rs45449298   C    A  0.004918   \n",
       "178     11794442_C_G           rs200688214   C    G  0.001639   \n",
       "185     11795125_C_T           rs543016186   C    T  0.000820   \n",
       "187     11795341_T_G           rs142612062   T    G  0.006557   \n",
       "191     11796278_G_A            rs34279942   G    A  0.005738   \n",
       "193     11796319_C_T           rs150847674   C    T  0.001639   \n",
       "208     11802878_T_C             rs1413355   T    C  0.007377   \n",
       "211     11806126_C_T            rs13306560   C    T  0.001642   \n",
       "157     11792354_C_A            rs45449298   C    A  0.004918   \n",
       "154     11792354_C_A            rs45449298   C    A  0.004918   \n",
       "155     11792354_C_A            rs45449298   C    A  0.004918   \n",
       "\n",
       "     CADD_PHRED_SCORE  \n",
       "245             24.50  \n",
       "246             23.50  \n",
       "247             11.45  \n",
       "251             22.90  \n",
       "126             25.90  \n",
       "129             22.00  \n",
       "131             20.70  \n",
       "123             12.93  \n",
       "128             22.00  \n",
       "130             20.70  \n",
       "137             14.07  \n",
       "4               13.42  \n",
       "85              10.91  \n",
       "84              10.91  \n",
       "5               13.42  \n",
       "3               13.42  \n",
       "219             24.30  \n",
       "223             17.60  \n",
       "226             14.57  \n",
       "192             10.02  \n",
       "152             23.00  \n",
       "156             22.00  \n",
       "178             27.80  \n",
       "185             24.10  \n",
       "187             22.70  \n",
       "191             10.02  \n",
       "193             22.90  \n",
       "208             19.93  \n",
       "211             20.30  \n",
       "157             22.00  \n",
       "154             22.00  \n",
       "155             22.00  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display rare variants, and their associated clinical disease phenotype\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype = deleterious_variants_with_known_phenotype[deleterious_variants_with_known_phenotype.IH_AF<=0.01][['GENE','Clndn','ID','VAR_NAME','REF','ALT','IH_AF','CADD_PHRED_SCORE']].sort_values(by=['GENE','Clndn','ID'], ascending=True)\n",
    "rare_deleterious_variants_with_known_phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of variants with known phenotypes, CADD >=10 and that are common\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype.ID.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the frequencies of the rare variants with known phenotypes and CADD Phred scores in Africans >=10 compare to that of other global populations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ALFA global allele count data to African data in rare_deleterious_variants_with_known_phenotype dataframe\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa = pd.merge(rare_deleterious_variants_with_known_phenotype.drop_duplicates(subset=['VAR_NAME','REF','ALT']).drop(columns=['Clndn','ID','CADD_PHRED_SCORE']), alfa_global, how='left', left_on=['VAR_NAME','REF','ALT'], right_on=['variant_id','reference_allele','alternate_allele'])\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa = rare_deleterious_variants_with_known_phenotype_incl_alfa.drop(columns=['variant_id','reference_allele','alternate_allele'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>ALFA_AF_East Asian</th>\n",
       "      <th>ALFA_AF_South Asian</th>\n",
       "      <th>ALFA_AF_European</th>\n",
       "      <th>ALFA_AF_Latin American 1</th>\n",
       "      <th>ALFA_AF_Latin American 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGT</td>\n",
       "      <td>rs143479528</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGT</td>\n",
       "      <td>rs61751077</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGT</td>\n",
       "      <td>rs61731499</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.004926</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGT</td>\n",
       "      <td>rs139685563</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000509</td>\n",
       "      <td>0.001256</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs145803736</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs149335605</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs111785152</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.003953</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs148748734</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AP4B1</td>\n",
       "      <td>rs143769705</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>chr13:110149715A-AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>rs140978802</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IL10</td>\n",
       "      <td>rs568879359</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IL10</td>\n",
       "      <td>rs139352858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL10</td>\n",
       "      <td>rs139073251</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs34279942</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs145641996</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.002008</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs45449298</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs200688214</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs543016186</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs142612062</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs150847674</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs1413355</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.007377</td>\n",
       "      <td>0.007538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>rs13306560</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>0.039773</td>\n",
       "      <td>0.053909</td>\n",
       "      <td>0.046575</td>\n",
       "      <td>0.024429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE              VAR_NAME REF  ALT     IH_AF  ALFA_AF_East Asian  \\\n",
       "0      AGT           rs143479528   G    A  0.004098            0.000000   \n",
       "1      AGT            rs61751077   G    A  0.005738            0.000000   \n",
       "2      AGT            rs61731499   A    G  0.006557            0.000000   \n",
       "3      AGT           rs139685563   C    T  0.001639            0.000000   \n",
       "4    AP4B1           rs145803736   C    T  0.000820            0.000000   \n",
       "5    AP4B1           rs149335605   A    T  0.000820            0.000000   \n",
       "6    AP4B1           rs111785152   C    A  0.008197            0.000000   \n",
       "7    AP4B1           rs148748734   A    C  0.001639            0.000000   \n",
       "8    AP4B1           rs143769705   T    C  0.000820            0.000000   \n",
       "9   COL4A1  chr13:110149715A-AAT   A  AAT  0.000825                 NaN   \n",
       "10  COL4A1           rs140978802   T    C  0.001639            0.000000   \n",
       "11    IL10           rs568879359   T    G  0.001642            0.000000   \n",
       "12    IL10           rs139352858   A    G  0.001639            0.000000   \n",
       "13    IL10           rs139073251   C    T  0.000820            0.000000   \n",
       "14   MTHFR            rs34279942   G    A  0.005738            0.000000   \n",
       "15   MTHFR           rs145641996   T    G  0.001639            0.000000   \n",
       "16   MTHFR            rs45449298   C    A  0.004918            0.000000   \n",
       "17   MTHFR           rs200688214   C    G  0.001639            0.000000   \n",
       "18   MTHFR           rs543016186   C    T  0.000820            0.000000   \n",
       "19   MTHFR           rs142612062   T    G  0.006557            0.000000   \n",
       "20   MTHFR           rs150847674   C    T  0.001639            0.000000   \n",
       "21   MTHFR             rs1413355   T    C  0.007377            0.007538   \n",
       "22   MTHFR            rs13306560   C    T  0.001642            0.001894   \n",
       "\n",
       "    ALFA_AF_South Asian  ALFA_AF_European  ALFA_AF_Latin American 1  \\\n",
       "0              0.000000          0.000006                  0.000000   \n",
       "1              0.000000          0.000013                  0.000000   \n",
       "2              0.000000          0.000026                  0.004926   \n",
       "3              0.000000          0.000509                  0.001256   \n",
       "4              0.000000          0.000033                  0.000000   \n",
       "5              0.000000          0.000000                  0.000000   \n",
       "6              0.000000          0.000010                  0.003953   \n",
       "7              0.000000          0.000000                  0.000000   \n",
       "8              0.000000          0.000000                  0.000000   \n",
       "9                   NaN               NaN                       NaN   \n",
       "10             0.000000          0.000000                  0.000000   \n",
       "11             0.000000          0.000000                  0.000000   \n",
       "12             0.000000          0.000000                  0.000000   \n",
       "13             0.000000          0.000062                  0.000000   \n",
       "14             0.000000          0.000441                  0.001969   \n",
       "15             0.000000          0.000023                  0.002008   \n",
       "16             0.000000          0.000331                  0.000000   \n",
       "17             0.000000          0.000038                  0.000000   \n",
       "18             0.000000          0.000353                  0.000000   \n",
       "19             0.000000          0.000000                  0.000000   \n",
       "20             0.000000          0.000006                  0.000000   \n",
       "21             0.000000          0.000000                  0.000000   \n",
       "22             0.039773          0.053909                  0.046575   \n",
       "\n",
       "    ALFA_AF_Latin American 2  \n",
       "0                   0.000000  \n",
       "1                   0.001053  \n",
       "2                   0.000000  \n",
       "3                   0.000000  \n",
       "4                   0.000000  \n",
       "5                   0.000000  \n",
       "6                   0.000000  \n",
       "7                   0.000000  \n",
       "8                   0.000000  \n",
       "9                        NaN  \n",
       "10                  0.000000  \n",
       "11                  0.000000  \n",
       "12                  0.000000  \n",
       "13                  0.000000  \n",
       "14                  0.000000  \n",
       "15                  0.000000  \n",
       "16                  0.000000  \n",
       "17                  0.000000  \n",
       "18                  0.000000  \n",
       "19                  0.000000  \n",
       "20                  0.000000  \n",
       "21                  0.000000  \n",
       "22                  0.024429  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate the alternate allele frequencies for the global populations\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_East Asian'] = rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_East Asian']/(rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_East Asian']+rare_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_East Asian'])\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_South Asian'] = rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_South Asian']/(rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_South Asian']+rare_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_South Asian'])\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_European'] = rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_European']/(rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_European']+rare_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_European'])\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_Latin American 1'] = rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 1']/(rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 1']+rare_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_Latin American 1'])\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa['ALFA_AF_Latin American 2'] = rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 2']/(rare_deleterious_variants_with_known_phenotype_incl_alfa['ALT_CT_ALFA_Latin American 2']+rare_deleterious_variants_with_known_phenotype_incl_alfa['REF_CT_ALFA_Latin American 2'])\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa.drop(columns=['ALT_CT_ALFA_East Asian','ALT_CT_ALFA_South Asian','ALT_CT_ALFA_European','ALT_CT_ALFA_Latin American 1','ALT_CT_ALFA_Latin American 2','REF_CT_ALFA_East Asian','REF_CT_ALFA_South Asian','REF_CT_ALFA_European','REF_CT_ALFA_Latin American 1','REF_CT_ALFA_Latin American 2'], inplace=True)\n",
    "rare_deleterious_variants_with_known_phenotype_incl_alfa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the frequencies of the rare variants with known phenotypes and CADD Phred scores >=10 in each African ethnolinguistic population group?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of variants to get frequencies for.\n",
    "\n",
    "rare_deleterious_variants_with_known_phenotype_list = list(rare_deleterious_variants_with_known_phenotype.ID.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract ethnolinguistic population frequencies for each of these variants. \n",
    "\n",
    "ih_afr_subpops_rare_deleterious_variants = ih_afr[(ih_afr.ID.isin(rare_deleterious_variants_with_known_phenotype_list)) & ~ (ih_afr.REG == 'Recent African')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out irrelevant data. Pivot data.\n",
    "\n",
    "ih_afr_subpops_rare_deleterious_variants_filtered = ih_afr_subpops_rare_deleterious_variants.drop(columns=['POS','REF','ALT','IH_ALT_CTS','IH_TOTAL_CTS','IH_REF_CTS','VARIANT_TYPE','REG'])\n",
    "ih_afr_subpops_rare_deleterious_variants_filtered = ih_afr_subpops_rare_deleterious_variants_filtered.pivot(index=['ID','VAR_NAME','GENE'], columns=['SUB_POP'], values=['IH_AF']).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_2872\\320947548.py:3: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left, 2 on the right)\n",
      "  ih_afr_subpops_rare_deleterious_variants_filtered = pd.merge(left=rare_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_rare_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n",
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_2872\\320947548.py:3: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "  ih_afr_subpops_rare_deleterious_variants_filtered = pd.merge(left=rare_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_rare_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>(ID, )</th>\n",
       "      <th>(GENE, )</th>\n",
       "      <th>(IH_AF, Bantu Kenya)</th>\n",
       "      <th>(IH_AF, Bantu South Africa)</th>\n",
       "      <th>(IH_AF, Biaka Pygmy)</th>\n",
       "      <th>(IH_AF, Esan)</th>\n",
       "      <th>(IH_AF, Luhya)</th>\n",
       "      <th>(IH_AF, Mandenka)</th>\n",
       "      <th>(IH_AF, Mandinka)</th>\n",
       "      <th>(IH_AF, Mbuti Pygmy)</th>\n",
       "      <th>(IH_AF, Mende)</th>\n",
       "      <th>(IH_AF, San)</th>\n",
       "      <th>(IH_AF, Yoruba)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs143479528</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>230703157_G_A</td>\n",
       "      <td>AGT</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014563</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs61751077</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>230703274_G_A</td>\n",
       "      <td>AGT</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs61731499</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>230704308_A_G</td>\n",
       "      <td>AGT</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.007246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs139685563</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>230706171_C_T</td>\n",
       "      <td>AGT</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs145803736</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>113897925_C_T</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs149335605</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>113900051_A_T</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs111785152</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>113900120_C_A</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.025862</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs148748734</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>113895442_A_C</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs143769705</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>113901851_T_C</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr13:110149715A-AAT</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>110149715_A_AAT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs140978802</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>110206692_T_C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs568879359</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>206768666_T_G</td>\n",
       "      <td>IL10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs139352858</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>206771057_A_G</td>\n",
       "      <td>IL10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs139073251</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>206772380_C_T</td>\n",
       "      <td>IL10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.004310</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs34279942</td>\n",
       "      <td>0.005738</td>\n",
       "      <td>11796278_G_A</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027174</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs145641996</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>11792285_T_G</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs45449298</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>11792354_C_A</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.016304</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs200688214</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>11794442_C_G</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs543016186</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>11795125_C_T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004854</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs142612062</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>11795341_T_G</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.021552</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs150847674</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>11796319_C_T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs1413355</td>\n",
       "      <td>0.007377</td>\n",
       "      <td>11802878_T_C</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs13306560</td>\n",
       "      <td>0.001642</td>\n",
       "      <td>11806126_C_T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                VAR_NAME     IH_AF           (ID, ) (GENE, )  \\\n",
       "0            rs143479528  0.004098    230703157_G_A      AGT   \n",
       "1             rs61751077  0.005738    230703274_G_A      AGT   \n",
       "2             rs61731499  0.006557    230704308_A_G      AGT   \n",
       "3            rs139685563  0.001639    230706171_C_T      AGT   \n",
       "4            rs145803736  0.000820    113897925_C_T    AP4B1   \n",
       "5            rs149335605  0.000820    113900051_A_T    AP4B1   \n",
       "6            rs111785152  0.008197    113900120_C_A    AP4B1   \n",
       "7            rs148748734  0.001639    113895442_A_C    AP4B1   \n",
       "8            rs143769705  0.000820    113901851_T_C    AP4B1   \n",
       "9   chr13:110149715A-AAT  0.000825  110149715_A_AAT   COL4A1   \n",
       "10           rs140978802  0.001639    110206692_T_C   COL4A1   \n",
       "11           rs568879359  0.001642    206768666_T_G     IL10   \n",
       "12           rs139352858  0.001639    206771057_A_G     IL10   \n",
       "13           rs139073251  0.000820    206772380_C_T     IL10   \n",
       "14            rs34279942  0.005738     11796278_G_A    MTHFR   \n",
       "15           rs145641996  0.001639     11792285_T_G    MTHFR   \n",
       "16            rs45449298  0.004918     11792354_C_A    MTHFR   \n",
       "17           rs200688214  0.001639     11794442_C_G    MTHFR   \n",
       "18           rs543016186  0.000820     11795125_C_T    MTHFR   \n",
       "19           rs142612062  0.006557     11795341_T_G    MTHFR   \n",
       "20           rs150847674  0.001639     11796319_C_T    MTHFR   \n",
       "21             rs1413355  0.007377     11802878_T_C    MTHFR   \n",
       "22            rs13306560  0.001642     11806126_C_T    MTHFR   \n",
       "\n",
       "    (IH_AF, Bantu Kenya)  (IH_AF, Bantu South Africa)  (IH_AF, Biaka Pygmy)  \\\n",
       "0                   0.00                          0.0                   0.0   \n",
       "1                   0.00                          0.0                   0.0   \n",
       "2                   0.05                          0.0                   0.0   \n",
       "3                   0.00                          0.0                   0.0   \n",
       "4                   0.00                          0.0                   0.0   \n",
       "5                   0.00                          0.0                   0.0   \n",
       "6                   0.00                          0.0                   0.0   \n",
       "7                   0.00                          0.0                   0.0   \n",
       "8                   0.00                          0.0                   0.0   \n",
       "9                   0.00                          0.0                   0.0   \n",
       "10                  0.00                          0.0                   0.0   \n",
       "11                  0.00                          0.0                   0.0   \n",
       "12                  0.00                          0.0                   0.0   \n",
       "13                  0.00                          0.0                   0.0   \n",
       "14                  0.00                          0.0                   0.0   \n",
       "15                  0.00                          0.0                   0.0   \n",
       "16                  0.00                          0.0                   0.0   \n",
       "17                  0.05                          0.0                   0.0   \n",
       "18                  0.00                          0.0                   0.0   \n",
       "19                  0.00                          0.0                   0.0   \n",
       "20                  0.00                          0.0                   0.0   \n",
       "21                  0.10                          0.0                   0.0   \n",
       "22                  0.00                          0.0                   0.0   \n",
       "\n",
       "    (IH_AF, Esan)  (IH_AF, Luhya)  (IH_AF, Mandenka)  (IH_AF, Mandinka)  \\\n",
       "0        0.014563        0.000000               0.00           0.000000   \n",
       "1        0.000000        0.000000               0.00           0.008621   \n",
       "2        0.000000        0.021739               0.00           0.000000   \n",
       "3        0.000000        0.010870               0.00           0.000000   \n",
       "4        0.000000        0.005435               0.00           0.000000   \n",
       "5        0.000000        0.005435               0.00           0.000000   \n",
       "6        0.000000        0.005435               0.00           0.025862   \n",
       "7        0.004854        0.000000               0.00           0.000000   \n",
       "8        0.000000        0.000000               0.00           0.000000   \n",
       "9        0.000000        0.000000               0.00           0.000000   \n",
       "10       0.000000        0.000000               0.00           0.008621   \n",
       "11       0.000000        0.000000               0.00           0.000000   \n",
       "12       0.004854        0.000000               0.00           0.000000   \n",
       "13       0.000000        0.000000               0.00           0.004310   \n",
       "14       0.000000        0.027174               0.00           0.000000   \n",
       "15       0.000000        0.000000               0.00           0.000000   \n",
       "16       0.004854        0.016304               0.00           0.000000   \n",
       "17       0.000000        0.005435               0.00           0.000000   \n",
       "18       0.004854        0.000000               0.00           0.000000   \n",
       "19       0.000000        0.000000               0.05           0.021552   \n",
       "20       0.009709        0.000000               0.00           0.000000   \n",
       "21       0.019417        0.005435               0.00           0.000000   \n",
       "22       0.000000        0.000000               0.00           0.008621   \n",
       "\n",
       "    (IH_AF, Mbuti Pygmy)  (IH_AF, Mende)  (IH_AF, San)  (IH_AF, Yoruba)  \n",
       "0               0.000000        0.012048      0.000000         0.000000  \n",
       "1               0.000000        0.024096      0.000000         0.003623  \n",
       "2               0.000000        0.000000      0.083333         0.007246  \n",
       "3               0.000000        0.000000      0.000000         0.000000  \n",
       "4               0.000000        0.000000      0.000000         0.000000  \n",
       "5               0.000000        0.000000      0.000000         0.000000  \n",
       "6               0.000000        0.018072      0.000000         0.000000  \n",
       "7               0.000000        0.000000      0.000000         0.003623  \n",
       "8               0.041667        0.000000      0.000000         0.000000  \n",
       "9               0.000000        0.000000      0.000000         0.003623  \n",
       "10              0.000000        0.000000      0.000000         0.000000  \n",
       "11              0.000000        0.006024      0.000000         0.003623  \n",
       "12              0.000000        0.000000      0.000000         0.003623  \n",
       "13              0.000000        0.000000      0.000000         0.000000  \n",
       "14              0.000000        0.000000      0.000000         0.007246  \n",
       "15              0.000000        0.006024      0.000000         0.003623  \n",
       "16              0.000000        0.000000      0.000000         0.007246  \n",
       "17              0.000000        0.000000      0.000000         0.000000  \n",
       "18              0.000000        0.000000      0.000000         0.000000  \n",
       "19              0.000000        0.006024      0.000000         0.000000  \n",
       "20              0.000000        0.000000      0.000000         0.000000  \n",
       "21              0.000000        0.000000      0.000000         0.007246  \n",
       "22              0.000000        0.000000      0.000000         0.000000  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add overall African frequencies\n",
    "\n",
    "ih_afr_subpops_rare_deleterious_variants_filtered = pd.merge(left=rare_deleterious_variants_with_known_phenotype_incl_alfa[['VAR_NAME','IH_AF']], right=ih_afr_subpops_rare_deleterious_variants_filtered, how='left', on=['VAR_NAME'])\n",
    "ih_afr_subpops_rare_deleterious_variants_filtered"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
