{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistical comparison of genetic variant frequencies: In-house vs ALFA database\n",
    "\n",
    "This analysis aimed to statistically compare the frequencies of variant alleles derived from in-house African genetic data to that of data obtained from the [NCBI ALFA database](https://www.ncbi.nlm.nih.gov/snp/docs/gsr/alfa/) for African, European and Asian populations using a two-tailed Fisher's Exact Test. \n",
    "\n",
    "To achieve this, the following steps were performed: \n",
    "\n",
    "1. In-house Data Generation: In-house genetic variant allele count data was generated from genetic information obtained from the [GnomAD 1000 Genomes and HGDP datasets](https://gnomad.broadinstitute.org/news/2021-10-gnomad-v3-1-2-minor-release/) for African population. This data was prepared for analysis by following the steps outlined in the `Notebooks\\Data_preparation\\2-IH_allele_counts.ipynb` Jupyter notebook. The prepared data is stored in `Data\\Processed\\IH_allele_counts_fishers.csv` and `Data\\Processed\\IH_allele_counts.csv`.\n",
    "2. ALFA Data Retrieval: Variant allele count data for variants with rsIDs was retrieved from the NCBI ALFA database. This data was prepared for analysis by following the steps outlined in the  `Notebooks\\Data_preparation\\3-ALFA_allele_counts_a.ipynb` and `Notebooks\\Data_preparation\\3-ALFA_allele_counts_b.ipynb` notebooks. The prepared data is stored in `Data\\Processed\\ALFA_allele_counts_b.csv`\n",
    "3. Data Integration: The variant allele count data generated in-house and the data fetched from ALFA were combined into a unified dataframe for analysis.\n",
    "4. Data Filtering: The data was filtered to encompass only the data for specific population groups including the ALFA African Others, EUR, EAS, SAS populations, as well as the averaged Recent African population.\n",
    "5. Fisher's Exact Tests: Two-tailed Fisher's Exact Tests were conducted to test for significant differences in variant allele frequencies among the selected population groups. Additionally, to account for multiple comparisons, the Bonferroni method was employed to correct the Fisher's test results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None \n",
    "import json\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from upsetplot import plot\n",
    "from requests import get, codes as http_code\n",
    "from warnings import simplefilter\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import in-house variant allele count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th>ALT_CT_IH_ACB</th>\n",
       "      <th>ALT_CT_IH_ASW</th>\n",
       "      <th>ALT_CT_IH_African</th>\n",
       "      <th>ALT_CT_IH_CA</th>\n",
       "      <th>...</th>\n",
       "      <th>ALT_CT_IH_SA</th>\n",
       "      <th>ALT_CT_IH_WA</th>\n",
       "      <th>CORR_REF_CT_IH_ACB</th>\n",
       "      <th>CORR_REF_CT_IH_ASW</th>\n",
       "      <th>CORR_REF_CT_IH_African</th>\n",
       "      <th>CORR_REF_CT_IH_CA</th>\n",
       "      <th>CORR_REF_CT_IH_EA</th>\n",
       "      <th>CORR_REF_CT_IH_Recent African</th>\n",
       "      <th>CORR_REF_CT_IH_SA</th>\n",
       "      <th>CORR_REF_CT_IH_WA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:34438836T-C</td>\n",
       "      <td>34438836_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1977</td>\n",
       "      <td>80</td>\n",
       "      <td>217</td>\n",
       "      <td>1607</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11:34438889G-C</td>\n",
       "      <td>34438889_C_G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438889</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1977</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1607</td>\n",
       "      <td>27</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:34438910C-T</td>\n",
       "      <td>34438910_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438910</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:34439179A-G</td>\n",
       "      <td>34439179_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439179</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:34439188C-G</td>\n",
       "      <td>34439188_G_C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439188</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            VAR_NAME            ID REF ALT GENE       POS  ALT_CT_IH_ACB  \\\n",
       "0  chr11:34438836T-C  34438836_C_T   T   C  CAT  34438836              0   \n",
       "1  chr11:34438889G-C  34438889_C_G   G   C  CAT  34438889              0   \n",
       "2  chr11:34438910C-T  34438910_T_C   C   T  CAT  34438910              0   \n",
       "3  chr11:34439179A-G  34439179_G_A   A   G  CAT  34439179              0   \n",
       "4  chr11:34439188C-G  34439188_G_C   C   G  CAT  34439188              0   \n",
       "\n",
       "   ALT_CT_IH_ASW  ALT_CT_IH_African  ALT_CT_IH_CA  ...  ALT_CT_IH_SA  \\\n",
       "0              0                  1             0  ...             0   \n",
       "1              0                  1             0  ...             1   \n",
       "2              0                  0             0  ...             0   \n",
       "3              0                  0             0  ...             0   \n",
       "4              0                  0             0  ...             0   \n",
       "\n",
       "   ALT_CT_IH_WA  CORR_REF_CT_IH_ACB  CORR_REF_CT_IH_ASW  \\\n",
       "0             0                 228                 142   \n",
       "1             0                 228                 142   \n",
       "2             0                 228                 142   \n",
       "3             0                 228                 142   \n",
       "4             0                 228                 142   \n",
       "\n",
       "   CORR_REF_CT_IH_African  CORR_REF_CT_IH_CA  CORR_REF_CT_IH_EA  \\\n",
       "0                    1977                 80                217   \n",
       "1                    1977                 80                218   \n",
       "2                    1978                 80                218   \n",
       "3                    1978                 80                218   \n",
       "4                    1978                 80                218   \n",
       "\n",
       "   CORR_REF_CT_IH_Recent African  CORR_REF_CT_IH_SA  CORR_REF_CT_IH_WA  \n",
       "0                           1607                 28               1282  \n",
       "1                           1607                 27               1282  \n",
       "2                           1608                 28               1282  \n",
       "3                           1608                 28               1282  \n",
       "4                           1608                 28               1282  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts_fishers.csv\",\n",
    "    )\n",
    ").drop(columns=\"Unnamed: 0\")\n",
    "\n",
    "ih_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import ALFA variant allele count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>ALT_CT_ALFA_African</th>\n",
       "      <th>ALT_CT_ALFA_African American</th>\n",
       "      <th>ALT_CT_ALFA_African Others</th>\n",
       "      <th>ALT_CT_ALFA_Asian</th>\n",
       "      <th>ALT_CT_ALFA_East Asian</th>\n",
       "      <th>ALT_CT_ALFA_European</th>\n",
       "      <th>ALT_CT_ALFA_Latin American 1</th>\n",
       "      <th>...</th>\n",
       "      <th>REF_CT_ALFA_African Others</th>\n",
       "      <th>REF_CT_ALFA_Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_Latin American 1</th>\n",
       "      <th>REF_CT_ALFA_Latin American 2</th>\n",
       "      <th>REF_CT_ALFA_Other</th>\n",
       "      <th>REF_CT_ALFA_Other Asian</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000343</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>695.0</td>\n",
       "      <td>671.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>296.0</td>\n",
       "      <td>622.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>109377.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>2200.0</td>\n",
       "      <td>5457.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>127076.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000989</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1435.0</td>\n",
       "      <td>1388.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>21489.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>...</td>\n",
       "      <td>135.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>37269.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>2052.0</td>\n",
       "      <td>1371.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>3283.0</td>\n",
       "      <td>48171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000990</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>798.0</td>\n",
       "      <td>767.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>5355.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>...</td>\n",
       "      <td>83.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>8931.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>12125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1005573</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3120.0</td>\n",
       "      <td>3009.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>10693.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>4955.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>956.0</td>\n",
       "      <td>474.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>7130.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1007311</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1970.0</td>\n",
       "      <td>1903.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9154.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>11242.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>456.0</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>15040.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  variant_id reference_allele alternate_allele  ALT_CT_ALFA_African  \\\n",
       "0  rs1000343                C                T                695.0   \n",
       "1  rs1000989                T                C               1435.0   \n",
       "2  rs1000990                T                C                798.0   \n",
       "3  rs1005573                C                T               3120.0   \n",
       "4  rs1007311                A                G               1970.0   \n",
       "\n",
       "   ALT_CT_ALFA_African American  ALT_CT_ALFA_African Others  \\\n",
       "0                         671.0                        24.0   \n",
       "1                        1388.0                        47.0   \n",
       "2                         767.0                        31.0   \n",
       "3                        3009.0                       111.0   \n",
       "4                        1903.0                        67.0   \n",
       "\n",
       "   ALT_CT_ALFA_Asian  ALT_CT_ALFA_East Asian  ALT_CT_ALFA_European  \\\n",
       "0                0.0                     0.0                  49.0   \n",
       "1               80.0                    55.0               21489.0   \n",
       "2               42.0                    32.0                5355.0   \n",
       "3               63.0                    35.0               10693.0   \n",
       "4               88.0                    56.0                9154.0   \n",
       "\n",
       "   ALT_CT_ALFA_Latin American 1  ...  REF_CT_ALFA_African Others  \\\n",
       "0                           5.0  ...                       296.0   \n",
       "1                         123.0  ...                       135.0   \n",
       "2                          40.0  ...                        83.0   \n",
       "3                         209.0  ...                        15.0   \n",
       "4                          61.0  ...                        55.0   \n",
       "\n",
       "   REF_CT_ALFA_Asian  REF_CT_ALFA_East Asian  REF_CT_ALFA_European  \\\n",
       "0              622.0                   490.0              109377.0   \n",
       "1              158.0                   109.0               37269.0   \n",
       "2               70.0                    54.0                8931.0   \n",
       "3              101.0                    69.0                4955.0   \n",
       "4               80.0                    56.0               11242.0   \n",
       "\n",
       "   REF_CT_ALFA_Latin American 1  REF_CT_ALFA_Latin American 2  \\\n",
       "0                         673.0                        2200.0   \n",
       "1                         273.0                        2052.0   \n",
       "2                         106.0                         349.0   \n",
       "3                          87.0                         956.0   \n",
       "4                          85.0                         456.0   \n",
       "\n",
       "   REF_CT_ALFA_Other  REF_CT_ALFA_Other Asian  REF_CT_ALFA_South Asian  \\\n",
       "0             5457.0                    132.0                    184.0   \n",
       "1             1371.0                     49.0                   3283.0   \n",
       "2              459.0                     16.0                     62.0   \n",
       "3              474.0                     32.0                     31.0   \n",
       "4             1538.0                     24.0                     55.0   \n",
       "\n",
       "   REF_CT_ALFA_Total  \n",
       "0           127076.0  \n",
       "1            48171.0  \n",
       "2            12125.0  \n",
       "3             7130.0  \n",
       "4            15040.0  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alfa_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"ALFA_allele_counts_b.csv\",\n",
    "    )\n",
    ").drop(columns=\"Unnamed: 0\")\n",
    "\n",
    "alfa_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data integration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter in-house data for variants with rsIDs. ALFA data is only available for variants with rsIDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_data_rsids = ih_data[ih_data.VAR_NAME.str.contains(\"rs\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge in-house and ALFA variant count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th>ALT_CT_IH_ACB</th>\n",
       "      <th>ALT_CT_IH_ASW</th>\n",
       "      <th>ALT_CT_IH_African</th>\n",
       "      <th>ALT_CT_IH_CA</th>\n",
       "      <th>...</th>\n",
       "      <th>REF_CT_ALFA_African Others</th>\n",
       "      <th>REF_CT_ALFA_Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_Latin American 1</th>\n",
       "      <th>REF_CT_ALFA_Latin American 2</th>\n",
       "      <th>REF_CT_ALFA_Other</th>\n",
       "      <th>REF_CT_ALFA_Other Asian</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000343</td>\n",
       "      <td>34452510_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34452510</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>172</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>296.0</td>\n",
       "      <td>622.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>109377.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>2200.0</td>\n",
       "      <td>5457.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>127076.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000989</td>\n",
       "      <td>110174956_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110174956</td>\n",
       "      <td>56</td>\n",
       "      <td>44</td>\n",
       "      <td>522</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>135.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>37269.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>2052.0</td>\n",
       "      <td>1371.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>3283.0</td>\n",
       "      <td>48171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000990</td>\n",
       "      <td>110174971_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110174971</td>\n",
       "      <td>56</td>\n",
       "      <td>44</td>\n",
       "      <td>519</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>83.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>8931.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>12125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1005573</td>\n",
       "      <td>33026408_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>33026408</td>\n",
       "      <td>204</td>\n",
       "      <td>118</td>\n",
       "      <td>1731</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>4955.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>956.0</td>\n",
       "      <td>474.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>7130.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1007311</td>\n",
       "      <td>150998920_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>150998920</td>\n",
       "      <td>145</td>\n",
       "      <td>88</td>\n",
       "      <td>1133</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>11242.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>456.0</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>15040.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    VAR_NAME             ID REF ALT    GENE        POS  ALT_CT_IH_ACB  \\\n",
       "0  rs1000343   34452510_T_C   C   T     CAT   34452510             24   \n",
       "1  rs1000989  110174956_C_T   T   C  COL4A1  110174956             56   \n",
       "2  rs1000990  110174971_C_T   T   C  COL4A1  110174971             56   \n",
       "3  rs1005573   33026408_T_C   C   T   OLIG2   33026408            204   \n",
       "4  rs1007311  150998920_G_A   A   G    NOS3  150998920            145   \n",
       "\n",
       "   ALT_CT_IH_ASW  ALT_CT_IH_African  ALT_CT_IH_CA  ...  \\\n",
       "0             17                172             1  ...   \n",
       "1             44                522            29  ...   \n",
       "2             44                519            29  ...   \n",
       "3            118               1731            70  ...   \n",
       "4             88               1133            24  ...   \n",
       "\n",
       "   REF_CT_ALFA_African Others  REF_CT_ALFA_Asian  REF_CT_ALFA_East Asian  \\\n",
       "0                       296.0              622.0                   490.0   \n",
       "1                       135.0              158.0                   109.0   \n",
       "2                        83.0               70.0                    54.0   \n",
       "3                        15.0              101.0                    69.0   \n",
       "4                        55.0               80.0                    56.0   \n",
       "\n",
       "   REF_CT_ALFA_European  REF_CT_ALFA_Latin American 1  \\\n",
       "0              109377.0                         673.0   \n",
       "1               37269.0                         273.0   \n",
       "2                8931.0                         106.0   \n",
       "3                4955.0                          87.0   \n",
       "4               11242.0                          85.0   \n",
       "\n",
       "   REF_CT_ALFA_Latin American 2  REF_CT_ALFA_Other  REF_CT_ALFA_Other Asian  \\\n",
       "0                        2200.0             5457.0                    132.0   \n",
       "1                        2052.0             1371.0                     49.0   \n",
       "2                         349.0              459.0                     16.0   \n",
       "3                         956.0              474.0                     32.0   \n",
       "4                         456.0             1538.0                     24.0   \n",
       "\n",
       "   REF_CT_ALFA_South Asian  REF_CT_ALFA_Total  \n",
       "0                    184.0           127076.0  \n",
       "1                   3283.0            48171.0  \n",
       "2                     62.0            12125.0  \n",
       "3                     31.0             7130.0  \n",
       "4                     55.0            15040.0  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_alfa_data = ih_data_rsids.merge(\n",
    "    alfa_data,\n",
    "    how=\"left\",\n",
    "    left_on=[\"VAR_NAME\", \"REF\", \"ALT\"],\n",
    "    right_on=[\"variant_id\", \"reference_allele\", \"alternate_allele\"],\n",
    ").drop(columns=[\"variant_id\", \"reference_allele\", \"alternate_allele\"])\n",
    "\n",
    "ih_alfa_data = ih_alfa_data.replace(np.NAN, 0)\n",
    "\n",
    "ih_alfa_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data filtering\n",
    "Filter the data to include only the specific comparison populations of interest. Retain the ALFA African Others, EUR, EAS, SAS populations, as well as the averaged Recent African population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ALT_CT_IH_Recent African</th>\n",
       "      <th>REF_CT_IH_Recent African</th>\n",
       "      <th>ALT_CT_ALFA_African Others</th>\n",
       "      <th>REF_CT_ALFA_African Others</th>\n",
       "      <th>ALT_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>ALT_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "      <th>ALT_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14929</th>\n",
       "      <td>110254294_C_T</td>\n",
       "      <td>rs9805608</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>110254294</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>206</td>\n",
       "      <td>1402</td>\n",
       "      <td>33.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>9552.0</td>\n",
       "      <td>21776.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14930</th>\n",
       "      <td>48242167_A_G</td>\n",
       "      <td>rs9807869</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>48242167</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>479</td>\n",
       "      <td>1129</td>\n",
       "      <td>30.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9808.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14931</th>\n",
       "      <td>110198689_C_T</td>\n",
       "      <td>rs995223</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>110198689</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>695</td>\n",
       "      <td>913</td>\n",
       "      <td>71.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>34291.0</td>\n",
       "      <td>96137.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14932</th>\n",
       "      <td>110198495_G_A</td>\n",
       "      <td>rs995224</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>110198495</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>339</td>\n",
       "      <td>1269</td>\n",
       "      <td>10.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>8596.0</td>\n",
       "      <td>25474.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14933</th>\n",
       "      <td>33027838_T_G</td>\n",
       "      <td>rs9982080</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>33027838</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>232</td>\n",
       "      <td>1376</td>\n",
       "      <td>13.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>1513.0</td>\n",
       "      <td>12773.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID   VAR_NAME REF ALT        POS    GENE  \\\n",
       "14929  110254294_C_T  rs9805608   T   C  110254294  COL4A1   \n",
       "14930   48242167_A_G  rs9807869   G   A   48242167   CARD8   \n",
       "14931  110198689_C_T   rs995223   T   C  110198689  COL4A1   \n",
       "14932  110198495_G_A   rs995224   A   G  110198495  COL4A1   \n",
       "14933   33027838_T_G  rs9982080   G   T   33027838   OLIG2   \n",
       "\n",
       "       ALT_CT_IH_Recent African  REF_CT_IH_Recent African  \\\n",
       "14929                       206                      1402   \n",
       "14930                       479                      1129   \n",
       "14931                       695                       913   \n",
       "14932                       339                      1269   \n",
       "14933                       232                      1376   \n",
       "\n",
       "       ALT_CT_ALFA_African Others  REF_CT_ALFA_African Others  \\\n",
       "14929                        33.0                       175.0   \n",
       "14930                        30.0                        84.0   \n",
       "14931                        71.0                       139.0   \n",
       "14932                        10.0                        80.0   \n",
       "14933                        13.0                       101.0   \n",
       "\n",
       "       ALT_CT_ALFA_East Asian  REF_CT_ALFA_East Asian  \\\n",
       "14929                    23.0                    75.0   \n",
       "14930                     1.0                    85.0   \n",
       "14931                    40.0                   454.0   \n",
       "14932                     3.0                   101.0   \n",
       "14933                    10.0                    76.0   \n",
       "\n",
       "       ALT_CT_ALFA_South Asian  REF_CT_ALFA_South Asian  ALT_CT_ALFA_European  \\\n",
       "14929                     53.0                     61.0                9552.0   \n",
       "14930                      0.0                     98.0                  16.0   \n",
       "14931                     24.0                    132.0               34291.0   \n",
       "14932                      1.0                     59.0                8596.0   \n",
       "14933                      9.0                     89.0                1513.0   \n",
       "\n",
       "       REF_CT_ALFA_European  \n",
       "14929               21776.0  \n",
       "14930                9808.0  \n",
       "14931               96137.0  \n",
       "14932               25474.0  \n",
       "14933               12773.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_alfa_comparisons = ih_alfa_data[\n",
    "    [\n",
    "        \"ID\",\n",
    "        \"VAR_NAME\",\n",
    "        \"REF\", \n",
    "        \"ALT\",\n",
    "        \"POS\",\n",
    "        \"GENE\",\n",
    "        \"ALT_CT_IH_Recent African\",\n",
    "        \"CORR_REF_CT_IH_Recent African\",\n",
    "        \"ALT_CT_ALFA_African Others\",\n",
    "        \"REF_CT_ALFA_African Others\",\n",
    "        \"ALT_CT_ALFA_East Asian\",\n",
    "        \"REF_CT_ALFA_East Asian\",\n",
    "        \"ALT_CT_ALFA_South Asian\",\n",
    "        \"REF_CT_ALFA_South Asian\",\n",
    "        \"ALT_CT_ALFA_European\",\n",
    "        \"REF_CT_ALFA_European\",\n",
    "    ]\n",
    "].rename(columns={\"CORR_REF_CT_IH_Recent African\": \"REF_CT_IH_Recent African\"})\n",
    "\n",
    "ih_alfa_comparisons.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replace missing allele frequencies with 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_alfa_comparisons = ih_alfa_comparisons.replace(np.NAN, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove rows with frequencies of 0. We only want to include variants that are present in all comparison populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows where population groups alt allele counts are 0. These rows contain variants that are not present in one of the populations.\n",
    "\n",
    "ih_alfa_comparisons = ih_alfa_comparisons[\n",
    "    (ih_alfa_comparisons[\"ALT_CT_IH_Recent African\"] != 0)\n",
    "    & (ih_alfa_comparisons[\"ALT_CT_ALFA_African Others\"] != 0)\n",
    "    & (ih_alfa_comparisons[\"ALT_CT_ALFA_East Asian\"] != 0)\n",
    "    & (ih_alfa_comparisons[\"ALT_CT_ALFA_South Asian\"] != 0)\n",
    "    & (ih_alfa_comparisons[\"ALT_CT_ALFA_European\"] != 0)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ALT_CT_IH_Recent African</th>\n",
       "      <th>REF_CT_IH_Recent African</th>\n",
       "      <th>ALT_CT_ALFA_African Others</th>\n",
       "      <th>REF_CT_ALFA_African Others</th>\n",
       "      <th>ALT_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>ALT_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "      <th>ALT_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110174956_C_T</td>\n",
       "      <td>rs1000989</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>110174956</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>422</td>\n",
       "      <td>1186</td>\n",
       "      <td>47.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>1685.0</td>\n",
       "      <td>3283.0</td>\n",
       "      <td>21489.0</td>\n",
       "      <td>37269.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110174971_C_T</td>\n",
       "      <td>rs1000990</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>110174971</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>419</td>\n",
       "      <td>1189</td>\n",
       "      <td>31.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>5355.0</td>\n",
       "      <td>8931.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33026408_T_C</td>\n",
       "      <td>rs1005573</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>33026408</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>1409</td>\n",
       "      <td>199</td>\n",
       "      <td>111.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>10693.0</td>\n",
       "      <td>4955.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>150998920_G_A</td>\n",
       "      <td>rs1007311</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>150998920</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>900</td>\n",
       "      <td>708</td>\n",
       "      <td>67.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>9154.0</td>\n",
       "      <td>11242.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>151008522_C_T</td>\n",
       "      <td>rs1008140</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>151008522</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>356</td>\n",
       "      <td>1252</td>\n",
       "      <td>24.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>829.0</td>\n",
       "      <td>39989.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID   VAR_NAME REF ALT        POS    GENE  \\\n",
       "1  110174956_C_T  rs1000989   T   C  110174956  COL4A1   \n",
       "2  110174971_C_T  rs1000990   T   C  110174971  COL4A1   \n",
       "3   33026408_T_C  rs1005573   C   T   33026408   OLIG2   \n",
       "4  150998920_G_A  rs1007311   A   G  150998920    NOS3   \n",
       "5  151008522_C_T  rs1008140   T   C  151008522    NOS3   \n",
       "\n",
       "   ALT_CT_IH_Recent African  REF_CT_IH_Recent African  \\\n",
       "1                       422                      1186   \n",
       "2                       419                      1189   \n",
       "3                      1409                       199   \n",
       "4                       900                       708   \n",
       "5                       356                      1252   \n",
       "\n",
       "   ALT_CT_ALFA_African Others  REF_CT_ALFA_African Others  \\\n",
       "1                        47.0                       135.0   \n",
       "2                        31.0                        83.0   \n",
       "3                       111.0                        15.0   \n",
       "4                        67.0                        55.0   \n",
       "5                        24.0                       100.0   \n",
       "\n",
       "   ALT_CT_ALFA_East Asian  REF_CT_ALFA_East Asian  ALT_CT_ALFA_South Asian  \\\n",
       "1                    55.0                   109.0                   1685.0   \n",
       "2                    32.0                    54.0                     36.0   \n",
       "3                    35.0                    69.0                     79.0   \n",
       "4                    56.0                    56.0                     43.0   \n",
       "5                    18.0                   126.0                      1.0   \n",
       "\n",
       "   REF_CT_ALFA_South Asian  ALT_CT_ALFA_European  REF_CT_ALFA_European  \n",
       "1                   3283.0               21489.0               37269.0  \n",
       "2                     62.0                5355.0                8931.0  \n",
       "3                     31.0               10693.0                4955.0  \n",
       "4                     55.0                9154.0               11242.0  \n",
       "5                    113.0                 829.0               39989.0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_alfa_comparisons.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fisher's Exact Test\n",
    "Perform a two-tailed Fisher's Exact Test to determine if population differences in allele frequencies for variants are statistically significant "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate comparison population combinations\n",
    "\n",
    "comp_populations = [\n",
    "        (\"IH_Recent African\", \"ALFA_East Asian\"),\n",
    "        (\"IH_Recent African\", \"ALFA_South Asian\"),\n",
    "        (\"IH_Recent African\", \"ALFA_European\"),\n",
    "        (\"IH_Recent African\", \"ALFA_African Others\"),\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_East Asian</th>\n",
       "      <th>OR_IH_Recent African_ALFA_East Asian</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_South Asian</th>\n",
       "      <th>OR_IH_Recent African_ALFA_South Asian</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_European</th>\n",
       "      <th>OR_IH_Recent African_ALFA_European</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_African Others</th>\n",
       "      <th>OR_IH_Recent African_ALFA_African Others</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110174956_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110174956</td>\n",
       "      <td>5.200182e-02</td>\n",
       "      <td>0.705166</td>\n",
       "      <td>7.213253e-09</td>\n",
       "      <td>0.693264</td>\n",
       "      <td>3.934938e-18</td>\n",
       "      <td>0.617105</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.022030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110174971_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110174971</td>\n",
       "      <td>3.241684e-02</td>\n",
       "      <td>0.594670</td>\n",
       "      <td>2.502565e-02</td>\n",
       "      <td>0.606906</td>\n",
       "      <td>2.904355e-20</td>\n",
       "      <td>0.587723</td>\n",
       "      <td>0.825431</td>\n",
       "      <td>0.943514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>33026408_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>33026408</td>\n",
       "      <td>4.683343e-34</td>\n",
       "      <td>13.958507</td>\n",
       "      <td>1.829078e-05</td>\n",
       "      <td>2.778386</td>\n",
       "      <td>3.296705e-67</td>\n",
       "      <td>3.280968</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.956811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>150998920_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>150998920</td>\n",
       "      <td>2.382782e-01</td>\n",
       "      <td>1.271186</td>\n",
       "      <td>2.128952e-02</td>\n",
       "      <td>1.625936</td>\n",
       "      <td>1.174584e-17</td>\n",
       "      <td>1.561140</td>\n",
       "      <td>0.850243</td>\n",
       "      <td>1.043511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>151008522_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NOS3</td>\n",
       "      <td>151008522</td>\n",
       "      <td>5.660992e-03</td>\n",
       "      <td>1.990415</td>\n",
       "      <td>7.276266e-11</td>\n",
       "      <td>32.130990</td>\n",
       "      <td>1.482368e-222</td>\n",
       "      <td>13.716133</td>\n",
       "      <td>0.501688</td>\n",
       "      <td>1.184771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID REF ALT    GENE        POS  \\\n",
       "0  110174956_C_T   T   C  COL4A1  110174956   \n",
       "0  110174971_C_T   T   C  COL4A1  110174971   \n",
       "0   33026408_T_C   C   T   OLIG2   33026408   \n",
       "0  150998920_G_A   A   G    NOS3  150998920   \n",
       "0  151008522_C_T   T   C    NOS3  151008522   \n",
       "\n",
       "   PVALUE_IH_Recent African_ALFA_East Asian  \\\n",
       "0                              5.200182e-02   \n",
       "0                              3.241684e-02   \n",
       "0                              4.683343e-34   \n",
       "0                              2.382782e-01   \n",
       "0                              5.660992e-03   \n",
       "\n",
       "   OR_IH_Recent African_ALFA_East Asian  \\\n",
       "0                              0.705166   \n",
       "0                              0.594670   \n",
       "0                             13.958507   \n",
       "0                              1.271186   \n",
       "0                              1.990415   \n",
       "\n",
       "   PVALUE_IH_Recent African_ALFA_South Asian  \\\n",
       "0                               7.213253e-09   \n",
       "0                               2.502565e-02   \n",
       "0                               1.829078e-05   \n",
       "0                               2.128952e-02   \n",
       "0                               7.276266e-11   \n",
       "\n",
       "   OR_IH_Recent African_ALFA_South Asian  \\\n",
       "0                               0.693264   \n",
       "0                               0.606906   \n",
       "0                               2.778386   \n",
       "0                               1.625936   \n",
       "0                              32.130990   \n",
       "\n",
       "   PVALUE_IH_Recent African_ALFA_European  OR_IH_Recent African_ALFA_European  \\\n",
       "0                            3.934938e-18                            0.617105   \n",
       "0                            2.904355e-20                            0.587723   \n",
       "0                            3.296705e-67                            3.280968   \n",
       "0                            1.174584e-17                            1.561140   \n",
       "0                           1.482368e-222                           13.716133   \n",
       "\n",
       "   PVALUE_IH_Recent African_ALFA_African Others  \\\n",
       "0                                      1.000000   \n",
       "0                                      0.825431   \n",
       "0                                      1.000000   \n",
       "0                                      0.850243   \n",
       "0                                      0.501688   \n",
       "\n",
       "   OR_IH_Recent African_ALFA_African Others  \n",
       "0                                  1.022030  \n",
       "0                                  0.943514  \n",
       "0                                  0.956811  \n",
       "0                                  1.043511  \n",
       "0                                  1.184771  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate fisher's exact odds ratios and p-values for each variant for the different population combination\n",
    "\n",
    "fishers_results = functions.fishers_test(ih_alfa_comparisons, comp_populations)\n",
    "fishers_results.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Correct for multiple testing using the Bonferroni method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since multiple Fisher's Tests were performed, the results need to be corrected for multiple testing. Perform this correction using the Bonferroni method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>variable</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_African Others</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_East Asian</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_European</th>\n",
       "      <th>PVALUE_IH_Recent African_ALFA_South Asian</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.591452e-48</td>\n",
       "      <td>1.592468e-81</td>\n",
       "      <td>6.490057e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110152180_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.697572e-12</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110152715_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.004425e-171</td>\n",
       "      <td>4.529113e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110153931_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.671083e-04</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110154143_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.202546e-20</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "variable             ID REF ALT  PVALUE_IH_Recent African_ALFA_African Others  \\\n",
       "0         110149776_T_G   G   T                                           1.0   \n",
       "1         110152180_T_C   C   T                                           1.0   \n",
       "2         110152715_A_G   G   A                                           1.0   \n",
       "3         110153931_A_G   G   A                                           1.0   \n",
       "4         110154143_A_G   G   A                                           1.0   \n",
       "\n",
       "variable  PVALUE_IH_Recent African_ALFA_East Asian  \\\n",
       "0                                     1.591452e-48   \n",
       "1                                     1.000000e+00   \n",
       "2                                     1.000000e+00   \n",
       "3                                     1.000000e+00   \n",
       "4                                     1.202546e-20   \n",
       "\n",
       "variable  PVALUE_IH_Recent African_ALFA_European  \\\n",
       "0                                   1.592468e-81   \n",
       "1                                   2.697572e-12   \n",
       "2                                  4.004425e-171   \n",
       "3                                   2.671083e-04   \n",
       "4                                   1.000000e+00   \n",
       "\n",
       "variable  PVALUE_IH_Recent African_ALFA_South Asian  \n",
       "0                                      6.490057e-10  \n",
       "1                                      1.000000e+00  \n",
       "2                                      4.529113e-02  \n",
       "3                                      1.000000e+00  \n",
       "4                                      1.000000e+00  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate a list with column names for all subpopulation p-values in the fishers_results dataframe\n",
    "p_value_combinations_list = []\n",
    "for combination in comp_populations:\n",
    "    first_pop = combination[0]\n",
    "    second_pop = combination[1]\n",
    "    p_value_combinations = \"PVALUE_{}_{}\".format(first_pop, second_pop)\n",
    "    p_value_combinations_list.append(p_value_combinations)\n",
    "\n",
    "# Correct for multiple testing for the entire dataset\n",
    "multipletests_results = functions.multipletest_correction_wholedf(\n",
    "    fishers_results, p_value_combinations_list, 0.05, \"bonferroni\"\n",
    ")\n",
    "\n",
    "multipletests_results.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format Bonferroni-corrected Fisher's Test results\n",
    "Combine the Bonferroni-corrected p-values with the rest of the Fisher's Test results, including the odds ratios. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>COMP_POPS</th>\n",
       "      <th>CORR_PVALUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110152180_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110152715_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110153931_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110154143_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID REF ALT                              COMP_POPS  CORR_PVALUE\n",
       "0  110149776_T_G   G   T  IH_Recent African_ALFA_African Others          1.0\n",
       "1  110152180_T_C   C   T  IH_Recent African_ALFA_African Others          1.0\n",
       "2  110152715_A_G   G   A  IH_Recent African_ALFA_African Others          1.0\n",
       "3  110153931_A_G   G   A  IH_Recent African_ALFA_African Others          1.0\n",
       "4  110154143_A_G   G   A  IH_Recent African_ALFA_African Others          1.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Melt the dataframe containing the Bonferroni-corrected p-values so that this data can be combined with the the rest of the Fisher's results\n",
    "\n",
    "pvalue_table = multipletests_results.melt(id_vars=[\"ID\", \"REF\", \"ALT\"]).rename(\n",
    "    columns={\"variable\": \"COMP_POPS\", \"value\": \"CORR_PVALUE\"}\n",
    ")\n",
    "pvalue_table[\"COMP_POPS\"] = pvalue_table[\"COMP_POPS\"].apply(\n",
    "    lambda x: str(x).replace(\"PVALUE_\", \"\")\n",
    ")\n",
    "pvalue_table.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>COMP_POPS</th>\n",
       "      <th>OR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110174956_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_Recent African_ALFA_East Asian</td>\n",
       "      <td>0.705166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110174971_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_Recent African_ALFA_East Asian</td>\n",
       "      <td>0.594670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33026408_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_Recent African_ALFA_East Asian</td>\n",
       "      <td>13.958507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>150998920_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>IH_Recent African_ALFA_East Asian</td>\n",
       "      <td>1.271186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>151008522_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_Recent African_ALFA_East Asian</td>\n",
       "      <td>1.990415</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID REF ALT                          COMP_POPS         OR\n",
       "0  110174956_C_T   T   C  IH_Recent African_ALFA_East Asian   0.705166\n",
       "1  110174971_C_T   T   C  IH_Recent African_ALFA_East Asian   0.594670\n",
       "2   33026408_T_C   C   T  IH_Recent African_ALFA_East Asian  13.958507\n",
       "3  150998920_G_A   A   G  IH_Recent African_ALFA_East Asian   1.271186\n",
       "4  151008522_C_T   T   C  IH_Recent African_ALFA_East Asian   1.990415"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the dataframe containing the remaining Fisher's results, such as the odds ratios, into a suitable format for combination with the corrected p-values.\n",
    "\n",
    "fishers_df_OR_columns = fishers_results.columns.str.contains(\"OR_*|ID|REF|ALT\")\n",
    "or_table = (\n",
    "    fishers_results.iloc[:, fishers_df_OR_columns]\n",
    "    .melt(id_vars=[\"ID\", \"REF\", \"ALT\"])\n",
    "    .rename(columns={\"variable\": \"COMP_POPS\", \"value\": \"OR\"})\n",
    ")\n",
    "or_table[\"COMP_POPS\"] = or_table[\"COMP_POPS\"].apply(lambda x: str(x).replace(\"OR_\", \"\"))\n",
    "or_table.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>COMP_POPS</th>\n",
       "      <th>CORR_PVALUE</th>\n",
       "      <th>OR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.740627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110152180_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.281796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110152715_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.799468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110153931_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.064030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110154143_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.289451</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID REF ALT                              COMP_POPS  CORR_PVALUE  \\\n",
       "0  110149776_T_G   G   T  IH_Recent African_ALFA_African Others          1.0   \n",
       "1  110152180_T_C   C   T  IH_Recent African_ALFA_African Others          1.0   \n",
       "2  110152715_A_G   G   A  IH_Recent African_ALFA_African Others          1.0   \n",
       "3  110153931_A_G   G   A  IH_Recent African_ALFA_African Others          1.0   \n",
       "4  110154143_A_G   G   A  IH_Recent African_ALFA_African Others          1.0   \n",
       "\n",
       "         OR  \n",
       "0  1.740627  \n",
       "1  0.281796  \n",
       "2  1.799468  \n",
       "3  1.064030  \n",
       "4  1.289451  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine the corrected p-value data with the remaining Fisher's test data\n",
    "\n",
    "pvalue_or_table = pd.merge(pvalue_table, or_table, on=[\"ID\", \"REF\", \"ALT\", \"COMP_POPS\"])\n",
    "pvalue_or_table.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format the data so that is it useable for further analysis. Start by adding in the variant allele count information that was used to conduct the Fisher's Exact Tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REG</th>\n",
       "      <th>ALT_CTS</th>\n",
       "      <th>REF_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34452510_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_ACB</td>\n",
       "      <td>24.0</td>\n",
       "      <td>204.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110174956_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_ACB</td>\n",
       "      <td>56.0</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110174971_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_ACB</td>\n",
       "      <td>56.0</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33026408_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>IH_ACB</td>\n",
       "      <td>204.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>150998920_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>IH_ACB</td>\n",
       "      <td>145.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID REF ALT     REG  ALT_CTS  REF_CTS\n",
       "0   34452510_T_C   C   T  IH_ACB     24.0    204.0\n",
       "1  110174956_C_T   T   C  IH_ACB     56.0    172.0\n",
       "2  110174971_C_T   T   C  IH_ACB     56.0    172.0\n",
       "3   33026408_T_C   C   T  IH_ACB    204.0     24.0\n",
       "4  150998920_G_A   A   G  IH_ACB    145.0     83.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Temporarily rename the REF and ALT columns\n",
    "ih_alfa_data = ih_alfa_data.rename(columns={\"REF\": \"REF_AL\", \"ALT\": \"ALT_AL\"})\n",
    "\n",
    "# Subset ALT count data and format\n",
    "ih_ALFA_data_ALT_columns = ih_alfa_data.columns.str.contains(\n",
    "    \"ALT_CT_*|ID|REF_AL|ALT_AL\"\n",
    ")\n",
    "alt_count_table = ih_alfa_data.iloc[:, ih_ALFA_data_ALT_columns].melt(\n",
    "    id_vars=[\"ID\", \"REF_AL\", \"ALT_AL\"], value_name=\"ALT_CTS\", var_name=\"REG\"\n",
    ")\n",
    "\n",
    "# Remove unnecessary info from the comparison column\n",
    "alt_count_table[\"REG\"] = alt_count_table[\"REG\"].apply(\n",
    "    lambda x: str(x).replace(\"ALT_CT_\", \"\")\n",
    ")\n",
    "\n",
    "# Subset REF count data and format\n",
    "ih_ALFA_data_REF_columns = ih_alfa_data.columns.str.contains(\n",
    "    \"REF_CT_*|ID|REF_AL|ALT_AL\"\n",
    ")\n",
    "ref_count_table = ih_alfa_data.iloc[:, ih_ALFA_data_REF_columns].melt(\n",
    "    id_vars=[\"ID\", \"REF_AL\", \"ALT_AL\"], value_name=\"REF_CTS\", var_name=\"REG\"\n",
    ")\n",
    "\n",
    "# Remove unnecessary info from the comparison column\n",
    "ref_count_table[\"REG\"] = alt_count_table[\"REG\"].apply(\n",
    "    lambda x: str(x).replace(\"REF_CT_\", \"\")\n",
    ")\n",
    "\n",
    "# Combine formatted alt and ref count data\n",
    "count_table = pd.merge(\n",
    "    alt_count_table, ref_count_table, on=[\"ID\", \"REF_AL\", \"ALT_AL\", \"REG\"]\n",
    ").rename(columns={\"REF_AL\": \"REF\", \"ALT_AL\": \"ALT\"})\n",
    "count_table.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Refine the dataset by retaining only the rows with significant p-values. Then, merge the significant p-value, odds ratio, and allele count data. Finally, compute the variant alternate allele frequencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>COMP_POPS</th>\n",
       "      <th>CORR_PVALUE</th>\n",
       "      <th>OR</th>\n",
       "      <th>POP1_REF_CTS</th>\n",
       "      <th>POP1_ALT_CTS</th>\n",
       "      <th>POP2_REF_CTS</th>\n",
       "      <th>POP2_ALT_CTS</th>\n",
       "      <th>POP1_ALT_FREQ</th>\n",
       "      <th>POP2_ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>110158027_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.035053e-03</td>\n",
       "      <td>24.975000</td>\n",
       "      <td>960.0</td>\n",
       "      <td>648.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.402985</td>\n",
       "      <td>0.026316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>110159006_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>1.275537e-45</td>\n",
       "      <td>0.001015</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.000622</td>\n",
       "      <td>0.380000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>110170107_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>3.579765e-28</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.000622</td>\n",
       "      <td>0.240741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>110176544_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>4.786871e-02</td>\n",
       "      <td>3.688627</td>\n",
       "      <td>1187.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.261816</td>\n",
       "      <td>0.087719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>110188529_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IH_Recent African_ALFA_African Others</td>\n",
       "      <td>2.801245e-65</td>\n",
       "      <td>0.001101</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.000622</td>\n",
       "      <td>0.361111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID REF ALT                              COMP_POPS  \\\n",
       "22   110158027_G_A   A   G  IH_Recent African_ALFA_African Others   \n",
       "31   110159006_A_G   G   A  IH_Recent African_ALFA_African Others   \n",
       "69   110170107_A_G   G   A  IH_Recent African_ALFA_African Others   \n",
       "84   110176544_C_T   T   C  IH_Recent African_ALFA_African Others   \n",
       "126  110188529_C_T   T   C  IH_Recent African_ALFA_African Others   \n",
       "\n",
       "      CORR_PVALUE         OR  POP1_REF_CTS  POP1_ALT_CTS  POP2_REF_CTS  \\\n",
       "22   1.035053e-03  24.975000         960.0         648.0          37.0   \n",
       "31   1.275537e-45   0.001015        1607.0           1.0          62.0   \n",
       "69   3.579765e-28   0.001963        1607.0           1.0          82.0   \n",
       "84   4.786871e-02   3.688627        1187.0         421.0         104.0   \n",
       "126  2.801245e-65   0.001101        1607.0           1.0         115.0   \n",
       "\n",
       "     POP2_ALT_CTS  POP1_ALT_FREQ  POP2_ALT_FREQ  \n",
       "22            1.0       0.402985       0.026316  \n",
       "31           38.0       0.000622       0.380000  \n",
       "69           26.0       0.000622       0.240741  \n",
       "84           10.0       0.261816       0.087719  \n",
       "126          65.0       0.000622       0.361111  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter p-value and odds ratio table by significant p-values\n",
    "\n",
    "sign_pvalue_or_table = pvalue_or_table[pvalue_or_table.CORR_PVALUE < 0.05]\n",
    "\n",
    "# Add empty count columns to p-value and odds ratio table\n",
    "sign_pvalue_or_table.loc[:,\"POP1_REF_CTS\"] = \"\"\n",
    "sign_pvalue_or_table.loc[:,\"POP1_ALT_CTS\"] = \"\"\n",
    "sign_pvalue_or_table.loc[:,\"POP2_REF_CTS\"] = \"\"\n",
    "sign_pvalue_or_table.loc[:,\"POP2_ALT_CTS\"] = \"\"\n",
    "\n",
    "# Iterate over row in p-value and odds ratio table and append count data\n",
    "for index, row in sign_pvalue_or_table.iterrows():\n",
    "    # Extract variant ID, ref allele and alt allele information\n",
    "    id = row[\"ID\"]\n",
    "    ref = row[\"REF\"]\n",
    "    alt = row[\"ALT\"]\n",
    "\n",
    "    # Split comparison population column into the two comparison populations\n",
    "    comp_pops = row[\"COMP_POPS\"]\n",
    "    split_pops = comp_pops.split(\"_\", 2)\n",
    "    split_pops = [\"_\".join(split_pops[0:2])] + split_pops[2:]\n",
    "    comp_pop1 = split_pops[0]\n",
    "    comp_pop2 = split_pops[1]\n",
    "\n",
    "    # Fetch row in count table with information for each comparison population\n",
    "    pop1_count_table_row = count_table.loc[\n",
    "        (count_table.ID == id)\n",
    "        & (count_table.REF == ref)\n",
    "        & (count_table.ALT == alt)\n",
    "        & (count_table.REG == comp_pop1)\n",
    "    ]\n",
    "    pop2_count_table_row = count_table.loc[\n",
    "        (count_table.ID == id)\n",
    "        & (count_table.REF == ref)\n",
    "        & (count_table.ALT == alt)\n",
    "        & (count_table.REG == comp_pop2)\n",
    "    ]\n",
    "\n",
    "    # Extract ref and alt count information for relevant comparison population\n",
    "    pop1_ref = pop1_count_table_row[\"REF_CTS\"].values\n",
    "    pop1_alt = pop1_count_table_row[\"ALT_CTS\"].values\n",
    "    pop2_ref = pop2_count_table_row[\"REF_CTS\"].values\n",
    "    pop2_alt = pop2_count_table_row[\"ALT_CTS\"].values\n",
    "\n",
    "    # Append ref and alt count information to p-value and odds ratio table\n",
    "    sign_pvalue_or_table.at[index, \"POP1_REF_CTS\"] = pop1_ref\n",
    "    sign_pvalue_or_table.at[index, \"POP1_ALT_CTS\"] = pop1_alt\n",
    "    sign_pvalue_or_table.at[index, \"POP2_REF_CTS\"] = pop2_ref\n",
    "    sign_pvalue_or_table.at[index, \"POP2_ALT_CTS\"] = pop2_alt\n",
    "\n",
    "\n",
    "# # Change count data format from list to float\n",
    "sign_pvalue_or_table[\"POP1_REF_CTS\"] = sign_pvalue_or_table[\"POP1_REF_CTS\"].str[0]\n",
    "sign_pvalue_or_table[\"POP1_ALT_CTS\"] = sign_pvalue_or_table[\"POP1_ALT_CTS\"].str[0]\n",
    "sign_pvalue_or_table[\"POP2_REF_CTS\"] = sign_pvalue_or_table[\"POP2_REF_CTS\"].str[0]\n",
    "sign_pvalue_or_table[\"POP2_ALT_CTS\"] = sign_pvalue_or_table[\"POP2_ALT_CTS\"].str[0]\n",
    "\n",
    "# Calculate alt allele frequencies and append results to table\n",
    "sign_pvalue_or_table[\"POP1_ALT_FREQ\"] = sign_pvalue_or_table[\"POP1_ALT_CTS\"].astype(\n",
    "    float\n",
    ") / (\n",
    "    sign_pvalue_or_table[\"POP1_ALT_CTS\"].astype(float)\n",
    "    + sign_pvalue_or_table[\"POP1_REF_CTS\"].astype(float)\n",
    ")\n",
    "sign_pvalue_or_table[\"POP2_ALT_FREQ\"] = sign_pvalue_or_table[\"POP2_ALT_CTS\"].astype(\n",
    "    float\n",
    ") / (\n",
    "    sign_pvalue_or_table[\"POP2_ALT_CTS\"].astype(float)\n",
    "    + sign_pvalue_or_table[\"POP2_REF_CTS\"].astype(float)\n",
    ")\n",
    "\n",
    "sign_pvalue_or_table.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the formatted Bonferroni-corrected Fisher's Test results\n",
    "\n",
    "sign_pvalue_or_table.to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Results\",\n",
    "        \"Tables\",\n",
    "        \"Significant_Fishers_results_bonferroni_corrected.csv\",\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "23b4a3e8622309bcc6db3d5cc6eb73d60ab98d9ec23bad6a26b709981ccb403a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
