{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieval of variant phenotype data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genetic variant phenotype association data was retrieved using [the Functional Annotation of Variants - Online Resource v2.0 (FAVOR)](https://favor.genohub.org/batch-annotation) for the variants identified in-house in African populations. To retrieve the data via the FAVOR database user interface, the variant input data needed to be formatted in a specific way. This notebook details the preparation of the input data. Since the phenotype data was retrieved using a user interface, the retrieval process itself is not documented here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Loading"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compile a list of variants analysed in African populations for which we want to obtain phenotype data, we first need to load the in-house African variant data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Bantu Kenya</td>\n",
       "      <td>EA</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>WA</td>\n",
       "      <td>276</td>\n",
       "      <td>0</td>\n",
       "      <td>276</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>San</td>\n",
       "      <td>SA</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mende</td>\n",
       "      <td>WA</td>\n",
       "      <td>166</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110148882_C_CT</td>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mbuti Pygmy</td>\n",
       "      <td>CA</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID             VAR_NAME VARIANT_TYPE        POS REF ALT  \\\n",
       "0  110148882_C_CT  chr13:110148882C-CT        INDEL  110148882   C  CT   \n",
       "1  110148882_C_CT  chr13:110148882C-CT        INDEL  110148882   C  CT   \n",
       "2  110148882_C_CT  chr13:110148882C-CT        INDEL  110148882   C  CT   \n",
       "3  110148882_C_CT  chr13:110148882C-CT        INDEL  110148882   C  CT   \n",
       "4  110148882_C_CT  chr13:110148882C-CT        INDEL  110148882   C  CT   \n",
       "\n",
       "     GENE      SUB_POP REG  IH_REF_CTS  IH_ALT_CTS  IH_TOTAL_CTS  IH_AF  \n",
       "0  COL4A1  Bantu Kenya  EA          20           0            20    0.0  \n",
       "1  COL4A1       Yoruba  WA         276           0           276    0.0  \n",
       "2  COL4A1          San  SA          12           0            12    0.0  \n",
       "3  COL4A1        Mende  WA         166           0           166    0.0  \n",
       "4  COL4A1  Mbuti Pygmy  CA          24           0            24    0.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_afr = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "ih_afr.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each of the variants listed in the DataFrame above resides in a particular gene. Load information on the genomic coordinates of each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COL4A1</td>\n",
       "      <td>13</td>\n",
       "      <td>110148863</td>\n",
       "      <td>110307257</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NOS3</td>\n",
       "      <td>7</td>\n",
       "      <td>150990917</td>\n",
       "      <td>151014688</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL6</td>\n",
       "      <td>7</td>\n",
       "      <td>22725784</td>\n",
       "      <td>22732102</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL1B</td>\n",
       "      <td>2</td>\n",
       "      <td>112829651</td>\n",
       "      <td>112836943</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGT</td>\n",
       "      <td>1</td>\n",
       "      <td>230690676</td>\n",
       "      <td>230745683</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE  chromosome      start       stop  strand\n",
       "0  COL4A1          13  110148863  110307257      -1\n",
       "1    NOS3           7  150990917  151014688       1\n",
       "2     IL6           7   22725784   22732102       1\n",
       "3    IL1B           2  112829651  112836943      -1\n",
       "4     AGT           1  230690676  230745683      -1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coords = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Metadata\",\n",
    "        \"gene_locations.csv\",\n",
    "    )\n",
    ")\n",
    "gene_coords = gene_coords.rename(columns={\"location_name\":\"GENE\"})\n",
    "gene_coords.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compile data into required input format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select aggregated variant data on Recent Africans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_afr = ih_afr[ih_afr.REG == \"Recent African\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove variants that have an alternate allele count of 0 in Recent Africans. These variants were not found in Recent Africans."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>110148891_C_G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>110148917_C_G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1101</td>\n",
       "      <td>119</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.097541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>110149176_T_A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149176</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>110149349_G_A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1215</td>\n",
       "      <td>5</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.004098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>110149494_C_T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID     VAR_NAME VARIANT_TYPE        POS REF ALT    GENE  \\\n",
       "12   110148891_C_G  rs552586867          SNP  110148891   C   G  COL4A1   \n",
       "28   110148917_C_G   rs59409892          SNP  110148917   C   G  COL4A1   \n",
       "144  110149176_T_A  rs546124548          SNP  110149176   T   A  COL4A1   \n",
       "252  110149349_G_A  rs139916479          SNP  110149349   G   A  COL4A1   \n",
       "288  110149494_C_T  rs552877576          SNP  110149494   C   T  COL4A1   \n",
       "\n",
       "    SUB_POP             REG  IH_REF_CTS  IH_ALT_CTS  IH_TOTAL_CTS     IH_AF  \n",
       "12      NaN  Recent African        1219           1          1220  0.000820  \n",
       "28      NaN  Recent African        1101         119          1220  0.097541  \n",
       "144     NaN  Recent African        1219           1          1220  0.000820  \n",
       "252     NaN  Recent African        1215           5          1220  0.004098  \n",
       "288     NaN  Recent African        1219           1          1220  0.000820  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_afr_filtered = ih_afr[ih_afr.IH_ALT_CTS != 0]\n",
    "\n",
    "ih_afr_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join the chromosome information from the `gene_coords` DataFrame with the variant information in the `ih_afr_filtered` DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>REG</th>\n",
       "      <th>IH_REF_CTS</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>IH_AF</th>\n",
       "      <th>CHROM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>110148891_C_G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110148917_C_G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1101</td>\n",
       "      <td>119</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.097541</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110149176_T_A</td>\n",
       "      <td>rs546124548</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149176</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>110149349_G_A</td>\n",
       "      <td>rs139916479</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1215</td>\n",
       "      <td>5</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>110149494_C_T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>SNP</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>1219</td>\n",
       "      <td>1</td>\n",
       "      <td>1220</td>\n",
       "      <td>0.000820</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID     VAR_NAME VARIANT_TYPE        POS REF ALT    GENE SUB_POP  \\\n",
       "0  110148891_C_G  rs552586867          SNP  110148891   C   G  COL4A1     NaN   \n",
       "1  110148917_C_G   rs59409892          SNP  110148917   C   G  COL4A1     NaN   \n",
       "2  110149176_T_A  rs546124548          SNP  110149176   T   A  COL4A1     NaN   \n",
       "3  110149349_G_A  rs139916479          SNP  110149349   G   A  COL4A1     NaN   \n",
       "4  110149494_C_T  rs552877576          SNP  110149494   C   T  COL4A1     NaN   \n",
       "\n",
       "              REG  IH_REF_CTS  IH_ALT_CTS  IH_TOTAL_CTS     IH_AF  CHROM  \n",
       "0  Recent African        1219           1          1220  0.000820     13  \n",
       "1  Recent African        1101         119          1220  0.097541     13  \n",
       "2  Recent African        1219           1          1220  0.000820     13  \n",
       "3  Recent African        1215           5          1220  0.004098     13  \n",
       "4  Recent African        1219           1          1220  0.000820     13  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_afr_filtered = pd.merge(ih_afr_filtered, gene_coords[[\"GENE\",\"chromosome\"]], on=\"GENE\").rename(columns={\"chromosome\":\"CHROM\"})\n",
    "\n",
    "ih_afr_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select variant chromosome, position, reference allele and alternate allele information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110149176</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT\n",
       "0     13  110148891   C   G\n",
       "1     13  110148917   C   G\n",
       "2     13  110149176   T   A\n",
       "3     13  110149349   G   A\n",
       "4     13  110149494   C   T"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_variant_info = ih_afr_filtered[[\"CHROM\", \"POS\", \"REF\", \"ALT\"]]\n",
    "selected_variant_info.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format variant chromosome, position, reference allele and alternate allele information correctly to allow retrieval of variant annotation info from https://favor.genohub.org/batch-annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "formatted_variant_info = pd.DataFrame(data=selected_variant_info[\"CHROM\"].astype(str) + \"-\" + selected_variant_info[\"POS\"].astype(str) + \"-\" + selected_variant_info[\"REF\"] + \"-\" + selected_variant_info[\"ALT\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save formatted data to txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "formatted_variant_info.to_csv('Variant_phenotype_formatted_variants.txt', header=False, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
