{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of ALFA global variant count data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare the African variant count data generated in-house from the [GnomAD 1000 Genomes and HGDP datasets](https://gnomad.broadinstitute.org/news/2021-10-gnomad-v3-1-2-minor-release/), with data from other global populations, global population variant data on European, Asian and Latin American populations was retrieved from the [NCBI ALFA database](https://www.ncbi.nlm.nih.gov/snp/docs/gsr/alfa/). \n",
    "\n",
    "The data retrieval process from the NCBI ALFA database is documented in the `Notebooks/1-Data_retrieval/ALFA_allele_counts.ipynb` notebook. The raw data obtained is stored in the `Data/Raw/ALFA/ALFA_allele_counts.csv` file. To prepare this data for further analysis, it must be organized and formatted appropriately. This section outlines the steps taken to achieve this."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import pandas as pd\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data loading"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the raw data from the`Data/Raw/ALFA/ALFA_allele_counts.csv` file into a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_code</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>population_code</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>allele_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492705</td>\n",
       "      <td>C</td>\n",
       "      <td>{'G': 0, 'C': 14050}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492695</td>\n",
       "      <td>C</td>\n",
       "      <td>{'G': 0, 'C': 9690}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_code   variant_id population_code reference_allele  \\\n",
       "0  PRJNA507278  rs552586867    SAMN10492705                C   \n",
       "1  PRJNA507278  rs552586867    SAMN10492695                C   \n",
       "\n",
       "          allele_counts  \n",
       "0  {'G': 0, 'C': 14050}  \n",
       "1   {'G': 0, 'C': 9690}  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_alfa_allele_counts = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"ALFA\",\n",
    "        \"ALFA_allele_count_data.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "all_alfa_allele_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The information stored in the DataFrame columns is as follows:\n",
    "\n",
    "* `study_code`: A unique identifier for the study or project from which the data was derived. It helps in tracking the source of the data.\n",
    "* `variant_id`: A unique identifier for the genetic variant. It typically refers to a specific location in the genome where a variation (such as a single nucleotide polymorphism, SNP) has been observed.\n",
    "* `population_code`: A code representing the population or subpopulation from which the data was collected. This helps in understanding the genetic diversity and allele frequencies across different groups.\n",
    "* `reference_allele`: The allele that is considered the reference or standard for the variant. It is usually the most common allele observed at that genomic location in a reference population.\n",
    "* `allele_counts`: The counts of each allele observed in the population. It represents the frequency of the reference allele and possibly other alternative alleles in the given population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unpack column values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `allele_counts` column contains information stored in a dictionary format. To make the data more accessible for analysis, expand the dictionary content into multiple separate columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_code</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>population_code</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>AAAAAAAAAAA</th>\n",
       "      <th>AAAAAAAAAAAAA</th>\n",
       "      <th>...</th>\n",
       "      <th>TATATATATA</th>\n",
       "      <th>TATATATATATA</th>\n",
       "      <th>ATACACACACACACA</th>\n",
       "      <th>ATACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACACACACACA</th>\n",
       "      <th>ATCTATACACACA</th>\n",
       "      <th>TAAAATA</th>\n",
       "      <th>TAAAATAAAATA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492705</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14050.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492695</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9690.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 562 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_code   variant_id population_code reference_allele    G        C  \\\n",
       "0  PRJNA507278  rs552586867    SAMN10492705                C  0.0  14050.0   \n",
       "1  PRJNA507278  rs552586867    SAMN10492695                C  0.0   9690.0   \n",
       "\n",
       "    A   T  AAAAAAAAAAA  AAAAAAAAAAAAA  ...  TATATATATA  TATATATATATA  \\\n",
       "0 NaN NaN          NaN            NaN  ...         NaN           NaN   \n",
       "1 NaN NaN          NaN            NaN  ...         NaN           NaN   \n",
       "\n",
       "   ATACACACACACACA  ATACACACACACACACA  ATATATACACACACACACACA  \\\n",
       "0              NaN                NaN                    NaN   \n",
       "1              NaN                NaN                    NaN   \n",
       "\n",
       "   ATATATACACACACACACACACA  ATATATACACACACACACACACACACACA  ATCTATACACACA  \\\n",
       "0                      NaN                            NaN            NaN   \n",
       "1                      NaN                            NaN            NaN   \n",
       "\n",
       "   TAAAATA  TAAAATAAAATA  \n",
       "0      NaN           NaN  \n",
       "1      NaN           NaN  \n",
       "\n",
       "[2 rows x 562 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Unpack dictionary values into multiple separate columns\n",
    "normalised_count_columns = (\n",
    "    all_alfa_allele_counts[\"allele_counts\"].map(eval).apply(pd.Series)\n",
    ")\n",
    "\n",
    "# Drop the old allele count column from the dataframe\n",
    "drop_count_column = all_alfa_allele_counts.drop(columns=\"allele_counts\")\n",
    "\n",
    "# Re-add the new, unpacked allele count column\n",
    "normalised_alfa_allele_counts = pd.concat(\n",
    "    [\n",
    "        drop_count_column.reset_index(drop=True),\n",
    "        normalised_count_columns.reset_index(drop=True),\n",
    "    ],\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "normalised_alfa_allele_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translate coded column values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The names of populations have been represented by codes in the `population_code` column. To make the data easier to understand, translate the coded population names into legible and recognisable population names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_code</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>population_code</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>AAAAAAAAAAA</th>\n",
       "      <th>AAAAAAAAAAAAA</th>\n",
       "      <th>...</th>\n",
       "      <th>TATATATATATA</th>\n",
       "      <th>ATACACACACACACA</th>\n",
       "      <th>ATACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACACA</th>\n",
       "      <th>ATATATACACACACACACACACACACACA</th>\n",
       "      <th>ATCTATACACACA</th>\n",
       "      <th>TAAAATA</th>\n",
       "      <th>TAAAATAAAATA</th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492705</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14050.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>SAMN10492695</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9690.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 563 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_code   variant_id population_code reference_allele    G        C  \\\n",
       "0  PRJNA507278  rs552586867    SAMN10492705                C  0.0  14050.0   \n",
       "1  PRJNA507278  rs552586867    SAMN10492695                C  0.0   9690.0   \n",
       "\n",
       "    A   T  AAAAAAAAAAA  AAAAAAAAAAAAA  ...  TATATATATATA  ATACACACACACACA  \\\n",
       "0 NaN NaN          NaN            NaN  ...           NaN              NaN   \n",
       "1 NaN NaN          NaN            NaN  ...           NaN              NaN   \n",
       "\n",
       "   ATACACACACACACACA  ATATATACACACACACACACA  ATATATACACACACACACACACA  \\\n",
       "0                NaN                    NaN                      NaN   \n",
       "1                NaN                    NaN                      NaN   \n",
       "\n",
       "   ATATATACACACACACACACACACACACA  ATCTATACACACA  TAAAATA  TAAAATAAAATA  \\\n",
       "0                            NaN            NaN      NaN           NaN   \n",
       "1                            NaN            NaN      NaN           NaN   \n",
       "\n",
       "   Population  \n",
       "0       Total  \n",
       "1    European  \n",
       "\n",
       "[2 rows x 563 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fetch ALFA study population metadata\n",
    "metadata_json = functions.get_metadata()\n",
    "\n",
    "# Generate a list of all unique study and population codes that need to be queried\n",
    "study_codes = [\n",
    "    study_code for study_code in normalised_alfa_allele_counts.study_code.unique()\n",
    "]\n",
    "population_codes = [\n",
    "    population_code\n",
    "    for population_code in normalised_alfa_allele_counts.population_code.unique()\n",
    "]\n",
    "\n",
    "# Fetch study and population name metadata. \n",
    "# Code modified from https://github.com/ncbi/dbsnp/blob/master/tutorials/Variation%20Services/Jupyter_Notebook/by_rsid.ipynb.\n",
    "metadata = {}\n",
    "for project_json in metadata_json:\n",
    "    p = {}\n",
    "    p[\"json\"] = project_json\n",
    "    p[\"pops\"] = {}\n",
    "    metadata[project_json[\"bioproject_id\"]] = p\n",
    "for prj_id, prj in metadata.items():\n",
    "    functions.add_all_pops(prj[\"json\"][\"populations\"], prj)\n",
    "\n",
    "for study_id in study_codes:\n",
    "    study_name = metadata[study_id][\"json\"][\"short_name\"]\n",
    "\n",
    "    pop_dict = {\n",
    "        \"SAMN10492696\": \"African Others\",\n",
    "        \"SAMN10492698\": \"African American\",\n",
    "        \"SAMN10492697\": \"East Asian\",\n",
    "        \"SAMN10492701\": \"Other Asian\",\n",
    "    }\n",
    "    for pop_id in population_codes:\n",
    "        if pop_id not in [\n",
    "            \"SAMN10492696\",\n",
    "            \"SAMN10492698\",\n",
    "            \"SAMN10492697\",\n",
    "            \"SAMN10492701\",\n",
    "        ]:\n",
    "            population_name = metadata[study_id][\"pops\"][pop_id][\"name\"]\n",
    "            temp_pop_dict = {pop_id: population_name}\n",
    "        pop_dict.update(temp_pop_dict)\n",
    "\n",
    "# Add column in population allele count dataframe for population names\n",
    "normalised_alfa_allele_counts[\"Population\"] = normalised_alfa_allele_counts[\n",
    "    \"population_code\"\n",
    "].map(pop_dict)\n",
    "\n",
    "normalised_alfa_allele_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the data to include only the specific populations of interest. Retain information on the European, East Asian, South Asian, Latin American, and Latin American 2 population groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a list of included population groups\n",
    "incl_pops = [\n",
    "    \"European\",\n",
    "    \"East Asian\",\n",
    "    \"South Asian\",\n",
    "    \"Latin American 1\",\n",
    "    \"Latin American 2\",\n",
    "]\n",
    "\n",
    "# Filter data by included population groups\n",
    "filtered_alfa_allele_counts = normalised_alfa_allele_counts[\n",
    "    normalised_alfa_allele_counts[\"Population\"].isin(incl_pops)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reshape data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To enable a meaningful comparison with our in-house African allele count data, reformat the ALFA allele count data to match the structure used for the in-house dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Melt the DataFrame to reshape it for easier analysis\n",
    "melted_alfa_allele_counts = filtered_alfa_allele_counts.melt(\n",
    "    id_vars=[\"study_code\", \"variant_id\", \"population_code\", \"reference_allele\", \"Population\"],\n",
    "    value_vars=filtered_alfa_allele_counts.iloc[:, 4:].columns,\n",
    "    var_name=\"allele\",\n",
    "    value_name=\"count\",\n",
    ").dropna(subset=[\"count\"])\n",
    "\n",
    "# Separate reference allele and allele count information\n",
    "ref_allele_info = melted_alfa_allele_counts[[\"study_code\", \"variant_id\", \"population_code\", \"Population\", \"reference_allele\"]]\n",
    "allele_count_info = melted_alfa_allele_counts[[\"study_code\", \"variant_id\", \"population_code\", \"Population\", \"allele\", \"count\"]]\n",
    "\n",
    "# Merge allele count information with reference allele information\n",
    "ref_alfa_counts = ref_allele_info.merge(\n",
    "    allele_count_info,\n",
    "    left_on=[\"study_code\", \"variant_id\", \"population_code\", \"Population\", \"reference_allele\"],\n",
    "    right_on=[\"study_code\", \"variant_id\", \"population_code\", \"Population\", \"allele\"]\n",
    ").rename(columns={\"count\": \"alfa_ref_cts\"}).drop(columns=\"allele\").drop_duplicates(subset=[\"study_code\",\"variant_id\",\"population_code\",\"Population\",\"reference_allele\"])\n",
    "\n",
    "# Merge remaining allele count information for alternate alleles\n",
    "alt_alfa_counts = melted_alfa_allele_counts.merge(\n",
    "    ref_alfa_counts,\n",
    "    on=[\"study_code\", \"variant_id\", \"population_code\", \"reference_allele\", \"Population\"],\n",
    ")\n",
    "\n",
    "# Filter alternate allele counts\n",
    "alfa_counts = alt_alfa_counts[alt_alfa_counts[\"reference_allele\"] != alt_alfa_counts[\"allele\"]]\n",
    "\n",
    "# Rename columns\n",
    "alfa_counts = alfa_counts.rename(columns={\"study_code\":\"STUDY_CODE\",\"population_code\":\"POP_ID\",\"Population\":\"POP\",\"variant_id\":\"VAR_NAME\", \"allele\": \"ALT\", \"count\": \"ALT_CTS\", \"reference_allele\":\"REF\", \"alfa_ref_cts\":\"REF_CTS\"})\n",
    "\n",
    "# Calculate alternate allele frequencies\n",
    "alfa_counts[\"AF\"] = alfa_counts[\"ALT_CTS\"] / (alfa_counts[\"REF_CTS\"] + alfa_counts[\"ALT_CTS\"])\n",
    "\n",
    "# Pivot the data for analysis\n",
    "alfa_pivot_data = alfa_counts.pivot_table(\n",
    "    index=[\"VAR_NAME\", \"REF\", \"ALT\"],\n",
    "    columns=\"POP\",\n",
    "    values=[\"ALT_CTS\", \"REF_CTS\"]\n",
    ").reset_index()\n",
    "\n",
    "# Separate data into alternate and reference count dataframes for renaming\n",
    "alfa_data_alt = alfa_pivot_data[[\"VAR_NAME\", \"REF\", \"ALT\", \"ALT_CTS\"]].droplevel(level=0, axis=1).reset_index(drop=True)\n",
    "alfa_data_alt.columns.values[0:3] = [\"VAR_NAME\", \"REF\", \"ALT\"]\n",
    "alfa_data_ref = alfa_pivot_data[[\"VAR_NAME\", \"REF\", \"ALT\", \"REF_CTS\"]].droplevel(level=0, axis=1).reset_index(drop=True)\n",
    "alfa_data_ref.columns.values[0:3] = [\"VAR_NAME\", \"REF\", \"ALT\"]\n",
    "\n",
    "# Add appropriate prefixes to alt and ref column names\n",
    "alfa_data_alt = functions.add_prefix_dataframe_col_names(\n",
    "    alfa_data_alt, alfa_data_alt.iloc[:, 3:], \"ALT_CT_ALFA_\"\n",
    ")\n",
    "\n",
    "alfa_data_ref = functions.add_prefix_dataframe_col_names(\n",
    "    alfa_data_ref, alfa_data_ref.iloc[:, 3:], \"REF_CT_ALFA_\"\n",
    ")\n",
    "\n",
    "# Merge renamed alternate and reference count data\n",
    "alfa_grouped_data = alfa_data_alt.merge(\n",
    "    alfa_data_ref, on=[\"VAR_NAME\", \"REF\", \"ALT\"]\n",
    ")\n",
    "\n",
    "# Remove index name\n",
    "alfa_grouped_data.columns.name = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display and save the prepared data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_CT_ALFA_East Asian</th>\n",
       "      <th>ALT_CT_ALFA_European</th>\n",
       "      <th>ALT_CT_ALFA_Latin American 1</th>\n",
       "      <th>ALT_CT_ALFA_Latin American 2</th>\n",
       "      <th>ALT_CT_ALFA_South Asian</th>\n",
       "      <th>REF_CT_ALFA_East Asian</th>\n",
       "      <th>REF_CT_ALFA_European</th>\n",
       "      <th>REF_CT_ALFA_Latin American 1</th>\n",
       "      <th>REF_CT_ALFA_Latin American 2</th>\n",
       "      <th>REF_CT_ALFA_South Asian</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000343</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>109377.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>2200.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000989</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>55.0</td>\n",
       "      <td>21489.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>1330.0</td>\n",
       "      <td>1685.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>37269.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>2052.0</td>\n",
       "      <td>3283.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    VAR_NAME REF ALT  ALT_CT_ALFA_East Asian  ALT_CT_ALFA_European  \\\n",
       "0  rs1000343   C   T                     0.0                  49.0   \n",
       "1  rs1000989   T   C                    55.0               21489.0   \n",
       "\n",
       "   ALT_CT_ALFA_Latin American 1  ALT_CT_ALFA_Latin American 2  \\\n",
       "0                           5.0                          10.0   \n",
       "1                         123.0                        1330.0   \n",
       "\n",
       "   ALT_CT_ALFA_South Asian  REF_CT_ALFA_East Asian  REF_CT_ALFA_European  \\\n",
       "0                      0.0                   490.0              109377.0   \n",
       "1                   1685.0                   109.0               37269.0   \n",
       "\n",
       "   REF_CT_ALFA_Latin American 1  REF_CT_ALFA_Latin American 2  \\\n",
       "0                         673.0                        2200.0   \n",
       "1                         273.0                        2052.0   \n",
       "\n",
       "   REF_CT_ALFA_South Asian  \n",
       "0                    184.0  \n",
       "1                   3283.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alfa_grouped_data.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A description of the data contained in each column is as follows:\n",
    "\n",
    "* `VAR_NAME`: Name or identifier of the genetic variant.\n",
    "* `REF`: Reference allele (original allele).\n",
    "* `ALT`: Alternate allele (mutated allele).\n",
    "* `ALT_CT_ALFA_East Asian`: Count of the alternate allele in the East Asian population (ALFA database).\n",
    "* `ALT_CT_ALFA_European`: Count of the alternate allele in the European population (ALFA database).\n",
    "* `ALT_CT_ALFA_Latin American 1`: Count of the alternate allele in the Latin American 1 population (ALFA database).\n",
    "* `ALT_CT_ALFA_Latin American 2`: Count of the alternate allele in the Latin American 2 population (ALFA database).\n",
    "* `ALT_CT_ALFA_South Asian`: Count of the alternate allele in the South Asian population (ALFA database).\n",
    "* `REF_CT_ALFA_East Asian`: Count of the reference allele in the East Asian population (ALFA database).\n",
    "* `REF_CT_ALFA_European`: Count of the reference allele in the European population (ALFA database).\n",
    "* `REF_CT_ALFA_Latin American 1`: Count of the reference allele in the Latin American 1 population (ALFA database).\n",
    "* `REF_CT_ALFA_Latin American 2`: Count of the reference allele in the Latin American 2 population (ALFA database).\n",
    "* `REF_CT_ALFA_South Asian`: Count of the reference allele in the South Asian population (ALFA database)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "alfa_grouped_data.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"ALFA_allele_counts.csv\",\n",
    "    ),\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
