{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation of variant count data for each African population group\n",
    "\n",
    "Data on genetic variation found in African population groups was generated in-house from genetic data obtained from the [GnomAD 1000 Genomes and HGDP datasets](https://gnomad.broadinstitute.org/news/2021-10-gnomad-v3-1-2-minor-release/), and stored in `Data\\Raw\\SUB\\{gene_name}.Count.csv` files. Details on the generation of this data are not included here. Each `Data\\Raw\\SUB\\{gene_name}.Count.csv` file contains the data for a particular gene. Here, `{gene_name}` would be replaced with the specific name of the gene. The files include information on:\n",
    "\n",
    "* Genetic variant names: These are identifiers for specific genetic differences (variants) in a population.\n",
    "* Variant positions in the genome: This tells us where these genetic variations are located in the genetic code.\n",
    "* Genetic alleles: An allele refers to the different forms of a specific gene that can exist at a particular genetic position. For each variant, there are two alleles, the normal form known as the reference allele (REF), and the altered form known as the alternate allele (ALT). These alleles define the genetic variation at a given position.\n",
    "* Total copies of each variant alternate allele in the population: This shows how many times each genetic variant's alternate allele (ALT) appears in all the samples of a population.\n",
    "* Total copies of both variant alternate and reference alleles in the population: This provides the overall count of all genetic variants' alleles (both REF and ALT) mentioned in the data.\n",
    "\n",
    "The `Data\\Raw\\SUB\\{gene_name}.Count.csv` data was prepared for further analysis by performing the following steps: \n",
    "1. The data for all genes was merged into a single dataset\n",
    "2. The merged data was melted into a suitable format\n",
    "3. A unique ID was assigned to each variant, as some variants did not have unique names. \n",
    "4. Duplicate variant entries were removed\n",
    "3. Additional features were added, such as the total count of regular-non variant sites, the African continental region information, and regional variant counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change working directory\n",
    "\n",
    "import os\n",
    "\n",
    "os.chdir(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\\Analysis\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules and packages\n",
    "\n",
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\"\n",
    ")\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import in-house variant count data for each gene and subpopulation. Combine into a single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ACB_ac</th>\n",
       "      <th>ACB_tc</th>\n",
       "      <th>GWD_ac</th>\n",
       "      <th>GWD_tc</th>\n",
       "      <th>ESN_ac</th>\n",
       "      <th>ESN_tc</th>\n",
       "      <th>...</th>\n",
       "      <th>BantuSouthAfrica_tc</th>\n",
       "      <th>BantuKenya_ac</th>\n",
       "      <th>BantuKenya_tc</th>\n",
       "      <th>YRI_ac</th>\n",
       "      <th>YRI_tc</th>\n",
       "      <th>LWK_ac</th>\n",
       "      <th>LWK_tc</th>\n",
       "      <th>ASW_ac</th>\n",
       "      <th>ASW_tc</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>25</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>35</td>\n",
       "      <td>350</td>\n",
       "      <td>13</td>\n",
       "      <td>194</td>\n",
       "      <td>14</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT  ACB_ac  ACB_tc  GWD_ac  GWD_tc  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT       0     228       0     352   \n",
       "1          rs552586867  110148891   C   G       0     228       0     352   \n",
       "2           rs59409892  110148917   C   G      26     228      38     352   \n",
       "3          rs535182970  110148920   G   C       1     228       0     352   \n",
       "4           rs56406633  110148959   A   G       1     228       0     352   \n",
       "\n",
       "   ESN_ac  ESN_tc  ...  BantuSouthAfrica_tc  BantuKenya_ac  BantuKenya_tc  \\\n",
       "0       0     296  ...                   16              0             24   \n",
       "1       0     296  ...                   16              0             24   \n",
       "2      25     296  ...                   16              3             24   \n",
       "3       0     296  ...                   16              0             24   \n",
       "4       0     296  ...                   16              0             24   \n",
       "\n",
       "   YRI_ac  YRI_tc  LWK_ac  LWK_tc  ASW_ac  ASW_tc    GENE  \n",
       "0       0     350       0     194       0     142  COL4A1  \n",
       "1       2     350       0     194       0     142  COL4A1  \n",
       "2      35     350      13     194      14     142  COL4A1  \n",
       "3       0     350       0     194       0     142  COL4A1  \n",
       "4       0     350       0     194       0     142  COL4A1  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = pd.DataFrame()\n",
    "for gene in constants.GENES:\n",
    "    gene_variant_count_path = os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"SUB\",\n",
    "        \"{}_Count.csv\".format(gene),\n",
    "    )\n",
    "    gene_variant_df = pd.DataFrame()\n",
    "    if os.path.exists(gene_variant_count_path):\n",
    "        gene_variant_df = pd.read_csv(gene_variant_count_path, sep=\",\").rename(columns={\"ID\":\"VAR_NAME\"})\n",
    "        gene_variant_df[\"GENE\"] = gene\n",
    "    variants = pd.concat([variants, gene_variant_df]).drop(columns=\"Unnamed: 0\")\n",
    "variants.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the dataframe above:\n",
    "\n",
    "`ALT` represents the alternate form of a specific genetic position.\n",
    "\n",
    "`REF` represents the normal form at the genomic position, which is essentially reference genetic information.\n",
    "\n",
    "The total copies of the variant for a population are represented as `{population_group}_ac`. Here, `{population_group}` would be replaced with the specific name of the population group or subpopulation. The _ac stands for \"alternate allele count,\" which indicates the total number of copies of the specific genetic variant in the population.\n",
    "\n",
    "The total copies of both variants and non-variants for the same population are represented as `{population_group}_tc`. Again, `{population_group}` should be replaced with the actual name of the population group. The _tc stands for \"total count,\" which represents the overall count of all genetic variants (including the specific one) and the non-variant genetic information in the population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Melt data into a suitable format for further analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Change the format of this data so that the population data is displayed separately in a new column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1     ACB           0   \n",
       "1          rs552586867  110148891   C   G  COL4A1     ACB           0   \n",
       "2           rs59409892  110148917   C   G  COL4A1     ACB          26   \n",
       "3          rs535182970  110148920   G   C  COL4A1     ACB           1   \n",
       "4           rs56406633  110148959   A   G  COL4A1     ACB           1   \n",
       "\n",
       "   IH_TOTAL_CTS  \n",
       "0           228  \n",
       "1           228  \n",
       "2           228  \n",
       "3           228  \n",
       "4           228  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Separate total count and alternate count information\n",
    "alt_ct_columns = variants.filter(regex=\"_ac|VAR_NAME|POS|REF|ALT|GENE\")\n",
    "total_ct_columns = variants.filter(regex=\"_tc|VAR_NAME|POS|REF|ALT|GENE\")\n",
    "\n",
    "# Melt information\n",
    "alt_ct_columns = alt_ct_columns.melt(\n",
    "    id_vars=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\"],\n",
    "    var_name=\"SUB_POP\",\n",
    "    value_name=\"IH_ALT_CTS\",\n",
    ")\n",
    "total_ct_columns = total_ct_columns.melt(\n",
    "    id_vars=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\"],\n",
    "    var_name=\"SUB_POP\",\n",
    "    value_name=\"IH_TOTAL_CTS\",\n",
    ")\n",
    "\n",
    "# Remove information after underscore in SUB_POP column\n",
    "alt_ct_columns[\"SUB_POP\"] = alt_ct_columns[\"SUB_POP\"].str.rsplit(\"_\", n=0).str.get(0)\n",
    "total_ct_columns[\"SUB_POP\"] = (\n",
    "    total_ct_columns[\"SUB_POP\"].str.rsplit(\"_\", n=0).str.get(0)\n",
    ")\n",
    "\n",
    "# Combine formatted information\n",
    "\n",
    "ih_allele_counts = pd.merge(\n",
    "    alt_ct_columns, total_ct_columns, on=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\", \"SUB_POP\"]\n",
    ")\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign a unique ID to each variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148882_CT_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148891_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>110148917_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148920_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148959_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1     ACB           0   \n",
       "1          rs552586867  110148891   C   G  COL4A1     ACB           0   \n",
       "2           rs59409892  110148917   C   G  COL4A1     ACB          26   \n",
       "3          rs535182970  110148920   G   C  COL4A1     ACB           1   \n",
       "4           rs56406633  110148959   A   G  COL4A1     ACB           1   \n",
       "\n",
       "   IH_TOTAL_CTS              ID  \n",
       "0           228  110148882_CT_C  \n",
       "1           228   110148891_G_C  \n",
       "2           228   110148917_G_C  \n",
       "3           228   110148920_C_G  \n",
       "4           228   110148959_G_A  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts[\"ID\"] = (\n",
    "    ih_allele_counts[[\"POS\", \"ALT\", \"REF\"]].astype(\"str\").agg(\"_\".join, axis=1)\n",
    ")\n",
    "\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drop duplicate entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts = ih_allele_counts.drop_duplicates(\n",
    "    [\"ID\", \"SUB_POP\", \"SUB_POP\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add additional data features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Correct total allele counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the data above, the total count of all alleles (REF and ALT) is included as `IH_TOTAL_CTS`. Some samples, from which this data was generated, lacked variant ALT data for specific positions in the genome. The total count was calculated only using the available data, which might make the variant frequency estimates inaccurate for a population.\n",
    "\n",
    "To improve accuracy, we can assume that if data is missing for a certain genetic variant at a position, that variant does not exist in those samples. Based on this assumption, we can correct the total allele count by considering the missing data as non-variants (REF alleles only). This correction helps us get more reliable estimates of the variant frequency for the population.\n",
    "\n",
    "In simple terms, the total allele counts `IH_TOTAL_CTS` were adjusted to account for missing information, assuming that if data is missing, it means the variant doesn't exist. This will give us a more accurate idea of how common the variants are in the population.\n",
    "\n",
    "To perform this correction, the following steps were performed: \n",
    "\n",
    "1. For each subpopulation, the total number of individuals were calculated.\n",
    "2. The number of individuals were multiplied by 2 to account for the fact that each individual has two copies of an allele (one from each parent) in either a variant or non-variant form. This value was then the corrected `IH_TOTAL_CTS` value, `CORR_IH_TOTAL_CTS`.\n",
    "3. The total number of alternate allele counts `IH_ALT_CTS` were subtracted from `CORR_IH_TOTAL_CTS` to get the count of non-variant alleles `IH_REF_CTS`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148882_CT_C</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148920_C_G</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148959_G_A</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1     ACB           0   \n",
       "1          rs552586867  110148891   C   G  COL4A1     ACB           0   \n",
       "2           rs59409892  110148917   C   G  COL4A1     ACB          26   \n",
       "3          rs535182970  110148920   G   C  COL4A1     ACB           1   \n",
       "4           rs56406633  110148959   A   G  COL4A1     ACB           1   \n",
       "\n",
       "   IH_TOTAL_CTS              ID  CORR_IH_TOTAL_CTS  \n",
       "0           228  110148882_CT_C                228  \n",
       "1           228   110148891_G_C                228  \n",
       "2           228   110148917_G_C                228  \n",
       "3           228   110148920_C_G                228  \n",
       "4           228   110148959_G_A                228  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import sample and population data\n",
    "sample_subpopulations = pd.read_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Sample_populations.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Group samples by population to get the number of samples (individuals) per population\n",
    "\n",
    "grouped_sample_subpopulations = functions.group_and_count(sample_subpopulations, [\"SUB\"]).reset_index().rename(columns={\"SAMPLE_NAME\":\"SAMPLE_COUNT\"}).drop(columns=\"REG\")\n",
    "\n",
    "# Calculate the total allele count per subpopulation\n",
    "grouped_sample_subpopulations[\"CORR_IH_TOTAL_CTS\"] = (\n",
    "    grouped_sample_subpopulations.SAMPLE_COUNT * 2\n",
    ")\n",
    "\n",
    "# Add total allele count to allele count dataframe\n",
    "ih_allele_counts = (\n",
    "    ih_allele_counts.merge(\n",
    "        grouped_sample_subpopulations, how=\"left\", left_on=\"SUB_POP\", right_on=\"SUB\"\n",
    "    )\n",
    "    .reset_index()\n",
    "    .drop(columns=[\"SUB\", \"SAMPLE_COUNT\", \"index\", \"Unnamed: 0\"])\n",
    ")\n",
    "\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculate the corrected reference allele counts, and the corrected alternate allele frequency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate corrected reference allele counts\n",
    "ih_allele_counts[\"CORR_IH_REF_CTS\"] = (\n",
    "    ih_allele_counts[\"CORR_IH_TOTAL_CTS\"]\n",
    "    - ih_allele_counts[\"IH_ALT_CTS\"]\n",
    ")\n",
    "\n",
    "# Calculate corrected alternate allele frequencies\n",
    "ih_allele_counts[\"CORR_IH_AF\"] = (\n",
    "    ih_allele_counts[\"IH_ALT_CTS\"]\n",
    "    / ih_allele_counts[\"CORR_IH_TOTAL_CTS\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add regional information for each subpopulation group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>CORR_IH_AF</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148882_CT_C</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>228</td>\n",
       "      <td>202</td>\n",
       "      <td>0.114035</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148920_C_G</td>\n",
       "      <td>228</td>\n",
       "      <td>227</td>\n",
       "      <td>0.004386</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148959_G_A</td>\n",
       "      <td>228</td>\n",
       "      <td>227</td>\n",
       "      <td>0.004386</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1     ACB           0   \n",
       "1          rs552586867  110148891   C   G  COL4A1     ACB           0   \n",
       "2           rs59409892  110148917   C   G  COL4A1     ACB          26   \n",
       "3          rs535182970  110148920   G   C  COL4A1     ACB           1   \n",
       "4           rs56406633  110148959   A   G  COL4A1     ACB           1   \n",
       "\n",
       "   IH_TOTAL_CTS              ID  CORR_IH_TOTAL_CTS  CORR_IH_REF_CTS  \\\n",
       "0           228  110148882_CT_C                228              228   \n",
       "1           228   110148891_G_C                228              228   \n",
       "2           228   110148917_G_C                228              202   \n",
       "3           228   110148920_C_G                228              227   \n",
       "4           228   110148959_G_A                228              227   \n",
       "\n",
       "   CORR_IH_AF  REG  \n",
       "0    0.000000  ACB  \n",
       "1    0.000000  ACB  \n",
       "2    0.114035  ACB  \n",
       "3    0.004386  ACB  \n",
       "4    0.004386  ACB  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts[\"REG\"] = ih_allele_counts[\"SUB_POP\"].map(\n",
    "    constants.LD_REGIONAL_CLASSIFICATION\n",
    ")\n",
    "\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add continental African count information\n",
    "\n",
    "Total African Population: This includes all populations, including African American `ASW` and African Caribbean `ACB` samples. By considering all African populations together, we can analyze the frequency of a variant across the entire African continent.\n",
    "\n",
    "Recent Africans Residing in Africa: This group excludes African American and Caribbean populations. It specifically focuses on individuals of recent African origin who still reside on the African continent. This allows us to study the frequency of a variant within African populations without the influence of African American and Caribbean genetic contributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>CORR_IH_AF</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>376363</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>San</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376364</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>BantuSouthAfrica</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376365</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>YRI</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>350</td>\n",
       "      <td>350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376366</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376367</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1588</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>1608</td>\n",
       "      <td>1608</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Recent African</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 VAR_NAME       POS REF ALT   GENE           SUB_POP  \\\n",
       "376363  chr19:48256362G-T  48256362   G   T  CARD8               San   \n",
       "376364  chr19:48256362G-T  48256362   G   T  CARD8  BantuSouthAfrica   \n",
       "376365  chr19:48256362G-T  48256362   G   T  CARD8               YRI   \n",
       "376366  chr19:48256362G-T  48256362   G   T  CARD8               ACB   \n",
       "376367  chr19:48256362G-T  48256362   G   T  CARD8               NaN   \n",
       "\n",
       "        IH_ALT_CTS  IH_TOTAL_CTS            ID  CORR_IH_TOTAL_CTS  \\\n",
       "376363           0            10  48256362_T_G                 12   \n",
       "376364           0            16  48256362_T_G                 16   \n",
       "376365           0           350  48256362_T_G                350   \n",
       "376366           0           228  48256362_T_G                228   \n",
       "376367           0          1588  48256362_T_G               1608   \n",
       "\n",
       "        CORR_IH_REF_CTS  CORR_IH_AF             REG  \n",
       "376363               12         0.0              SA  \n",
       "376364               16         0.0              SA  \n",
       "376365              350         0.0              WA  \n",
       "376366              228         0.0             ACB  \n",
       "376367             1608         0.0  Recent African  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract African allele count data using a custom function\n",
    "\n",
    "total_africa_ct = functions.grouped_pop_allele_counts(ih_allele_counts.copy(), \"African\")\n",
    "\n",
    "# Extract Recent African allele count data using a custom function\n",
    "\n",
    "filters = (ih_allele_counts.REG != \"ACB\") & (\n",
    "    ih_allele_counts.REG != \"ASW\"\n",
    ")\n",
    "\n",
    "native_africa_ct = functions.grouped_pop_allele_counts(\n",
    "    ih_allele_counts.copy(), \"Recent African\", filters\n",
    ")\n",
    "\n",
    "# Concatenate the African and Recent African allele count data with the African subpopulation allele count data\n",
    "\n",
    "ih_allele_counts_grouped = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            ih_allele_counts,\n",
    "            total_africa_ct,\n",
    "            native_africa_ct,\n",
    "        ]\n",
    "    )\n",
    "    .sort_values(\"ID\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "ih_allele_counts_grouped.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Classify variants as SNPs or INDELs\n",
    "\n",
    "Variants can be classified as:\n",
    "\n",
    "* Single nucleotide polymorphisms (SNPs) if they result in the exchange of one DNA base pair for another \n",
    "\n",
    "**OR**\n",
    "\n",
    "* Insertion/deletions (INDELs) if they result in the deletion or insertion of one or more DNA base pairs\n",
    "\n",
    "Adding these variant classifications may prove useful for further analysis of the variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>CORR_IH_AF</th>\n",
       "      <th>REG</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>376363</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>San</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SA</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376364</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>BantuSouthAfrica</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SA</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376365</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>YRI</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>350</td>\n",
       "      <td>350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376366</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACB</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376367</th>\n",
       "      <td>chr19:48256362G-T</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1588</td>\n",
       "      <td>48256362_T_G</td>\n",
       "      <td>1608</td>\n",
       "      <td>1608</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 VAR_NAME       POS REF ALT   GENE           SUB_POP  \\\n",
       "376363  chr19:48256362G-T  48256362   G   T  CARD8               San   \n",
       "376364  chr19:48256362G-T  48256362   G   T  CARD8  BantuSouthAfrica   \n",
       "376365  chr19:48256362G-T  48256362   G   T  CARD8               YRI   \n",
       "376366  chr19:48256362G-T  48256362   G   T  CARD8               ACB   \n",
       "376367  chr19:48256362G-T  48256362   G   T  CARD8               NaN   \n",
       "\n",
       "        IH_ALT_CTS  IH_TOTAL_CTS            ID  CORR_IH_TOTAL_CTS  \\\n",
       "376363           0            10  48256362_T_G                 12   \n",
       "376364           0            16  48256362_T_G                 16   \n",
       "376365           0           350  48256362_T_G                350   \n",
       "376366           0           228  48256362_T_G                228   \n",
       "376367           0          1588  48256362_T_G               1608   \n",
       "\n",
       "        CORR_IH_REF_CTS  CORR_IH_AF             REG VARIANT_TYPE  \n",
       "376363               12         0.0              SA          SNP  \n",
       "376364               16         0.0              SA          SNP  \n",
       "376365              350         0.0              WA          SNP  \n",
       "376366              228         0.0             ACB          SNP  \n",
       "376367             1608         0.0  Recent African          SNP  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts_grouped[\"VARIANT_TYPE\"] = np.where(\n",
    "    (ih_allele_counts_grouped.ALT.str.len() > 1)\n",
    "    | (ih_allele_counts_grouped.REF.str.len() > 1),\n",
    "    \"INDEL\",\n",
    "    \"SNP\",\n",
    ")\n",
    "\n",
    "ih_allele_counts_grouped.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save in-house allele count data to CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts_grouped.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts.csv\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare data in suitable format for Fisher's Tests\n",
    "\n",
    "The in-house allele count data needs to be formatted differently to conduct Fisher's Tests to test for significant differences in allele frequency between populations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>REG</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th>ALT_CT_IH_ACB</th>\n",
       "      <th>ALT_CT_IH_ASW</th>\n",
       "      <th>ALT_CT_IH_African</th>\n",
       "      <th>ALT_CT_IH_CA</th>\n",
       "      <th>...</th>\n",
       "      <th>ALT_CT_IH_SA</th>\n",
       "      <th>ALT_CT_IH_WA</th>\n",
       "      <th>CORR_REF_CT_IH_ACB</th>\n",
       "      <th>CORR_REF_CT_IH_ASW</th>\n",
       "      <th>CORR_REF_CT_IH_African</th>\n",
       "      <th>CORR_REF_CT_IH_CA</th>\n",
       "      <th>CORR_REF_CT_IH_EA</th>\n",
       "      <th>CORR_REF_CT_IH_Recent African</th>\n",
       "      <th>CORR_REF_CT_IH_SA</th>\n",
       "      <th>CORR_REF_CT_IH_WA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:34438836T-C</td>\n",
       "      <td>34438836_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1977</td>\n",
       "      <td>80</td>\n",
       "      <td>217</td>\n",
       "      <td>1607</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11:34438889G-C</td>\n",
       "      <td>34438889_C_G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438889</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1977</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1607</td>\n",
       "      <td>27</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:34438910C-T</td>\n",
       "      <td>34438910_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438910</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:34439179A-G</td>\n",
       "      <td>34439179_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439179</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:34439188C-G</td>\n",
       "      <td>34439188_G_C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439188</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>142</td>\n",
       "      <td>1978</td>\n",
       "      <td>80</td>\n",
       "      <td>218</td>\n",
       "      <td>1608</td>\n",
       "      <td>28</td>\n",
       "      <td>1282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "REG           VAR_NAME            ID REF ALT GENE       POS  ALT_CT_IH_ACB  \\\n",
       "0    chr11:34438836T-C  34438836_C_T   T   C  CAT  34438836              0   \n",
       "1    chr11:34438889G-C  34438889_C_G   G   C  CAT  34438889              0   \n",
       "2    chr11:34438910C-T  34438910_T_C   C   T  CAT  34438910              0   \n",
       "3    chr11:34439179A-G  34439179_G_A   A   G  CAT  34439179              0   \n",
       "4    chr11:34439188C-G  34439188_G_C   C   G  CAT  34439188              0   \n",
       "\n",
       "REG  ALT_CT_IH_ASW  ALT_CT_IH_African  ALT_CT_IH_CA  ...  ALT_CT_IH_SA  \\\n",
       "0                0                  1             0  ...             0   \n",
       "1                0                  1             0  ...             1   \n",
       "2                0                  0             0  ...             0   \n",
       "3                0                  0             0  ...             0   \n",
       "4                0                  0             0  ...             0   \n",
       "\n",
       "REG  ALT_CT_IH_WA  CORR_REF_CT_IH_ACB  CORR_REF_CT_IH_ASW  \\\n",
       "0               0                 228                 142   \n",
       "1               0                 228                 142   \n",
       "2               0                 228                 142   \n",
       "3               0                 228                 142   \n",
       "4               0                 228                 142   \n",
       "\n",
       "REG  CORR_REF_CT_IH_African  CORR_REF_CT_IH_CA  CORR_REF_CT_IH_EA  \\\n",
       "0                      1977                 80                217   \n",
       "1                      1977                 80                218   \n",
       "2                      1978                 80                218   \n",
       "3                      1978                 80                218   \n",
       "4                      1978                 80                218   \n",
       "\n",
       "REG  CORR_REF_CT_IH_Recent African  CORR_REF_CT_IH_SA  CORR_REF_CT_IH_WA  \n",
       "0                             1607                 28               1282  \n",
       "1                             1607                 27               1282  \n",
       "2                             1608                 28               1282  \n",
       "3                             1608                 28               1282  \n",
       "4                             1608                 28               1282  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate list of all unique regions in ih_allele_counts_grouped dataframe\n",
    "inhouse_populations = [region for region in set(ih_allele_counts_grouped.REG.values)]\n",
    "\n",
    "# Group allele counts by region\n",
    "\n",
    "ih_allele_counts_regions = (\n",
    "    ih_allele_counts_grouped.groupby(\n",
    "        by=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\", \"REG\"]\n",
    "    )\n",
    "    .sum(numeric_only=True)\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "ih_allele_counts_regions\n",
    "\n",
    "# Pivot data\n",
    "ih_allele_counts_pivot = ih_allele_counts_regions.pivot(\n",
    "    index=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\"],\n",
    "    columns=\"REG\",\n",
    "    values=[\"IH_ALT_CTS\", \"CORR_IH_REF_CTS\"],\n",
    ")\n",
    "\n",
    "# Separate alternate and reference count data into different dataframes to facilate renaming of count columns appropriately\n",
    "ih_data_alt = (\n",
    "    ih_allele_counts_pivot[[\"IH_ALT_CTS\"]].droplevel(level=0, axis=1).reset_index()\n",
    ")\n",
    "\n",
    "ih_data_corr_ref = (\n",
    "    ih_allele_counts_pivot[[\"CORR_IH_REF_CTS\"]].droplevel(level=0, axis=1).reset_index()\n",
    ")\n",
    "\n",
    "# Add appropriate prefixes to alt and ref columns\n",
    "ih_data_alt = functions.add_prefix_dataframe_col_names(\n",
    "    ih_data_alt, inhouse_populations, \"ALT_CT_IH_\"\n",
    ")\n",
    "\n",
    "ih_data_corr_ref = functions.add_prefix_dataframe_col_names(\n",
    "    ih_data_corr_ref, inhouse_populations, \"CORR_REF_CT_IH_\"\n",
    ")\n",
    "\n",
    "# Merge renamed alternate and reference count data\n",
    "ih_recent_afr = ih_data_alt.merge(\n",
    "    ih_data_corr_ref, on=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\"]\n",
    ")\n",
    "\n",
    "ih_recent_afr.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Fisher's Test data to a csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_recent_afr.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts_fishers.csv\",\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
