{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of variant count data for each African population group\n",
    "\n",
    "Data on genetic variation found in African population groups was generated in-house by processing genomic data obtained from [GnomAD 1000 Genomes and HGDP datasets](https://gnomad.broadinstitute.org/news/2021-10-gnomad-v3-1-2-minor-release/) through a [Snakemake bioinformatics pipeline](https://github.com/Tuks-ICMM/Pharmacogenetic-Analysis-Pipeline). The generated data was stored in `Data\\Raw\\SUB\\{gene_name}.Count.csv` files. Each `Data\\Raw\\SUB\\{gene_name}.Count.csv` file contains the data on variants identified in a particular gene. Here, `{gene_name}` would be replaced with the specific name of the gene. The files include information on:\n",
    "\n",
    "* Genetic variant names: These are identifiers for specific genetic differences (variants) in a population.\n",
    "* Variant positions in the genome: This tells us where these genetic variations are located in the genetic code.\n",
    "* Genetic alleles: An allele refers to the different forms of a specific gene that can exist at a particular genetic position. For each variant, there are two alleles, the normal form known as the reference allele (REF), and the altered form known as the alternate allele (ALT). These alleles define the genetic variation at a given position.\n",
    "* Total copies of each variant alternate allele in the population: This shows how many times each genetic variant's alternate allele (ALT) appears in all the samples of a population.\n",
    "* Total copies of both variant alternate and reference alleles in the population: This provides the overall count of all genetic variants' alleles (both REF and ALT) mentioned in the data.\n",
    "* Sample subpopulation group: The ethnolinguistic classification of the African population from which the genetic data sample originated.\n",
    "\n",
    "The `Data\\Raw\\SUB\\{gene_name}.Count.csv` data was prepared for further analysis by performing the following steps: \n",
    "1. The data for all genes was merged into a single dataset\n",
    "2. The merged data was melted into a suitable format\n",
    "3. A unique ID was assigned to each variant, as some variants did not have unique names. \n",
    "4. Duplicate variant entries were removed\n",
    "5. Remove irrelevant data\n",
    "6. Additional features were added, such as the total count of regular-non variant sites, and regional information on the African ethnolinguistic/subpopulation groups.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import in-house African variant data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ACB_ac</th>\n",
       "      <th>ACB_tc</th>\n",
       "      <th>GWD_ac</th>\n",
       "      <th>GWD_tc</th>\n",
       "      <th>ESN_ac</th>\n",
       "      <th>ESN_tc</th>\n",
       "      <th>...</th>\n",
       "      <th>BantuSouthAfrica_tc</th>\n",
       "      <th>BantuKenya_ac</th>\n",
       "      <th>BantuKenya_tc</th>\n",
       "      <th>YRI_ac</th>\n",
       "      <th>YRI_tc</th>\n",
       "      <th>LWK_ac</th>\n",
       "      <th>LWK_tc</th>\n",
       "      <th>ASW_ac</th>\n",
       "      <th>ASW_tc</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>25</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>35</td>\n",
       "      <td>350</td>\n",
       "      <td>13</td>\n",
       "      <td>194</td>\n",
       "      <td>14</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>296</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT  ACB_ac  ACB_tc  GWD_ac  GWD_tc  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT       0     228       0     352   \n",
       "1          rs552586867  110148891   C   G       0     228       0     352   \n",
       "2           rs59409892  110148917   C   G      26     228      38     352   \n",
       "3          rs535182970  110148920   G   C       1     228       0     352   \n",
       "4           rs56406633  110148959   A   G       1     228       0     352   \n",
       "\n",
       "   ESN_ac  ESN_tc  ...  BantuSouthAfrica_tc  BantuKenya_ac  BantuKenya_tc  \\\n",
       "0       0     296  ...                   16              0             24   \n",
       "1       0     296  ...                   16              0             24   \n",
       "2      25     296  ...                   16              3             24   \n",
       "3       0     296  ...                   16              0             24   \n",
       "4       0     296  ...                   16              0             24   \n",
       "\n",
       "   YRI_ac  YRI_tc  LWK_ac  LWK_tc  ASW_ac  ASW_tc    GENE  \n",
       "0       0     350       0     194       0     142  COL4A1  \n",
       "1       2     350       0     194       0     142  COL4A1  \n",
       "2      35     350      13     194      14     142  COL4A1  \n",
       "3       0     350       0     194       0     142  COL4A1  \n",
       "4       0     350       0     194       0     142  COL4A1  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import CSVs with variants identified in-house in African populations for genes of interest.\n",
    "\n",
    "variants = pd.DataFrame()\n",
    "\n",
    "genes = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Metadata\",\n",
    "        \"locations.csv\",\n",
    "    )\n",
    ").location_name\n",
    "\n",
    "for gene in genes:\n",
    "    gene_variant_count_path = os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"SUB\",\n",
    "        \"{}_Count.csv\".format(gene),\n",
    "    )\n",
    "\n",
    "    # Combine variant data on each gene into a single dataframe\n",
    "    gene_variant_df = pd.DataFrame()\n",
    "    if os.path.exists(gene_variant_count_path):\n",
    "        gene_variant_df = pd.read_csv(gene_variant_count_path, sep=\",\").rename(\n",
    "            columns={\"ID\": \"VAR_NAME\"}\n",
    "        )\n",
    "        gene_variant_df[\"GENE\"] = gene\n",
    "    variants = pd.concat([variants, gene_variant_df]).drop(columns=\"Unnamed: 0\")\n",
    "\n",
    "variants.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the dataframe above:\n",
    "\n",
    "`ALT` represents the alternate form of a specific genetic position.\n",
    "\n",
    "`REF` represents the normal form at the genomic position, which is essentially reference genetic information.\n",
    "\n",
    "The total copies of the variant for a population are represented as `{population_group}_ac`. Here, `{population_group}` would be replaced with the specific name of the population group or subpopulation. The _ac stands for \"alternate allele count,\" which indicates the total number of copies of the specific genetic variant in the population.\n",
    "\n",
    "The total copies of both variants and non-variants for the same population are represented as `{population_group}_tc`. Again, `{population_group}` should be replaced with the actual name of the population group. The _tc stands for \"total count,\" which represents the overall count of all genetic variants (including the specific one) and the non-variant genetic information in the population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display the data for each population group separately in a new column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE SUB_POP  IH_ALT_CTS  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1     ACB           0   \n",
       "1          rs552586867  110148891   C   G  COL4A1     ACB           0   \n",
       "2           rs59409892  110148917   C   G  COL4A1     ACB          26   \n",
       "3          rs535182970  110148920   G   C  COL4A1     ACB           1   \n",
       "4           rs56406633  110148959   A   G  COL4A1     ACB           1   \n",
       "\n",
       "   IH_TOTAL_CTS  \n",
       "0           228  \n",
       "1           228  \n",
       "2           228  \n",
       "3           228  \n",
       "4           228  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Separate total count and alternate count information\n",
    "alt_ct_columns = variants.filter(regex=\"_ac|VAR_NAME|POS|REF|ALT|GENE\")\n",
    "total_ct_columns = variants.filter(regex=\"_tc|VAR_NAME|POS|REF|ALT|GENE\")\n",
    "\n",
    "# Melt information\n",
    "alt_ct_columns = alt_ct_columns.melt(\n",
    "    id_vars=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\"],\n",
    "    var_name=\"SUB_POP\",\n",
    "    value_name=\"IH_ALT_CTS\",\n",
    ")\n",
    "total_ct_columns = total_ct_columns.melt(\n",
    "    id_vars=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\"],\n",
    "    var_name=\"SUB_POP\",\n",
    "    value_name=\"IH_TOTAL_CTS\",\n",
    ")\n",
    "\n",
    "# Remove information after underscore in SUB_POP column\n",
    "alt_ct_columns[\"SUB_POP\"] = alt_ct_columns[\"SUB_POP\"].str.rsplit(\"_\", n=0).str.get(0)\n",
    "total_ct_columns[\"SUB_POP\"] = (\n",
    "    total_ct_columns[\"SUB_POP\"].str.rsplit(\"_\", n=0).str.get(0)\n",
    ")\n",
    "\n",
    "# Combine formatted information\n",
    "\n",
    "ih_allele_counts = pd.merge(\n",
    "    alt_ct_columns,\n",
    "    total_ct_columns,\n",
    "    on=[\"VAR_NAME\", \"POS\", \"REF\", \"ALT\", \"GENE\", \"SUB_POP\"],\n",
    ")\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename subpopulations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts = ih_allele_counts.replace({\"SUB_POP\": constants.SUBPOP_RENAME})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign a unique ID to each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some variants do not have unique names. This will complicate downstream analysis of the data. Add a column with a unique ID for each variant to rectify this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>African Caribbean</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148882_CT_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>African Caribbean</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "      <td>110148891_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>African Caribbean</td>\n",
       "      <td>26</td>\n",
       "      <td>228</td>\n",
       "      <td>110148917_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>African Caribbean</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148920_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>African Caribbean</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>110148959_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              VAR_NAME        POS REF ALT    GENE            SUB_POP  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  COL4A1  African Caribbean   \n",
       "1          rs552586867  110148891   C   G  COL4A1  African Caribbean   \n",
       "2           rs59409892  110148917   C   G  COL4A1  African Caribbean   \n",
       "3          rs535182970  110148920   G   C  COL4A1  African Caribbean   \n",
       "4           rs56406633  110148959   A   G  COL4A1  African Caribbean   \n",
       "\n",
       "   IH_ALT_CTS  IH_TOTAL_CTS              ID  \n",
       "0           0           228  110148882_CT_C  \n",
       "1           0           228   110148891_G_C  \n",
       "2          26           228   110148917_G_C  \n",
       "3           1           228   110148920_C_G  \n",
       "4           1           228   110148959_G_A  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts[\"ID\"] = (\n",
    "    ih_allele_counts[[\"POS\", \"ALT\", \"REF\"]].astype(\"str\").agg(\"_\".join, axis=1)\n",
    ")\n",
    "\n",
    "ih_allele_counts.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drop duplicate entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts = ih_allele_counts.drop_duplicates([\"ID\", \"SUB_POP\", \"SUB_POP\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove irrelevant data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop African American and African Caribbean data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts_filtered = ih_allele_counts[\n",
    "    ~(\n",
    "        (ih_allele_counts.SUB_POP == \"African American\")\n",
    "        | (ih_allele_counts.SUB_POP == \"African Caribbean\")\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove rows with variants with alternate allele counts of 0. These variants are not present in the particular population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts_filtered = ih_allele_counts_filtered[\n",
    "    ~(ih_allele_counts_filtered.IH_ALT_CTS == 0.0)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add additional data features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Correct total allele counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the data above, the total count of all alleles (REF and ALT) is included as `IH_TOTAL_CTS`. Some samples, from which this data was generated, lacked variant ALT data for specific positions in the genome. The total count was calculated only using the available data, which might make the variant frequency estimates inaccurate for a population.\n",
    "\n",
    "To improve accuracy, we can assume that if data is missing for a certain variant ALT allele at a position, that variant ALT allele does not exist in those samples, and can thus be given a count of 0. Based on this assumption, we can correct the total allele count by considering the missing data as non-variants (REF alleles only). This correction helps us get more reliable estimates of the variant frequency for the population.\n",
    "\n",
    "In simple terms, the total allele counts `IH_TOTAL_CTS` were adjusted to account for missing information, assuming that if data is missing, it means the variant doesn't exist. This will give us a more accurate idea of how common the variants are in the population.\n",
    "\n",
    "To perform this correction, the following steps were performed: \n",
    "\n",
    "1. For each subpopulation, the total number of individuals were calculated.\n",
    "2. The number of individuals were multiplied by 2 to account for the fact that each individual has two copies of an allele (one from each parent) in either a variant (ALT) or non-variant (REF) form. This value was then the corrected `IH_TOTAL_CTS` value, `CORR_IH_TOTAL_CTS`.\n",
    "3. The total number of alternate allele counts `IH_ALT_CTS` were subtracted from `CORR_IH_TOTAL_CTS` to get the count of non-variant alleles `IH_REF_CTS`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs139916479</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149349_A_G</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs552877576</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149494_T_C</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13:110149646CTTTAT-C</td>\n",
       "      <td>110149646</td>\n",
       "      <td>CTTTAT</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149646_C_CTTTAT</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs13260</td>\n",
       "      <td>110149776</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>71</td>\n",
       "      <td>352</td>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  VAR_NAME        POS     REF ALT    GENE   SUB_POP  \\\n",
       "0               rs59409892  110148917       C   G  COL4A1  Mandinka   \n",
       "1              rs139916479  110149349       G   A  COL4A1  Mandinka   \n",
       "2              rs552877576  110149494       C   T  COL4A1  Mandinka   \n",
       "3  chr13:110149646CTTTAT-C  110149646  CTTTAT   C  COL4A1  Mandinka   \n",
       "4                  rs13260  110149776       G   T  COL4A1  Mandinka   \n",
       "\n",
       "   IH_ALT_CTS  IH_TOTAL_CTS                  ID  CORR_IH_TOTAL_CTS  \n",
       "0          38           352       110148917_G_C              352.0  \n",
       "1           2           352       110149349_A_G              352.0  \n",
       "2           2           352       110149494_T_C              352.0  \n",
       "3           2           352  110149646_C_CTTTAT              352.0  \n",
       "4          71           352       110149776_T_G              352.0  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import sample and population data\n",
    "sample_subpopulations = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Sample_populations.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Group samples by population to get the number of samples (individuals) per population\n",
    "\n",
    "grouped_sample_subpopulations = (\n",
    "    functions.group_and_count(sample_subpopulations, [\"SUB\"])\n",
    "    .reset_index()\n",
    "    .rename(columns={\"SAMPLE_NAME\": \"SAMPLE_COUNT\"})\n",
    "    .drop(columns=\"REG\")\n",
    ")\n",
    "\n",
    "# Calculate the total allele count per subpopulation\n",
    "grouped_sample_subpopulations[\"CORR_IH_TOTAL_CTS\"] = (\n",
    "    grouped_sample_subpopulations.SAMPLE_COUNT * 2\n",
    ")\n",
    "\n",
    "# Add total allele count to allele count dataframe\n",
    "ih_allele_counts_filtered = (\n",
    "    ih_allele_counts_filtered.merge(\n",
    "        grouped_sample_subpopulations, how=\"left\", left_on=\"SUB_POP\", right_on=\"SUB\"\n",
    "    )\n",
    "    .reset_index()\n",
    "    .drop(columns=[\"SUB\", \"SAMPLE_COUNT\", \"index\", \"Unnamed: 0\"])\n",
    ")\n",
    "\n",
    "ih_allele_counts_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculate the corrected reference allele counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>352.0</td>\n",
       "      <td>314.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs139916479</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149349_A_G</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs552877576</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149494_T_C</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13:110149646CTTTAT-C</td>\n",
       "      <td>110149646</td>\n",
       "      <td>CTTTAT</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149646_C_CTTTAT</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs13260</td>\n",
       "      <td>110149776</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>71</td>\n",
       "      <td>352</td>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>352.0</td>\n",
       "      <td>281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45300</th>\n",
       "      <td>rs111859112</td>\n",
       "      <td>33029010</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>2</td>\n",
       "      <td>194</td>\n",
       "      <td>33029010_G_T</td>\n",
       "      <td>194.0</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45301</th>\n",
       "      <td>rs13046814</td>\n",
       "      <td>33029069</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>10</td>\n",
       "      <td>194</td>\n",
       "      <td>33029069_G_T</td>\n",
       "      <td>194.0</td>\n",
       "      <td>184.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45302</th>\n",
       "      <td>rs547410200</td>\n",
       "      <td>33029136</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>1</td>\n",
       "      <td>194</td>\n",
       "      <td>33029136_C_T</td>\n",
       "      <td>194.0</td>\n",
       "      <td>193.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45303</th>\n",
       "      <td>rs182058038</td>\n",
       "      <td>33029193</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>1</td>\n",
       "      <td>194</td>\n",
       "      <td>33029193_A_T</td>\n",
       "      <td>194.0</td>\n",
       "      <td>193.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45304</th>\n",
       "      <td>rs7278343</td>\n",
       "      <td>33029257</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>28</td>\n",
       "      <td>194</td>\n",
       "      <td>33029257_G_A</td>\n",
       "      <td>194.0</td>\n",
       "      <td>166.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45305 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      VAR_NAME        POS     REF ALT    GENE   SUB_POP  \\\n",
       "0                   rs59409892  110148917       C   G  COL4A1  Mandinka   \n",
       "1                  rs139916479  110149349       G   A  COL4A1  Mandinka   \n",
       "2                  rs552877576  110149494       C   T  COL4A1  Mandinka   \n",
       "3      chr13:110149646CTTTAT-C  110149646  CTTTAT   C  COL4A1  Mandinka   \n",
       "4                      rs13260  110149776       G   T  COL4A1  Mandinka   \n",
       "...                        ...        ...     ...  ..     ...       ...   \n",
       "45300              rs111859112   33029010       T   G   OLIG2     Luhya   \n",
       "45301               rs13046814   33029069       T   G   OLIG2     Luhya   \n",
       "45302              rs547410200   33029136       T   C   OLIG2     Luhya   \n",
       "45303              rs182058038   33029193       T   A   OLIG2     Luhya   \n",
       "45304                rs7278343   33029257       A   G   OLIG2     Luhya   \n",
       "\n",
       "       IH_ALT_CTS  IH_TOTAL_CTS                  ID  CORR_IH_TOTAL_CTS  \\\n",
       "0              38           352       110148917_G_C              352.0   \n",
       "1               2           352       110149349_A_G              352.0   \n",
       "2               2           352       110149494_T_C              352.0   \n",
       "3               2           352  110149646_C_CTTTAT              352.0   \n",
       "4              71           352       110149776_T_G              352.0   \n",
       "...           ...           ...                 ...                ...   \n",
       "45300           2           194        33029010_G_T              194.0   \n",
       "45301          10           194        33029069_G_T              194.0   \n",
       "45302           1           194        33029136_C_T              194.0   \n",
       "45303           1           194        33029193_A_T              194.0   \n",
       "45304          28           194        33029257_G_A              194.0   \n",
       "\n",
       "       CORR_IH_REF_CTS  \n",
       "0                314.0  \n",
       "1                350.0  \n",
       "2                350.0  \n",
       "3                350.0  \n",
       "4                281.0  \n",
       "...                ...  \n",
       "45300            192.0  \n",
       "45301            184.0  \n",
       "45302            193.0  \n",
       "45303            193.0  \n",
       "45304            166.0  \n",
       "\n",
       "[45305 rows x 11 columns]"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate corrected reference allele counts\n",
    "ih_allele_counts_filtered[\"CORR_IH_REF_CTS\"] = (\n",
    "    ih_allele_counts_filtered[\"CORR_IH_TOTAL_CTS\"]\n",
    "    - ih_allele_counts_filtered[\"IH_ALT_CTS\"]\n",
    ")\n",
    "\n",
    "ih_allele_counts_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add regional information for each African subpopulation/ethnolinguistic group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add information on the African region (i.e., Southern Africa, Western Africa, Eastern Africa, Central Africa, America, Caribbean) from which a particular African subpopulation/ethnolinguistic group originates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>352.0</td>\n",
       "      <td>314.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs139916479</td>\n",
       "      <td>110149349</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149349_A_G</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs552877576</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149494_T_C</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13:110149646CTTTAT-C</td>\n",
       "      <td>110149646</td>\n",
       "      <td>CTTTAT</td>\n",
       "      <td>C</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>2</td>\n",
       "      <td>352</td>\n",
       "      <td>110149646_C_CTTTAT</td>\n",
       "      <td>352.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs13260</td>\n",
       "      <td>110149776</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>71</td>\n",
       "      <td>352</td>\n",
       "      <td>110149776_T_G</td>\n",
       "      <td>352.0</td>\n",
       "      <td>281.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  VAR_NAME        POS     REF ALT    GENE   SUB_POP  \\\n",
       "0               rs59409892  110148917       C   G  COL4A1  Mandinka   \n",
       "1              rs139916479  110149349       G   A  COL4A1  Mandinka   \n",
       "2              rs552877576  110149494       C   T  COL4A1  Mandinka   \n",
       "3  chr13:110149646CTTTAT-C  110149646  CTTTAT   C  COL4A1  Mandinka   \n",
       "4                  rs13260  110149776       G   T  COL4A1  Mandinka   \n",
       "\n",
       "   IH_ALT_CTS  IH_TOTAL_CTS                  ID  CORR_IH_TOTAL_CTS  \\\n",
       "0          38           352       110148917_G_C              352.0   \n",
       "1           2           352       110149349_A_G              352.0   \n",
       "2           2           352       110149494_T_C              352.0   \n",
       "3           2           352  110149646_C_CTTTAT              352.0   \n",
       "4          71           352       110149776_T_G              352.0   \n",
       "\n",
       "   CORR_IH_REF_CTS REG  \n",
       "0            314.0  WA  \n",
       "1            350.0  WA  \n",
       "2            350.0  WA  \n",
       "3            350.0  WA  \n",
       "4            281.0  WA  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts_filtered[\"REG\"] = ih_allele_counts_filtered[\"SUB_POP\"].map(\n",
    "    constants.REGIONAL_CLASSIFICATION\n",
    ")\n",
    "\n",
    "ih_allele_counts_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add grouped African count information\n",
    "\n",
    "Provide aggregated allele count information for Recent African populations. Recent African populations are defined as African populations currently residing on the African continent. This group excludes African American and African Caribbean populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>1608.0</td>\n",
       "      <td>1606.0</td>\n",
       "      <td>Recent African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>1000G Yoruba</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>38</td>\n",
       "      <td>352</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>352.0</td>\n",
       "      <td>314.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>152</td>\n",
       "      <td>1596</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>1608.0</td>\n",
       "      <td>1456.0</td>\n",
       "      <td>Recent African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>13</td>\n",
       "      <td>194</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>194.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>EA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      VAR_NAME        POS REF ALT    GENE       SUB_POP  IH_ALT_CTS  \\\n",
       "0  rs552586867  110148891   C   G  COL4A1           NaN           2   \n",
       "1  rs552586867  110148891   C   G  COL4A1  1000G Yoruba           2   \n",
       "2   rs59409892  110148917   C   G  COL4A1      Mandinka          38   \n",
       "3   rs59409892  110148917   C   G  COL4A1           NaN         152   \n",
       "4   rs59409892  110148917   C   G  COL4A1         Luhya          13   \n",
       "\n",
       "   IH_TOTAL_CTS             ID  CORR_IH_TOTAL_CTS  CORR_IH_REF_CTS  \\\n",
       "0           350  110148891_G_C             1608.0           1606.0   \n",
       "1           350  110148891_G_C                NaN              NaN   \n",
       "2           352  110148917_G_C              352.0            314.0   \n",
       "3          1596  110148917_G_C             1608.0           1456.0   \n",
       "4           194  110148917_G_C              194.0            181.0   \n",
       "\n",
       "              REG  \n",
       "0  Recent African  \n",
       "1             NaN  \n",
       "2              WA  \n",
       "3  Recent African  \n",
       "4              EA  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate Recent African allele count data using a custom function\n",
    "\n",
    "recent_africa_ct = functions.grouped_pop_allele_counts(\n",
    "    ih_allele_counts_filtered.copy(), \"Recent African\"\n",
    ")\n",
    "\n",
    "# Calculate Recent African CORR_IH_TOTAL_CTS and CORR_IH_REF_CTS. The aggregated values for these columns will not be correct.\n",
    "\n",
    "recent_africa_ct[\"CORR_IH_TOTAL_CTS\"] = (\n",
    "    grouped_sample_subpopulations.SAMPLE_COUNT.sum() * 2\n",
    ")\n",
    "recent_africa_ct[\"CORR_IH_REF_CTS\"] = (\n",
    "    recent_africa_ct[\"CORR_IH_TOTAL_CTS\"] - recent_africa_ct[\"IH_ALT_CTS\"]\n",
    ")\n",
    "\n",
    "\n",
    "# Concatenate the Recent African allele count data with the subpopulation allele count data\n",
    "\n",
    "ih_allele_counts_grouped = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            ih_allele_counts_filtered,\n",
    "            recent_africa_ct,\n",
    "        ]\n",
    "    )\n",
    "    .sort_values(\"ID\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "ih_allele_counts_grouped.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aggregate allele counts for the 1000G and HGDP Yoruban population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>1608.0</td>\n",
       "      <td>1606.0</td>\n",
       "      <td>Recent African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>392.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>40</td>\n",
       "      <td>392</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>392.0</td>\n",
       "      <td>352.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mende</td>\n",
       "      <td>18</td>\n",
       "      <td>196</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>196.0</td>\n",
       "      <td>178.0</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Bantu South Africa</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>16.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>SA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      VAR_NAME        POS REF ALT    GENE             SUB_POP  IH_ALT_CTS  \\\n",
       "0  rs552586867  110148891   C   G  COL4A1                 NaN           2   \n",
       "1  rs552586867  110148891   C   G  COL4A1              Yoruba           2   \n",
       "3   rs59409892  110148917   C   G  COL4A1              Yoruba          40   \n",
       "4   rs59409892  110148917   C   G  COL4A1               Mende          18   \n",
       "6   rs59409892  110148917   C   G  COL4A1  Bantu South Africa           2   \n",
       "\n",
       "   IH_TOTAL_CTS             ID  CORR_IH_TOTAL_CTS  CORR_IH_REF_CTS  \\\n",
       "0           350  110148891_G_C             1608.0           1606.0   \n",
       "1           350  110148891_G_C              392.0            390.0   \n",
       "3           392  110148917_G_C              392.0            352.0   \n",
       "4           196  110148917_G_C              196.0            178.0   \n",
       "6            16  110148917_G_C               16.0             14.0   \n",
       "\n",
       "              REG  \n",
       "0  Recent African  \n",
       "1             NaN  \n",
       "3             NaN  \n",
       "4              WA  \n",
       "6              SA  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate 1000G and HGDP Yoruban allele count data using a custom function\n",
    "\n",
    "subpop_filters = (ih_allele_counts_grouped[\"SUB_POP\"] == \"1000G Yoruba\") | (\n",
    "    ih_allele_counts_grouped[\"SUB_POP\"] == \"HGDP Yoruba\"\n",
    ")\n",
    "\n",
    "yoruban_ct = functions.grouped_pop_allele_counts(\n",
    "    ih_allele_counts_grouped.copy(), \"Yoruba\", subpop_filters=subpop_filters\n",
    ")\n",
    "\n",
    "# Calculate the corrected total and reference counts for the Yoruban population. The aggregated values for these columns will not be correct.\n",
    "\n",
    "yoruban_ct[\"CORR_IH_TOTAL_CTS\"] = (\n",
    "    sample_subpopulations[sample_subpopulations[\"SUB\"] == \"Yoruba\"][\n",
    "        \"SAMPLE_NAME\"\n",
    "    ].count()\n",
    "    * 2\n",
    ")\n",
    "\n",
    "yoruban_ct[\"CORR_IH_REF_CTS\"] = (\n",
    "    yoruban_ct[\"CORR_IH_TOTAL_CTS\"] - yoruban_ct[\"IH_ALT_CTS\"]\n",
    ")\n",
    "\n",
    "# Concatenate the Aggregated Yoruban allele count data with the subpopulation allele count data\n",
    "\n",
    "ih_allele_counts_grouped = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            ih_allele_counts_grouped,\n",
    "            yoruban_ct,\n",
    "        ]\n",
    "    )\n",
    "    .sort_values(\"ID\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Replace the 1000G and HGDP Yoruban allele count data with the aggregated Yoruban allele count data\n",
    "\n",
    "ih_allele_counts_grouped = ih_allele_counts_grouped[\n",
    "    (\n",
    "        (ih_allele_counts_grouped[\"SUB_POP\"] != \"1000G Yoruba\")\n",
    "        & (ih_allele_counts_grouped[\"SUB_POP\"] != \"HGDP Yoruba\")\n",
    "    )\n",
    "]\n",
    "\n",
    "ih_allele_counts_grouped.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculate allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>CORR_IH_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>1608.0</td>\n",
       "      <td>1606.0</td>\n",
       "      <td>Recent African</td>\n",
       "      <td>0.001244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>2</td>\n",
       "      <td>350</td>\n",
       "      <td>110148891_G_C</td>\n",
       "      <td>392.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.005102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>40</td>\n",
       "      <td>392</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>392.0</td>\n",
       "      <td>352.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.102041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Mende</td>\n",
       "      <td>18</td>\n",
       "      <td>196</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>196.0</td>\n",
       "      <td>178.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.091837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>Bantu South Africa</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>110148917_G_C</td>\n",
       "      <td>16.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>SA</td>\n",
       "      <td>0.125000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      VAR_NAME        POS REF ALT    GENE             SUB_POP  IH_ALT_CTS  \\\n",
       "0  rs552586867  110148891   C   G  COL4A1                 NaN           2   \n",
       "1  rs552586867  110148891   C   G  COL4A1              Yoruba           2   \n",
       "3   rs59409892  110148917   C   G  COL4A1              Yoruba          40   \n",
       "4   rs59409892  110148917   C   G  COL4A1               Mende          18   \n",
       "6   rs59409892  110148917   C   G  COL4A1  Bantu South Africa           2   \n",
       "\n",
       "   IH_TOTAL_CTS             ID  CORR_IH_TOTAL_CTS  CORR_IH_REF_CTS  \\\n",
       "0           350  110148891_G_C             1608.0           1606.0   \n",
       "1           350  110148891_G_C              392.0            390.0   \n",
       "3           392  110148917_G_C              392.0            352.0   \n",
       "4           196  110148917_G_C              196.0            178.0   \n",
       "6            16  110148917_G_C               16.0             14.0   \n",
       "\n",
       "              REG  CORR_IH_AF  \n",
       "0  Recent African    0.001244  \n",
       "1             NaN    0.005102  \n",
       "3             NaN    0.102041  \n",
       "4              WA    0.091837  \n",
       "6              SA    0.125000  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts_grouped[\"CORR_IH_AF\"] = (\n",
    "    ih_allele_counts_grouped[\"IH_ALT_CTS\"]\n",
    "    / ih_allele_counts_grouped[\"CORR_IH_TOTAL_CTS\"]\n",
    ")\n",
    "\n",
    "ih_allele_counts_grouped.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Classify variants as SNPs or INDELs\n",
    "\n",
    "Variants can be classified as:\n",
    "\n",
    "* Single nucleotide polymorphisms (SNPs) if they result in the exchange of one DNA base pair for another \n",
    "\n",
    "**OR**\n",
    "\n",
    "* Insertion/deletions (INDELs) if they result in the deletion or insertion of one or more DNA base pairs\n",
    "\n",
    "Adding these variant classifications may prove useful for further analysis of the variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>IH_ALT_CTS</th>\n",
       "      <th>IH_TOTAL_CTS</th>\n",
       "      <th>ID</th>\n",
       "      <th>CORR_IH_TOTAL_CTS</th>\n",
       "      <th>CORR_IH_REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "      <th>CORR_IH_AF</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60249</th>\n",
       "      <td>rs143171553</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>3</td>\n",
       "      <td>352</td>\n",
       "      <td>48256362_A_G</td>\n",
       "      <td>352.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.008523</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60250</th>\n",
       "      <td>rs143171553</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>Bantu Kenya</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>48256362_A_G</td>\n",
       "      <td>24.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>EA</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60251</th>\n",
       "      <td>rs143171553</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>Mandenka</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>48256362_A_G</td>\n",
       "      <td>46.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60253</th>\n",
       "      <td>rs143171553</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>Mende</td>\n",
       "      <td>8</td>\n",
       "      <td>194</td>\n",
       "      <td>48256362_A_G</td>\n",
       "      <td>196.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>WA</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60254</th>\n",
       "      <td>rs143171553</td>\n",
       "      <td>48256362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CARD8</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>3</td>\n",
       "      <td>350</td>\n",
       "      <td>48256362_A_G</td>\n",
       "      <td>392.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007653</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          VAR_NAME       POS REF ALT   GENE      SUB_POP  IH_ALT_CTS  \\\n",
       "60249  rs143171553  48256362   G   A  CARD8     Mandinka           3   \n",
       "60250  rs143171553  48256362   G   A  CARD8  Bantu Kenya           1   \n",
       "60251  rs143171553  48256362   G   A  CARD8     Mandenka           1   \n",
       "60253  rs143171553  48256362   G   A  CARD8        Mende           8   \n",
       "60254  rs143171553  48256362   G   A  CARD8       Yoruba           3   \n",
       "\n",
       "       IH_TOTAL_CTS            ID  CORR_IH_TOTAL_CTS  CORR_IH_REF_CTS  REG  \\\n",
       "60249           352  48256362_A_G              352.0            349.0   WA   \n",
       "60250            22  48256362_A_G               24.0             23.0   EA   \n",
       "60251            42  48256362_A_G               46.0             45.0   WA   \n",
       "60253           194  48256362_A_G              196.0            188.0   WA   \n",
       "60254           350  48256362_A_G              392.0            389.0  NaN   \n",
       "\n",
       "       CORR_IH_AF VARIANT_TYPE  \n",
       "60249    0.008523          SNP  \n",
       "60250    0.041667          SNP  \n",
       "60251    0.021739          SNP  \n",
       "60253    0.040816          SNP  \n",
       "60254    0.007653          SNP  "
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_allele_counts_grouped[\"VARIANT_TYPE\"] = np.where(\n",
    "    (ih_allele_counts_grouped.ALT.str.len() > 1)\n",
    "    | (ih_allele_counts_grouped.REF.str.len() > 1),\n",
    "    \"INDEL\",\n",
    "    \"SNP\",\n",
    ")\n",
    "\n",
    "ih_allele_counts_grouped.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save African in-house allele count data to a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_allele_counts_grouped.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts.csv\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare data in suitable format for Fisher's Tests\n",
    "\n",
    "The in-house allele count data needs to be formatted differently to conduct Fisher's Tests to test for significant differences in allele frequency between population regions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>SUB</th>\n",
       "      <th>SAMPLE_NAME</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>HG02461</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>HG02462</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>HG02463</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>HG02464</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Mandinka</td>\n",
       "      <td>HG02465</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>799</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>NA19475</td>\n",
       "      <td>EA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>800</td>\n",
       "      <td>Mandenka</td>\n",
       "      <td>SS6004470</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>801</td>\n",
       "      <td>Mbuti Pygmy</td>\n",
       "      <td>SS6004471</td>\n",
       "      <td>CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>802</td>\n",
       "      <td>San</td>\n",
       "      <td>SS6004473</td>\n",
       "      <td>SA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>803</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>SS6004475</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>804 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0          SUB SAMPLE_NAME REG\n",
       "0             0     Mandinka     HG02461  WA\n",
       "1             1     Mandinka     HG02462  WA\n",
       "2             2     Mandinka     HG02463  WA\n",
       "3             3     Mandinka     HG02464  WA\n",
       "4             4     Mandinka     HG02465  WA\n",
       "..          ...          ...         ...  ..\n",
       "799         799        Luhya     NA19475  EA\n",
       "800         800     Mandenka   SS6004470  WA\n",
       "801         801  Mbuti Pygmy   SS6004471  CA\n",
       "802         802          San   SS6004473  SA\n",
       "803         803       Yoruba   SS6004475  WA\n",
       "\n",
       "[804 rows x 4 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_subpopulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_subpopulations[sample_subpopulations[\"REG\"] == \"CA\"][\"SAMPLE_NAME\"].count() * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>REG</th>\n",
       "      <th>VAR_NAME</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th>ALT_CT_IH_CA</th>\n",
       "      <th>ALT_CT_IH_EA</th>\n",
       "      <th>ALT_CT_IH_Recent African</th>\n",
       "      <th>ALT_CT_IH_SA</th>\n",
       "      <th>ALT_CT_IH_WA</th>\n",
       "      <th>CORR_REF_CT_IH_CA</th>\n",
       "      <th>CORR_REF_CT_IH_EA</th>\n",
       "      <th>CORR_REF_CT_IH_Recent African</th>\n",
       "      <th>CORR_REF_CT_IH_SA</th>\n",
       "      <th>CORR_REF_CT_IH_WA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:34438836T-C</td>\n",
       "      <td>34438836_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11:34438889G-C</td>\n",
       "      <td>34438889_C_G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34438889</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:34439223G-C</td>\n",
       "      <td>34439223_C_G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1606.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1280.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:34439262C-T</td>\n",
       "      <td>34439262_T_C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:34439312G-A</td>\n",
       "      <td>34439312_A_G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CAT</td>\n",
       "      <td>34439312</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1607.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "REG           VAR_NAME            ID REF ALT GENE       POS  ALT_CT_IH_CA  \\\n",
       "0    chr11:34438836T-C  34438836_C_T   T   C  CAT  34438836           0.0   \n",
       "1    chr11:34438889G-C  34438889_C_G   G   C  CAT  34438889           0.0   \n",
       "2    chr11:34439223G-C  34439223_C_G   G   C  CAT  34439223           0.0   \n",
       "3    chr11:34439262C-T  34439262_T_C   C   T  CAT  34439262           0.0   \n",
       "4    chr11:34439312G-A  34439312_A_G   G   A  CAT  34439312           0.0   \n",
       "\n",
       "REG  ALT_CT_IH_EA  ALT_CT_IH_Recent African  ALT_CT_IH_SA  ALT_CT_IH_WA  \\\n",
       "0             1.0                       1.0           0.0           0.0   \n",
       "1             0.0                       1.0           1.0           0.0   \n",
       "2             0.0                       2.0           0.0           2.0   \n",
       "3             0.0                       1.0           0.0           0.0   \n",
       "4             0.0                       1.0           1.0           0.0   \n",
       "\n",
       "REG  CORR_REF_CT_IH_CA  CORR_REF_CT_IH_EA  CORR_REF_CT_IH_Recent African  \\\n",
       "0                  0.0              217.0                         1607.0   \n",
       "1                  0.0                0.0                         1607.0   \n",
       "2                  0.0                0.0                         1606.0   \n",
       "3                  0.0                0.0                         1607.0   \n",
       "4                  0.0                0.0                         1607.0   \n",
       "\n",
       "REG  CORR_REF_CT_IH_SA  CORR_REF_CT_IH_WA  \n",
       "0                  0.0                0.0  \n",
       "1                 27.0                0.0  \n",
       "2                  0.0             1280.0  \n",
       "3                  0.0                0.0  \n",
       "4                 27.0                0.0  "
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate a list of all unique regions in ih_allele_counts_grouped dataframe\n",
    "inhouse_populations = [region for region in set(ih_allele_counts_grouped.REG.values)]\n",
    "\n",
    "# Aggregate allele counts by region\n",
    "\n",
    "ih_allele_counts_regions = (\n",
    "    ih_allele_counts_grouped.groupby(\n",
    "        by=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\", \"REG\"]\n",
    "    )\n",
    "    .sum(numeric_only=True)\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Pivot data\n",
    "ih_allele_counts_pivot = ih_allele_counts_regions.pivot(\n",
    "    index=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\"],\n",
    "    columns=\"REG\",\n",
    "    values=[\"IH_ALT_CTS\", \"CORR_IH_REF_CTS\"],\n",
    ")\n",
    "\n",
    "# Separate alternate and reference count data into different dataframes to facilate renaming of count columns appropriately\n",
    "ih_data_alt = (\n",
    "    ih_allele_counts_pivot[[\"IH_ALT_CTS\"]].droplevel(level=0, axis=1).reset_index()\n",
    ")\n",
    "\n",
    "ih_data_corr_ref = (\n",
    "    ih_allele_counts_pivot[[\"CORR_IH_REF_CTS\"]].droplevel(level=0, axis=1).reset_index()\n",
    ")\n",
    "\n",
    "# Add appropriate prefixes to alt and ref columns\n",
    "ih_data_alt = functions.add_prefix_dataframe_col_names(\n",
    "    ih_data_alt, inhouse_populations, \"ALT_CT_IH_\"\n",
    ")\n",
    "\n",
    "ih_data_corr_ref = functions.add_prefix_dataframe_col_names(\n",
    "    ih_data_corr_ref, inhouse_populations, \"CORR_REF_CT_IH_\"\n",
    ")\n",
    "\n",
    "# Merge renamed alternate and reference count data\n",
    "ih_recent_afr = ih_data_alt.merge(\n",
    "    ih_data_corr_ref, on=[\"VAR_NAME\", \"ID\", \"REF\", \"ALT\", \"GENE\", \"POS\"]\n",
    ")\n",
    "\n",
    "# Calculate the corrected reference counts for each region. The aggregated values for these columns will not be correct.\n",
    "\n",
    "ih_recent_afr[\"CORR_REF_CT_IH_CA\"] = (\n",
    "    sample_subpopulations[sample_subpopulations[\"REG\"] == \"CA\"][\"SAMPLE_NAME\"].count()\n",
    "    * 2\n",
    ") - ih_recent_afr[\"ALT_CT_IH_CA\"]\n",
    "\n",
    "ih_recent_afr[\"CORR_REF_CT_IH_SA\"] = (\n",
    "    sample_subpopulations[sample_subpopulations[\"REG\"] == \"SA\"][\"SAMPLE_NAME\"].count()\n",
    "    * 2\n",
    ") - ih_recent_afr[\"ALT_CT_IH_SA\"]\n",
    "\n",
    "ih_recent_afr[\"CORR_REF_CT_IH_EA\"] = (\n",
    "    sample_subpopulations[sample_subpopulations[\"REG\"] == \"EA\"][\"SAMPLE_NAME\"].count()\n",
    "    * 2\n",
    ") - ih_recent_afr[\"ALT_CT_IH_EA\"]\n",
    "\n",
    "ih_recent_afr[\"CORR_REF_CT_IH_WA\"] = (\n",
    "    sample_subpopulations[sample_subpopulations[\"REG\"] == \"WA\"][\"SAMPLE_NAME\"].count()\n",
    "    * 2\n",
    ") - ih_recent_afr[\"ALT_CT_IH_WA\"]\n",
    "\n",
    "# Replace missing count values with 0\n",
    "\n",
    "ih_recent_afr = ih_recent_afr.replace(np.NAN, 0)\n",
    "\n",
    "ih_recent_afr.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Fisher's Test data to a csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "ih_recent_afr.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"IH_allele_counts_fishers.csv\",\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
