{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of variant consequences\n",
    "\n",
    "Genetic variants can be classified by the consequences that they have on the function of a gene. \n",
    "\n",
    "The consequences of a genetic variant can be broadly classified into several categories:\n",
    "\n",
    "* Synonymous: A genetic variant that does not result in a protein amino acid change. \n",
    "* Non-synonymous: A genetic variant that changes an amino acid in a protein. \n",
    "* Upstream: An upstream gene variant refers to a genetic change or alteration that occurs in the DNA sequence located before (or \"upstream\" of) a particular gene. Upstream variants can potentially affect the regulation or expression of the gene by influencing how the gene is transcribed or controlled.\n",
    "* Downstream: A downstream gene variant occurs in the DNA sequence located after (or \"downstream\" of) a specific gene. Downstream variants might impact processes related to the gene's transcript processing, translation, or overall function.\n",
    "* Intronic: Intronic variants are located in regions that do not directly encode genes. These variants may impact the splicing process of genes.\n",
    "* Regulatory: A genetic variant located in an intronic region that interferes with gene regulatory elements.\n",
    "* Splice site: A genetic variant within a site where genetic splicing takes place. \n",
    "* 3-prime/5-prime UTR: These variants are located before (3-prime) and after (5-prime) gene coding regions and may impact various gene regulatory functions.\n",
    "\n",
    "Consequence data for variants identified in-house in African population groups was retrieved from the [Ensembl Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) using [CADD v1.6](https://cadd.gs.washington.edu/score). The retrieved data was stored in `Data/VEP/GRCh38-v1.6_{gene_name}.tsv, where `{gene_name}` refers to the name of a specific gene. \n",
    "\n",
    "To prepare the data for further analysis, the following steps were performed: \n",
    "\n",
    "1. The consequence data for all genes was merged into a single dataset\n",
    "2. Only the relevant features, such as chromosome, reference and alternate allele information, variant position information, and variant consequences, were selected. These features were renamed if necessary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import pandas as pd\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant consequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_19232\\193486938.py:22: DtypeWarning: Columns (16,17,23,30,32,35,37) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  consequence_df = pd.read_csv(gene_consequence_path, sep=\"\\t\", skiprows=[0])\n",
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_19232\\193486938.py:22: DtypeWarning: Columns (13,16,17,30,35,37) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  consequence_df = pd.read_csv(gene_consequence_path, sep=\"\\t\", skiprows=[0])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Type</th>\n",
       "      <th>Length</th>\n",
       "      <th>AnnoType</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>ConsScore</th>\n",
       "      <th>ConsDetail</th>\n",
       "      <th>...</th>\n",
       "      <th>Rare10000bp</th>\n",
       "      <th>Sngl10000bp</th>\n",
       "      <th>EnsembleRegulatoryFeature</th>\n",
       "      <th>dbscSNV-ada_score</th>\n",
       "      <th>dbscSNV-rf_score</th>\n",
       "      <th>RemapOverlapTF</th>\n",
       "      <th>RemapOverlapCL</th>\n",
       "      <th>RawScore</th>\n",
       "      <th>PHRED</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>INS</td>\n",
       "      <td>1</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DOWNSTREAM</td>\n",
       "      <td>1</td>\n",
       "      <td>downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.437825</td>\n",
       "      <td>0.160</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>0</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DOWNSTREAM</td>\n",
       "      <td>1</td>\n",
       "      <td>downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.227221</td>\n",
       "      <td>0.446</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>0</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DOWNSTREAM</td>\n",
       "      <td>1</td>\n",
       "      <td>downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.269936</td>\n",
       "      <td>3.938</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>0</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DOWNSTREAM</td>\n",
       "      <td>1</td>\n",
       "      <td>downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.530972</td>\n",
       "      <td>6.825</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>0</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DOWNSTREAM</td>\n",
       "      <td>1</td>\n",
       "      <td>downstream</td>\n",
       "      <td>...</td>\n",
       "      <td>92</td>\n",
       "      <td>1315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.380228</td>\n",
       "      <td>14.950</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #Chrom        Pos Ref Alt Type  Length    AnnoType Consequence  ConsScore  \\\n",
       "0      13  110148882   C  CT  INS       1  Intergenic  DOWNSTREAM          1   \n",
       "1      13  110148891   C   G  SNV       0  Intergenic  DOWNSTREAM          1   \n",
       "2      13  110148917   C   G  SNV       0  Intergenic  DOWNSTREAM          1   \n",
       "3      13  110148920   G   C  SNV       0  Intergenic  DOWNSTREAM          1   \n",
       "4      13  110148959   A   G  SNV       0  Intergenic  DOWNSTREAM          1   \n",
       "\n",
       "   ConsDetail  ...  Rare10000bp  Sngl10000bp  EnsembleRegulatoryFeature  \\\n",
       "0  downstream  ...           91         1311                        NaN   \n",
       "1  downstream  ...           91         1314                        NaN   \n",
       "2  downstream  ...           91         1312                        NaN   \n",
       "3  downstream  ...           91         1312                        NaN   \n",
       "4  downstream  ...           92         1315                        NaN   \n",
       "\n",
       "  dbscSNV-ada_score  dbscSNV-rf_score  RemapOverlapTF RemapOverlapCL  \\\n",
       "0               NaN               NaN             NaN            NaN   \n",
       "1               NaN               NaN             NaN            NaN   \n",
       "2               NaN               NaN             NaN            NaN   \n",
       "3               NaN               NaN             NaN            NaN   \n",
       "4               NaN               NaN             NaN            NaN   \n",
       "\n",
       "   RawScore   PHRED    GENE  \n",
       "0 -0.437825   0.160  COL4A1  \n",
       "1 -0.227221   0.446  COL4A1  \n",
       "2  0.269936   3.938  COL4A1  \n",
       "3  0.530972   6.825  COL4A1  \n",
       "4  1.380228  14.950  COL4A1  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_consequence_data = pd.DataFrame()\n",
    "\n",
    "genes = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Metadata\",\n",
    "        \"locations.csv\",\n",
    "    )\n",
    ").location_name\n",
    "\n",
    "for gene in genes:\n",
    "    gene_consequence_path = os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"VEP\",\n",
    "        \"GRCh38-v1.6_{}.tsv\".format(gene),\n",
    "    )\n",
    "\n",
    "    consequence_df = pd.DataFrame()\n",
    "    if os.path.exists(gene_consequence_path):\n",
    "        consequence_df = pd.read_csv(gene_consequence_path, sep=\"\\t\", skiprows=[0])\n",
    "        consequence_df[\"GENE\"] = gene\n",
    "\n",
    "    gene_consequence_data = pd.concat([gene_consequence_data, consequence_df])\n",
    "\n",
    "gene_consequence_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the chromosome, position, reference allele, alternate allele, gene and consequence features for further analysis. Rename features if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>CONSEQUENCE_CLASSIFICATION</th>\n",
       "      <th>CONSEQUENCE</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>INS</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT TYPE CONSEQUENCE_CLASSIFICATION CONSEQUENCE  \\\n",
       "0     13  110148882   C  CT  INS                 downstream  downstream   \n",
       "1     13  110148891   C   G  SNV                 downstream  downstream   \n",
       "2     13  110148917   C   G  SNV                 downstream  downstream   \n",
       "3     13  110148920   G   C  SNV                 downstream  downstream   \n",
       "4     13  110148959   A   G  SNV                 downstream  downstream   \n",
       "\n",
       "     GENE  \n",
       "0  COL4A1  \n",
       "1  COL4A1  \n",
       "2  COL4A1  \n",
       "3  COL4A1  \n",
       "4  COL4A1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_consequence_data_filtered = gene_consequence_data.copy()[\n",
    "    [\"#Chrom\", \"Pos\", \"Ref\", \"Alt\", \"Type\", \"Consequence\", \"ConsDetail\", \"GeneName\"]\n",
    "].rename(\n",
    "    columns={\n",
    "        \"#Chrom\": \"CHROM\",\n",
    "        \"Pos\": \"POS\",\n",
    "        \"Ref\": \"REF\",\n",
    "        \"Alt\": \"ALT\",\n",
    "        \"Type\": \"TYPE\",\n",
    "        \"Consequence\": \"CONSEQUENCE_CLASSIFICATION\",\n",
    "        \"ConsDetail\": \"CONSEQUENCE\",\n",
    "        \"GeneName\": \"GENE\",\n",
    "    }\n",
    ")\n",
    "gene_consequence_data_filtered[\"CONSEQUENCE_CLASSIFICATION\"] = gene_consequence_data_filtered[\n",
    "    \"CONSEQUENCE_CLASSIFICATION\"\n",
    "].apply(str.lower)\n",
    "gene_consequence_data_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign a unique ID to each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add a column with a unique ID for each variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>CONSEQUENCE_CLASSIFICATION</th>\n",
       "      <th>CONSEQUENCE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>INS</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148882_CT_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148891_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148917_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148920_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>downstream</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>110148959_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT TYPE CONSEQUENCE_CLASSIFICATION CONSEQUENCE  \\\n",
       "0     13  110148882   C  CT  INS                 downstream  downstream   \n",
       "1     13  110148891   C   G  SNV                 downstream  downstream   \n",
       "2     13  110148917   C   G  SNV                 downstream  downstream   \n",
       "3     13  110148920   G   C  SNV                 downstream  downstream   \n",
       "4     13  110148959   A   G  SNV                 downstream  downstream   \n",
       "\n",
       "     GENE              ID  \n",
       "0  COL4A1  110148882_CT_C  \n",
       "1  COL4A1   110148891_G_C  \n",
       "2  COL4A1   110148917_G_C  \n",
       "3  COL4A1   110148920_C_G  \n",
       "4  COL4A1   110148959_G_A  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_consequence_data_filtered[\"ID\"] = (\n",
    "    gene_consequence_data_filtered[[\"POS\", \"ALT\", \"REF\"]].astype(\"str\").agg(\"_\".join, axis=1)\n",
    ")\n",
    "\n",
    "gene_consequence_data_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variant filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove variants that are not associated with the specified genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_consequence_data_filtered = gene_consequence_data_filtered.copy()[\n",
    "    gene_consequence_data_filtered[\"GENE\"].isin(genes)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save consequence data to a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_consequence_data_filtered.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_consequences.csv\",\n",
    "    ),\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
