{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of variant consequence data\n",
    "\n",
    "Genetic variants can be classified by the consequences that they have on the function of a gene. \n",
    "\n",
    "The consequences of a genetic variant can be broadly classified into several categories:\n",
    "\n",
    "* Synonymous: Some genetic variants do not alter the function of the gene or protein, resulting in no observable impact on an individual's health or traits.\n",
    "* Missense: This type of variant leads to the replacement of one amino acid with another in the protein sequence, potentially affecting its structure or function.\n",
    "* Nonsense: These variants result in a premature stop codon in the protein-coding sequence, leading to the production of a truncated and often non-functional protein.\n",
    "* Frameshift: These variants insert or delete genetic material, causing a shift in the reading frame of the gene and disrupting the correct synthesis of the protein.\n",
    "* Splice Site: Variants at the boundary regions of introns and exons can affect the process of splicing, leading to altered mRNA transcripts and, subsequently, abnormal protein production.\n",
    "* Deletion/Duplication: These variants involve the loss or duplication of a segment of DNA, potentially affecting gene function or dosage.\n",
    "\n",
    "Consequence data for variants identified in-house in African population groups from the [GnomAD 1000 Genomes and HGDP datasets](https://gnomad.broadinstitute.org/news/2021-10-gnomad-v3-1-2-minor-release/) was retrieved from the [Ensembl Variant Effect Predictor]() API. The retrieved data was stored in `Data/Raw/SUB/{gene_name}_VEP.csv` and `Data/Raw/SUPER/{gene_name}_VEP.csv`, where `{gene_name}` refers to the name of a specific gene. \n",
    "\n",
    "To prepare the data for further analysis, the following steps were performed: \n",
    "\n",
    "1. The consequence data for all genes was merged into a single dataset\n",
    "2. Only the relevant features, such as variant ID, reference and alternate allele information, variant position information, and variant consequences, were selected. These features were renamed if necessary.\n",
    "3. An additional feature was added to provide a broad classification of variant consequences. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.chdir(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\\Notebooks\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\"\n",
    ")\n",
    "import pandas as pd\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import variant consequence information for each gene and append to a single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Co-Located Variant</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Transcript Strand</th>\n",
       "      <th>Existing Variation</th>\n",
       "      <th>Start Coordinates</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>...</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>input</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>Polyphen_score</th>\n",
       "      <th>Polyphen_pred</th>\n",
       "      <th>CONDEL</th>\n",
       "      <th>CONDEL_pred</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>False</td>\n",
       "      <td>NM_001845.6</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-</td>\n",
       "      <td>110148882</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13:110148882-110148883:-1/CT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001845.6</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.446</td>\n",
       "      <td>13:110148891-110148891:1/G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001845.6</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3.938</td>\n",
       "      <td>13:110148917-110148917:1/G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001845.6</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6.825</td>\n",
       "      <td>13:110148920-110148920:1/C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001845.6</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>14.950</td>\n",
       "      <td>13:110148959-110148959:1/G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID        POS REF ALT  Co-Located Variant Transcript ID  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT               False   NM_001845.6   \n",
       "1          rs552586867  110148891   C   G                True   NM_001845.6   \n",
       "2           rs59409892  110148917   C   G                True   NM_001845.6   \n",
       "3          rs535182970  110148920   G   C                True   NM_001845.6   \n",
       "4           rs56406633  110148959   A   G                True   NM_001845.6   \n",
       "\n",
       "   Transcript Strand Existing Variation  Start Coordinates  \\\n",
       "0               -1.0                  -          110148882   \n",
       "1               -1.0        rs552586867          110148891   \n",
       "2               -1.0         rs59409892          110148917   \n",
       "3               -1.0        rs535182970          110148920   \n",
       "4               -1.0         rs56406633          110148959   \n",
       "\n",
       "               Consequence  ...         Biotype CADD_PHRED  \\\n",
       "0  downstream_gene_variant  ...  protein_coding        NaN   \n",
       "1  downstream_gene_variant  ...  protein_coding      0.446   \n",
       "2  downstream_gene_variant  ...  protein_coding      3.938   \n",
       "3  downstream_gene_variant  ...  protein_coding      6.825   \n",
       "4  downstream_gene_variant  ...  protein_coding     14.950   \n",
       "\n",
       "                          input SIFT_score  SIFT_pred Polyphen_score  \\\n",
       "0  13:110148882-110148883:-1/CT        NaN        NaN            NaN   \n",
       "1    13:110148891-110148891:1/G        NaN        NaN            NaN   \n",
       "2    13:110148917-110148917:1/G        NaN        NaN            NaN   \n",
       "3    13:110148920-110148920:1/C        NaN        NaN            NaN   \n",
       "4    13:110148959-110148959:1/G        NaN        NaN            NaN   \n",
       "\n",
       "   Polyphen_pred CONDEL  CONDEL_pred    GENE  \n",
       "0            NaN    NaN          NaN  COL4A1  \n",
       "1            NaN    NaN          NaN  COL4A1  \n",
       "2            NaN    NaN          NaN  COL4A1  \n",
       "3            NaN    NaN          NaN  COL4A1  \n",
       "4            NaN    NaN          NaN  COL4A1  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_vep_data = pd.DataFrame()\n",
    "for gene in constants.GENES:\n",
    "    gene_vep_path = os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"SUPER\",\n",
    "        \"{}_VEP.csv\".format(gene),\n",
    "    )\n",
    "    consequence_df = pd.DataFrame()\n",
    "    if os.path.exists(gene_vep_path):\n",
    "        consequence_df = pd.read_csv(gene_vep_path, sep=\",\")\n",
    "        consequence_df[\"GENE\"] = gene\n",
    "    gene_vep_data = pd.concat([gene_vep_data, consequence_df]).drop(columns=\"Unnamed: 0\")\n",
    "\n",
    "gene_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the variant ID, position, reference allele, alternate allele, gene and consequence features for further analysis. Rename features if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CONSEQUENCE</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID        POS REF ALT              CONSEQUENCE    GENE\n",
       "0  chr13:110148882C-CT  110148882   C  CT  downstream_gene_variant  COL4A1\n",
       "1          rs552586867  110148891   C   G  downstream_gene_variant  COL4A1\n",
       "2           rs59409892  110148917   C   G  downstream_gene_variant  COL4A1\n",
       "3          rs535182970  110148920   G   C  downstream_gene_variant  COL4A1\n",
       "4           rs56406633  110148959   A   G  downstream_gene_variant  COL4A1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_vep_data = gene_vep_data.copy()[[\"ID\", \"POS\", \"REF\", \"ALT\", \"Consequence\", \"GENE\"]].rename(columns={\"Consequence\": \"CONSEQUENCE\"})\n",
    "gene_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add a feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variant consequences can be quite diverse and specific, describing various effects on gene function or protein structure. Some of the consequences can often be grouped or classified into broader categories for easier interpretation and analysis. An additional feature was added to provide a broad classification of variant consequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CONSEQUENCE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>CONSEQUENCE_CLASSIFICATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13:110148882C-CT</td>\n",
       "      <td>110148882</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs552586867</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs59409892</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs535182970</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs56406633</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>downstream</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID        POS REF ALT              CONSEQUENCE    GENE  \\\n",
       "0  chr13:110148882C-CT  110148882   C  CT  downstream_gene_variant  COL4A1   \n",
       "1          rs552586867  110148891   C   G  downstream_gene_variant  COL4A1   \n",
       "2           rs59409892  110148917   C   G  downstream_gene_variant  COL4A1   \n",
       "3          rs535182970  110148920   G   C  downstream_gene_variant  COL4A1   \n",
       "4           rs56406633  110148959   A   G  downstream_gene_variant  COL4A1   \n",
       "\n",
       "  CONSEQUENCE_CLASSIFICATION  \n",
       "0                 downstream  \n",
       "1                 downstream  \n",
       "2                 downstream  \n",
       "3                 downstream  \n",
       "4                 downstream  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_vep_data[\"CONSEQUENCE_CLASSIFICATION\"] = gene_vep_data[\"CONSEQUENCE\"].map(\n",
    "    constants.VARIANT_CLASSIFICATION\n",
    ")\n",
    "\n",
    "gene_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save consequence data to a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_vep_data.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_consequences.csv\",\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
