{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of variant effect score data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variant effect scores are numerical values or metrics used to quantitatively assess the potential functional impact of genetic variants on genes or proteins. These scores are based on computational predictions which estimate how likely a genetic variant is to have a deleterious (pathogenic) effect.\n",
    "\n",
    "Genetic variant effect score and prediction data were retrieved for the variants identified in-house from [PredictSNP2](https://loschmidt.chemi.muni.cz/predictsnp2/). The retrieved data is stored at `Data/Raw/PredictSNP2`.\n",
    "\n",
    "The data was prepared for analysis by: \n",
    "1. Selecting features of interest, such as variant ID, position, reference and alternate allele, and consequence scores and predictions\n",
    "2. Adding additional features including information on the gene in which a particular variant was found"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.chdir(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\\Notebooks\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    r\"C:\\Users\\User\\Desktop\\Megan\\MSC2\\Results\\5._Posthoc_analysis\\Pipeline_GnomAD_14032023\"\n",
    ")\n",
    "import pandas as pd\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import PredictSNP2 variant effect data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CONS</th>\n",
       "      <th>EXON</th>\n",
       "      <th>SGNF</th>\n",
       "      <th>DBSNP</th>\n",
       "      <th>HAPLO</th>\n",
       "      <th>...</th>\n",
       "      <th>DANNC</th>\n",
       "      <th>FATE</th>\n",
       "      <th>FATS</th>\n",
       "      <th>FATC</th>\n",
       "      <th>FUNE</th>\n",
       "      <th>FUNS</th>\n",
       "      <th>FUNC</th>\n",
       "      <th>GWAVAE</th>\n",
       "      <th>GWAVAS</th>\n",
       "      <th>GWAVAC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>0.82</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.18997</td>\n",
       "      <td>0.77</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>downstream</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?se...</td>\n",
       "      <td>http://www.broadinstitute.org/mammals/haploreg...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.18963</td>\n",
       "      <td>0.77</td>\n",
       "      <td>neutral</td>\n",
       "      <td>8.51599164992024E-6</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>downstream</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>0.65</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.21184</td>\n",
       "      <td>0.75</td>\n",
       "      <td>?</td>\n",
       "      <td>0.622785975437062</td>\n",
       "      <td>0.49</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>rs56406633</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?se...</td>\n",
       "      <td>http://www.broadinstitute.org/mammals/haploreg...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.71</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.82375</td>\n",
       "      <td>0.82</td>\n",
       "      <td>neutral</td>\n",
       "      <td>7.52439751165574E-20</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148971</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>0.65</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.95264</td>\n",
       "      <td>0.91</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>1.19368630042138</td>\n",
       "      <td>0.74</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS          ID REF ALT        CONS EXON           SGNF  \\\n",
       "0     13  110148891           .   C   G  downstream  NaN              .   \n",
       "1     13  110148917  rs59409892   C   G  downstream  NaN              .   \n",
       "2     13  110148920           .   G   C  downstream  NaN              .   \n",
       "3     13  110148959  rs56406633   A   G        UTR3  NaN  Likely benign   \n",
       "4     13  110148971           .   G   C        UTR3  NaN              .   \n",
       "\n",
       "                                               DBSNP  \\\n",
       "0                                                  .   \n",
       "1  http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?se...   \n",
       "2                                                  .   \n",
       "3  http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?se...   \n",
       "4                                                  .   \n",
       "\n",
       "                                               HAPLO  ... DANNC         FATE  \\\n",
       "0                                                  .  ...  0.82      neutral   \n",
       "1  http://www.broadinstitute.org/mammals/haploreg...  ...  0.62      neutral   \n",
       "2                                                  .  ...  0.65      neutral   \n",
       "3  http://www.broadinstitute.org/mammals/haploreg...  ...  0.71  deleterious   \n",
       "4                                                  .  ...  0.65  deleterious   \n",
       "\n",
       "      FATS  FATC         FUNE                  FUNS  FUNC   GWAVAE  GWAVAS  \\\n",
       "0  0.18997  0.77            .                     .     .  neutral    0.30   \n",
       "1  0.18963  0.77      neutral   8.51599164992024E-6  0.80  neutral    0.32   \n",
       "2  0.21184  0.75            ?     0.622785975437062  0.49  neutral    0.32   \n",
       "3  0.82375  0.82      neutral  7.52439751165574E-20  0.80  neutral    0.44   \n",
       "4  0.95264  0.91  deleterious      1.19368630042138  0.74  neutral    0.44   \n",
       "\n",
       "  GWAVAC  \n",
       "0   0.68  \n",
       "1   0.71  \n",
       "2   0.71  \n",
       "3   0.53  \n",
       "4   0.53  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdtsnp2_vep_data = pd.DataFrame()\n",
    "vep_path = os.path.join(\n",
    "    constants.HOME_PATH, \"Data\", \"Raw\", \"PredictSNP2\", \"PredictSNP2.csv\"\n",
    ")\n",
    "if os.path.exists(vep_path):\n",
    "    pdtsnp2_vep_data = pd.read_csv(vep_path, sep=\";\")\n",
    "\n",
    "pdtsnp2_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select features of interest\n",
    "Select variant ID, position, reference and alternate allele, and consequence score and prediction data from various algorithms including FATHMM (FAT), CADD, PredictSNP (PSNP), DANN, FunSeq2 (FUN), and GWAVA. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SGNF</th>\n",
       "      <th>PSNPE</th>\n",
       "      <th>PSNPS</th>\n",
       "      <th>PSNPC</th>\n",
       "      <th>CADDE</th>\n",
       "      <th>...</th>\n",
       "      <th>DANNC</th>\n",
       "      <th>FATE</th>\n",
       "      <th>FATS</th>\n",
       "      <th>FATC</th>\n",
       "      <th>FUNE</th>\n",
       "      <th>FUNS</th>\n",
       "      <th>FUNC</th>\n",
       "      <th>GWAVAE</th>\n",
       "      <th>GWAVAS</th>\n",
       "      <th>GWAVAC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.88</td>\n",
       "      <td>neutral</td>\n",
       "      <td>...</td>\n",
       "      <td>0.82</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.18997</td>\n",
       "      <td>0.77</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.319012</td>\n",
       "      <td>0.74</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.18963</td>\n",
       "      <td>0.77</td>\n",
       "      <td>neutral</td>\n",
       "      <td>8.51599164992024E-6</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.320755</td>\n",
       "      <td>0.74</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>...</td>\n",
       "      <td>0.65</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.21184</td>\n",
       "      <td>0.75</td>\n",
       "      <td>?</td>\n",
       "      <td>0.622785975437062</td>\n",
       "      <td>0.49</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>rs56406633</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.456207</td>\n",
       "      <td>0.91</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>...</td>\n",
       "      <td>0.71</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.82375</td>\n",
       "      <td>0.82</td>\n",
       "      <td>neutral</td>\n",
       "      <td>7.52439751165574E-20</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148971</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.155078</td>\n",
       "      <td>0.73</td>\n",
       "      <td>neutral</td>\n",
       "      <td>...</td>\n",
       "      <td>0.65</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.95264</td>\n",
       "      <td>0.91</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>1.19368630042138</td>\n",
       "      <td>0.74</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS          ID REF ALT           SGNF        PSNPE     PSNPS  \\\n",
       "0     13  110148891           .   C   G              .      neutral -1.000000   \n",
       "1     13  110148917  rs59409892   C   G              .      neutral -0.319012   \n",
       "2     13  110148920           .   G   C              .      neutral -0.320755   \n",
       "3     13  110148959  rs56406633   A   G  Likely benign  deleterious  0.456207   \n",
       "4     13  110148971           .   G   C              .      neutral -0.155078   \n",
       "\n",
       "   PSNPC        CADDE  ...  DANNC         FATE     FATS  FATC         FUNE  \\\n",
       "0   0.88      neutral  ...   0.82      neutral  0.18997  0.77            .   \n",
       "1   0.74  deleterious  ...   0.62      neutral  0.18963  0.77      neutral   \n",
       "2   0.74  deleterious  ...   0.65      neutral  0.21184  0.75            ?   \n",
       "3   0.91  deleterious  ...   0.71  deleterious  0.82375  0.82      neutral   \n",
       "4   0.73      neutral  ...   0.65  deleterious  0.95264  0.91  deleterious   \n",
       "\n",
       "                   FUNS  FUNC   GWAVAE GWAVAS GWAVAC  \n",
       "0                     .     .  neutral   0.30   0.68  \n",
       "1   8.51599164992024E-6  0.80  neutral   0.32   0.71  \n",
       "2     0.622785975437062  0.49  neutral   0.32   0.71  \n",
       "3  7.52439751165574E-20  0.80  neutral   0.44   0.53  \n",
       "4      1.19368630042138  0.74  neutral   0.44   0.53  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdtsnp2_vep_data = pdtsnp2_vep_data.copy()[\n",
    "    [\n",
    "        \"CHROM\",\n",
    "        \"POS\",\n",
    "        \"ID\",\n",
    "        \"REF\",\n",
    "        \"ALT\",\n",
    "        \"SGNF\",\n",
    "        \"PSNPE\",\n",
    "        \"PSNPS\",\n",
    "        \"PSNPC\",\n",
    "        \"CADDE\",\n",
    "        \"CADDS\",\n",
    "        \"CADDC\",\n",
    "        \"DANNE\",\n",
    "        \"DANNS\",\n",
    "        \"DANNC\",\n",
    "        \"FATE\",\n",
    "        \"FATS\",\n",
    "        \"FATC\",\n",
    "        \"FUNE\",\n",
    "        \"FUNS\",\n",
    "        \"FUNC\",\n",
    "        \"GWAVAE\",\n",
    "        \"GWAVAS\",\n",
    "        \"GWAVAC\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "pdtsnp2_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Handle missing values\n",
    "Replace missing values, currently represented with decimal points, with Numpy NaN values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdtsnp2_vep_data = pdtsnp2_vep_data.replace(\".\", np.NaN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add features\n",
    "Add information on the gene in which a particular variant was found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SGNF</th>\n",
       "      <th>PSNPE</th>\n",
       "      <th>PSNPS</th>\n",
       "      <th>PSNPC</th>\n",
       "      <th>CADDE</th>\n",
       "      <th>CADDS</th>\n",
       "      <th>...</th>\n",
       "      <th>FATS</th>\n",
       "      <th>FATC</th>\n",
       "      <th>FUNE</th>\n",
       "      <th>FUNS</th>\n",
       "      <th>FUNC</th>\n",
       "      <th>GWAVAE</th>\n",
       "      <th>GWAVAS</th>\n",
       "      <th>GWAVAC</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.88</td>\n",
       "      <td>neutral</td>\n",
       "      <td>2.258</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18997</td>\n",
       "      <td>0.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.68</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.319012</td>\n",
       "      <td>0.74</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>9.461</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18963</td>\n",
       "      <td>0.77</td>\n",
       "      <td>neutral</td>\n",
       "      <td>8.51599164992024E-6</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>110148920</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.320755</td>\n",
       "      <td>0.74</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>9.779</td>\n",
       "      <td>...</td>\n",
       "      <td>0.21184</td>\n",
       "      <td>0.75</td>\n",
       "      <td>?</td>\n",
       "      <td>0.622785975437062</td>\n",
       "      <td>0.49</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.71</td>\n",
       "      <td>rs535182970</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>110148959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.456207</td>\n",
       "      <td>0.91</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>13.690</td>\n",
       "      <td>...</td>\n",
       "      <td>0.82375</td>\n",
       "      <td>0.82</td>\n",
       "      <td>neutral</td>\n",
       "      <td>7.52439751165574E-20</td>\n",
       "      <td>0.80</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "      <td>rs56406633</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>110148971</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neutral</td>\n",
       "      <td>-0.155078</td>\n",
       "      <td>0.73</td>\n",
       "      <td>neutral</td>\n",
       "      <td>6.450</td>\n",
       "      <td>...</td>\n",
       "      <td>0.95264</td>\n",
       "      <td>0.91</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>1.19368630042138</td>\n",
       "      <td>0.74</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.53</td>\n",
       "      <td>rs568536001</td>\n",
       "      <td>COL4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT           SGNF        PSNPE     PSNPS  PSNPC  \\\n",
       "0     13  110148891   C   G            NaN      neutral -1.000000   0.88   \n",
       "1     13  110148917   C   G            NaN      neutral -0.319012   0.74   \n",
       "2     13  110148920   G   C            NaN      neutral -0.320755   0.74   \n",
       "3     13  110148959   A   G  Likely benign  deleterious  0.456207   0.91   \n",
       "4     13  110148971   G   C            NaN      neutral -0.155078   0.73   \n",
       "\n",
       "         CADDE   CADDS  ...     FATS  FATC         FUNE                  FUNS  \\\n",
       "0      neutral   2.258  ...  0.18997  0.77          NaN                   NaN   \n",
       "1  deleterious   9.461  ...  0.18963  0.77      neutral   8.51599164992024E-6   \n",
       "2  deleterious   9.779  ...  0.21184  0.75            ?     0.622785975437062   \n",
       "3  deleterious  13.690  ...  0.82375  0.82      neutral  7.52439751165574E-20   \n",
       "4      neutral   6.450  ...  0.95264  0.91  deleterious      1.19368630042138   \n",
       "\n",
       "   FUNC   GWAVAE  GWAVAS GWAVAC           ID    GENE  \n",
       "0   NaN  neutral    0.30   0.68  rs552586867  COL4A1  \n",
       "1  0.80  neutral    0.32   0.71   rs59409892  COL4A1  \n",
       "2  0.49  neutral    0.32   0.71  rs535182970  COL4A1  \n",
       "3  0.80  neutral    0.44   0.53   rs56406633  COL4A1  \n",
       "4  0.74  neutral    0.44   0.53  rs568536001  COL4A1  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import gene and positional data from CSV file\n",
    "gene_positional_data = pd.read_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_consequences.csv\",\n",
    "    ),\n",
    "    usecols=[\"ID\", \"REF\", \"ALT\", \"POS\", \"GENE\"],\n",
    ")\n",
    "\n",
    "# Merge gene and positional data with frequency information\n",
    "pdtsnp2_vep_data = (\n",
    "    pdtsnp2_vep_data.merge(\n",
    "        gene_positional_data,\n",
    "        how=\"left\",\n",
    "        on=[\n",
    "            \"POS\",\n",
    "            \"REF\",\n",
    "            \"ALT\",\n",
    "        ],\n",
    "    )\n",
    "    .drop(columns=[\"ID_x\"])\n",
    "    .rename(columns={\"ID_y\": \"ID\"})\n",
    ")\n",
    "\n",
    "pdtsnp2_vep_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save data to CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdtsnp2_vep_data.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        constants.HOME_PATH,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_effects.csv\",\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
