{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of variant phenotype data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genetic variant phenotype association data was retrieved using [the Functional Annotation of Variants - Online Resource v2.0](https://favor.genohub.org/batch-annotation) for the variants identified in-house in African populations. The retrieved data was stored in `Data/PHENO/GRCh38-variant_phenotypes.tsv. \n",
    "\n",
    "The data was prepared for analysis by: \n",
    "Selecting features of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "PROJECT_ROOT = os.getenv(\"PROJECT_ROOT\")\n",
    "if PROJECT_ROOT not in sys.path:\n",
    "    os.chdir(PROJECT_ROOT + \"/Notebooks\")\n",
    "    sys.path.append(PROJECT_ROOT)\n",
    "\n",
    "import pandas as pd\n",
    "import Utils.constants as constants\n",
    "import Utils.functions as functions\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import variant phenotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>GenecodeComprehensiveCategory</th>\n",
       "      <th>GenecodeComprehensiveInfo</th>\n",
       "      <th>UcscInfo</th>\n",
       "      <th>Clnsig</th>\n",
       "      <th>Clnsigincl</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>Clndnincl</th>\n",
       "      <th>Clnrevstat</th>\n",
       "      <th>Clndisdb</th>\n",
       "      <th>Clndisdbincl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ENST00000375820.9,ENST00000649720.1,ENST000006...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>downstream</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ENST00000375820.9,ENST00000649720.1,ENST000006...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>COL4A1(ENST00000375820.10:c.*869G&gt;A,ENST000006...</td>\n",
       "      <td>ENST00000649720.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>13</td>\n",
       "      <td>110149715</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>COL4A1(ENST00000375820.10:c.*647_*646delAT,ENS...</td>\n",
       "      <td>ENST00000649720.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>13</td>\n",
       "      <td>110151168</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>intronic</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ENST00000649720.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf  Chromosome   Position         Rsid  \\\n",
       "0    13-110148917-C-G          13  110148917   rs59409892   \n",
       "1    13-110148891-C-G          13  110148891  rs552586867   \n",
       "2    13-110149494-C-T          13  110149494  rs552877576   \n",
       "3  13-110149715-AAT-A          13  110149715  rs886049952   \n",
       "4    13-110151168-C-T          13  110151168  rs557686466   \n",
       "\n",
       "  GenecodeComprehensiveCategory  \\\n",
       "0                    downstream   \n",
       "1                    downstream   \n",
       "2                          UTR3   \n",
       "3                          UTR3   \n",
       "4                      intronic   \n",
       "\n",
       "                           GenecodeComprehensiveInfo  \\\n",
       "0                                             COL4A1   \n",
       "1                                             COL4A1   \n",
       "2  COL4A1(ENST00000375820.10:c.*869G>A,ENST000006...   \n",
       "3  COL4A1(ENST00000375820.10:c.*647_*646delAT,ENS...   \n",
       "4                                             COL4A1   \n",
       "\n",
       "                                            UcscInfo Clnsig Clnsigincl Clndn  \\\n",
       "0  ENST00000375820.9,ENST00000649720.1,ENST000006...    NaN        NaN   NaN   \n",
       "1  ENST00000375820.9,ENST00000649720.1,ENST000006...    NaN        NaN   NaN   \n",
       "2                                  ENST00000649720.1    NaN        NaN   NaN   \n",
       "3                                  ENST00000649720.1    NaN        NaN   NaN   \n",
       "4                                  ENST00000649720.1    NaN        NaN   NaN   \n",
       "\n",
       "  Clndnincl Clnrevstat Clndisdb Clndisdbincl  \n",
       "0       NaN        NaN      NaN          NaN  \n",
       "1       NaN        NaN      NaN          NaN  \n",
       "2       NaN        NaN      NaN          NaN  \n",
       "3       NaN        NaN      NaN          NaN  \n",
       "4       NaN        NaN      NaN          NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_pheno = pd.read_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Raw\",\n",
    "        \"PHENO\",\n",
    "        \"GRCh38-Favorv2.0_ALLGENES.csv\",\n",
    "    )\n",
    ")\n",
    "\n",
    "variant_pheno.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select features of interest\n",
    "Select variant nomenclature, and associated clinical disease name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf         Rsid Clndn\n",
       "0    13-110148917-C-G   rs59409892   NaN\n",
       "1    13-110148891-C-G  rs552586867   NaN\n",
       "2    13-110149494-C-T  rs552877576   NaN\n",
       "3  13-110149715-AAT-A  rs886049952   NaN\n",
       "4    13-110151168-C-T  rs557686466   NaN"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_pheno_filtered = variant_pheno.copy()[['VariantVcf','Rsid','Clndn']]\n",
    "variant_pheno_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract variant reference and alternate allele information\n",
    "Extract variant chromosome, genomic position, reference and alternate allele information from VariantVCF column. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149715</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110151168</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf         Rsid Clndn CHROM        POS  REF ALT\n",
       "0    13-110148917-C-G   rs59409892   NaN    13  110148917    C   G\n",
       "1    13-110148891-C-G  rs552586867   NaN    13  110148891    C   G\n",
       "2    13-110149494-C-T  rs552877576   NaN    13  110149494    C   T\n",
       "3  13-110149715-AAT-A  rs886049952   NaN    13  110149715  AAT   A\n",
       "4    13-110151168-C-T  rs557686466   NaN    13  110151168    C   T"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_pheno_filtered[['CHROM','POS','REF','ALT']] = variant_pheno_filtered[\"VariantVcf\"].str.split(\"-\", expand=True)\n",
    "variant_pheno_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign a unique ID to each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add a column with a unique ID for each variant. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VariantVcf</th>\n",
       "      <th>Rsid</th>\n",
       "      <th>Clndn</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-110148917-C-G</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148917</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148917_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13-110148891-C-G</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110148891</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>110148891_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13-110149494-C-T</td>\n",
       "      <td>rs552877576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149494</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110149494_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13-110149715-AAT-A</td>\n",
       "      <td>rs886049952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110149715</td>\n",
       "      <td>AAT</td>\n",
       "      <td>A</td>\n",
       "      <td>110149715_AAT_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13-110151168-C-T</td>\n",
       "      <td>rs557686466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>110151168</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>110151168_C_T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           VariantVcf         Rsid Clndn CHROM        POS  REF ALT  \\\n",
       "0    13-110148917-C-G   rs59409892   NaN    13  110148917    C   G   \n",
       "1    13-110148891-C-G  rs552586867   NaN    13  110148891    C   G   \n",
       "2    13-110149494-C-T  rs552877576   NaN    13  110149494    C   T   \n",
       "3  13-110149715-AAT-A  rs886049952   NaN    13  110149715  AAT   A   \n",
       "4    13-110151168-C-T  rs557686466   NaN    13  110151168    C   T   \n",
       "\n",
       "                ID  \n",
       "0    110148917_C_G  \n",
       "1    110148891_C_G  \n",
       "2    110149494_C_T  \n",
       "3  110149715_AAT_A  \n",
       "4    110151168_C_T  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_pheno_filtered[\"ID\"] = (\n",
    "    variant_pheno_filtered[[\"POS\", \"REF\", \"ALT\"]].astype(\"str\").agg(\"_\".join, axis=1)\n",
    ")\n",
    "\n",
    "variant_pheno_filtered.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save variant phenotype data to a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_pheno_filtered.reset_index(drop=True).to_csv(\n",
    "    os.path.join(\n",
    "        PROJECT_ROOT,\n",
    "        \"Data\",\n",
    "        \"Processed\",\n",
    "        \"Variant_disease_phenotypes.csv\",\n",
    "    ),\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
